
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>TIGRAMITE &#8212; Tigramite 5.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinxdoc.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">Tigramite 5.1 documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">TIGRAMITE</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="tigramite">
<h1>TIGRAMITE<a class="headerlink" href="#tigramite" title="Permalink to this heading">¶</a></h1>
<p><a class="reference external" href="https://github.com/jakobrunge/tigramite/">Github repo</a></p>
<p>Tigramite is a causal time series analysis python package. It allows to efficiently estimate causal graphs from high-dimensional time series datasets (causal discovery) and to use these graphs for robust forecasting and the estimation and prediction of direct, total, and mediated effects. Causal discovery is based on linear as well as non-parametric conditional independence tests applicable to discrete or continuously-valued time series. Also includes functions for high-quality plots of the results. Please cite the following papers depending on which method you use:</p>
<ul class="simple">
<li><p>PCMCI: J. Runge, P. Nowack, M. Kretschmer, S. Flaxman, D. Sejdinovic, Detecting and quantifying causal associations in large nonlinear time series datasets. Sci. Adv. 5, eaau4996 (2019). <a class="reference external" href="https://advances.sciencemag.org/content/5/11/eaau4996">https://advances.sciencemag.org/content/5/11/eaau4996</a></p></li>
<li><p>PCMCI+: J. Runge (2020): Discovering contemporaneous and lagged causal relations in autocorrelated nonlinear time series datasets. Proceedings of the 36th Conference on Uncertainty in Artificial Intelligence, UAI 2020,Toronto, Canada, 2019, AUAI Press, 2020. <a class="reference external" href="http://auai.org/uai2020/proceedings/579_main_paper.pdf">http://auai.org/uai2020/proceedings/579_main_paper.pdf</a></p></li>
<li><p>LPCMCI: Gerhardus, A. &amp; Runge, J. High-recall causal discovery for autocorrelated time series with latent confounders Advances in Neural Information Processing Systems, 2020, 33. <a class="reference external" href="https://proceedings.neurips.cc/paper/2020/hash/94e70705efae423efda1088614128d0b-Abstract.html">https://proceedings.neurips.cc/paper/2020/hash/94e70705efae423efda1088614128d0b-Abstract.html</a></p></li>
<li><p>Generally: J. Runge (2018): Causal Network Reconstruction from Time Series: From Theoretical Assumptions to Practical Estimation. Chaos: An Interdisciplinary Journal of Nonlinear Science 28 (7): 075310. <a class="reference external" href="https://aip.scitation.org/doi/10.1063/1.5025050">https://aip.scitation.org/doi/10.1063/1.5025050</a></p></li>
<li><p>Nature Communications Perspective paper: <a class="reference external" href="https://www.nature.com/articles/s41467-019-10105-3">https://www.nature.com/articles/s41467-019-10105-3</a></p></li>
<li><p>Causal effects: J. Runge, Necessary and sufficient graphical conditions for optimal adjustment sets in causal graphical models with hidden variables, Advances in Neural Information Processing Systems, 2021, 34</p></li>
<li><p>Mediation class: J. Runge et al. (2015): Identifying causal gateways and mediators in complex spatio-temporal systems. Nature Communications, 6, 8502. <a class="reference external" href="http://doi.org/10.1038/ncomms9502">http://doi.org/10.1038/ncomms9502</a></p></li>
<li><p>Mediation class: J. Runge (2015): Quantifying information transfer and mediation along causal pathways in complex systems. Phys. Rev. E, 92(6), 62829. <a class="reference external" href="http://doi.org/10.1103/PhysRevE.92.062829">http://doi.org/10.1103/PhysRevE.92.062829</a></p></li>
<li><p>CMIknn: J. Runge (2018): Conditional Independence Testing Based on a Nearest-Neighbor Estimator of Conditional Mutual Information. In Proceedings of the 21st International Conference on Artificial Intelligence and Statistics. <a class="reference external" href="http://proceedings.mlr.press/v84/runge18a.html">http://proceedings.mlr.press/v84/runge18a.html</a></p></li>
</ul>
<div class="toctree-wrapper compound">
</div>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tigramite.pcmci.PCMCI" title="tigramite.pcmci.PCMCI"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tigramite.pcmci.PCMCI</span></code></a>(dataframe, cond_ind_test)</p></td>
<td><p>PCMCI causal discovery for time series datasets.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tigramite.lpcmci.LPCMCI" title="tigramite.lpcmci.LPCMCI"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tigramite.lpcmci.LPCMCI</span></code></a>(dataframe, cond_ind_test)</p></td>
<td><p>LPCMCI is an algorithm for causal discovery in large-scale times series that allows for latent confounders and learns lag-specific causal relationships.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tigramite.independence_tests.CondIndTest" title="tigramite.independence_tests.CondIndTest"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tigramite.independence_tests.CondIndTest</span></code></a>([...])</p></td>
<td><p>Base class of conditional independence tests.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tigramite.independence_tests.ParCorr" title="tigramite.independence_tests.ParCorr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tigramite.independence_tests.ParCorr</span></code></a>(**kwargs)</p></td>
<td><p>Partial correlation test.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tigramite.independence_tests.GPDC" title="tigramite.independence_tests.GPDC"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tigramite.independence_tests.GPDC</span></code></a>([...])</p></td>
<td><p>GPDC conditional independence test based on Gaussian processes and distance correlation.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tigramite.independence_tests.GPDCtorch" title="tigramite.independence_tests.GPDCtorch"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tigramite.independence_tests.GPDCtorch</span></code></a>([...])</p></td>
<td><p>GPDC conditional independence test based on Gaussian processes and distance correlation.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tigramite.independence_tests.CMIknn" title="tigramite.independence_tests.CMIknn"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tigramite.independence_tests.CMIknn</span></code></a>([knn, ...])</p></td>
<td><p>Conditional mutual information test based on nearest-neighbor estimator.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tigramite.independence_tests.CMIsymb" title="tigramite.independence_tests.CMIsymb"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tigramite.independence_tests.CMIsymb</span></code></a>([...])</p></td>
<td><p>Conditional mutual information test based on discrete estimator.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tigramite.independence_tests.OracleCI" title="tigramite.independence_tests.OracleCI"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tigramite.independence_tests.OracleCI</span></code></a>([...])</p></td>
<td><p>Oracle of conditional independence test X _|_ Y | Z given a graph.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tigramite.independence_tests.ParCorrMult" title="tigramite.independence_tests.ParCorrMult"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tigramite.independence_tests.ParCorrMult</span></code></a>([...])</p></td>
<td><p>Partial correlation test for multivariate X and Y.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tigramite.causal_effects.CausalEffects" title="tigramite.causal_effects.CausalEffects"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tigramite.causal_effects.CausalEffects</span></code></a>(...)</p></td>
<td><p>Causal effect estimation.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tigramite.models.Models" title="tigramite.models.Models"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tigramite.models.Models</span></code></a>(dataframe, model[, ...])</p></td>
<td><p>Base class for time series models.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tigramite.models.LinearMediation" title="tigramite.models.LinearMediation"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tigramite.models.LinearMediation</span></code></a>(dataframe)</p></td>
<td><p>Linear mediation analysis for time series models.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tigramite.models.Prediction" title="tigramite.models.Prediction"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tigramite.models.Prediction</span></code></a>(dataframe, ...)</p></td>
<td><p>Prediction class for time series models.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#module-tigramite.data_processing" title="tigramite.data_processing"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tigramite.data_processing</span></code></a></p></td>
<td><p>Tigramite data processing functions.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#module-tigramite.toymodels.structural_causal_processes" title="tigramite.toymodels.structural_causal_processes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tigramite.toymodels.structural_causal_processes</span></code></a></p></td>
<td><p>Tigramite toymodels.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="tigramite-pcmci-pcmci">
<h1><code class="xref py py-mod docutils literal notranslate"><span class="pre">tigramite.pcmci</span></code>: PCMCI<a class="headerlink" href="#tigramite-pcmci-pcmci" title="Permalink to this heading">¶</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tigramite.pcmci.</span></span><span class="sig-name descname"><span class="pre">PCMCI</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataframe</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cond_ind_test</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">selected_variables</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/pcmci.html#PCMCI"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.pcmci.PCMCI" title="Permalink to this definition">¶</a></dt>
<dd><p>PCMCI causal discovery for time series datasets.</p>
<p>PCMCI is a causal discovery framework for large-scale time series
datasets. This class contains several methods. The standard PCMCI method
addresses time-lagged causal discovery and is described in <a class="footnote-reference brackets" href="#id3" id="id1" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a> where
also further sub-variants are discussed. Lagged as well as contemporaneous
causal discovery is addressed with PCMCIplus and described in <a class="footnote-reference brackets" href="#id4" id="id2" role="doc-noteref"><span class="fn-bracket">[</span>5<span class="fn-bracket">]</span></a>. See the
tutorials for guidance in applying these methods.</p>
<p>PCMCI has:</p>
<ul class="simple">
<li><p>different conditional independence tests adapted to linear or
nonlinear dependencies, and continuously-valued or discrete data (
implemented in <code class="docutils literal notranslate"><span class="pre">tigramite.independence_tests</span></code>)</p></li>
<li><p>(mostly) hyperparameter optimization</p></li>
<li><p>easy parallelization (separate script)</p></li>
<li><p>handling of masked time series data</p></li>
<li><p>false discovery control and confidence interval estimation</p></li>
</ul>
<p class="rubric">Notes</p>
<a class="reference internal image-reference" href="_images/mci_schematic.png"><img alt="_images/mci_schematic.png" src="_images/mci_schematic.png" style="width: 200pt;" /></a>
<p>In the PCMCI framework, the dependency structure of a set of time series
variables is represented in a <em>time series graph</em> as shown in the Figure.
The nodes of a time series graph are defined as the variables at
different times and a link indicates a conditional dependency that can be
interpreted as a causal dependency under certain assumptions (see paper).
Assuming stationarity, the links are repeated in time. The parents
<img class="math" src="_images/math/fff7b4153a6590df59f8ed526be56220045b7f3b.png" alt="\mathcal{P}"/> of a variable are defined as the set of all nodes
with a link towards it (blue and red boxes in Figure).</p>
<p>The different PCMCI methods estimate causal links by iterative
conditional independence testing. PCMCI can be flexibly combined with
any kind of conditional independence test statistic adapted to the kind
of data (continuous or discrete) and its assumed dependency types.
These are available in <code class="docutils literal notranslate"><span class="pre">tigramite.independence_tests</span></code>.</p>
<p>NOTE: MCI test statistic values define a particular measure of causal
strength depending on the test statistic used. For example, ParCorr()
results in normalized values between -1 and 1. However, if you are
interested in quantifying causal effects, i.e., the effect of
hypothetical interventions, you may better look at the causal effect
estimation functionality of Tigramite.</p>
<p class="rubric">References</p>
<aside class="footnote brackets" id="id3" role="note">
<span class="label"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id1">1</a>,<a role="doc-backlink" href="#id5">2</a>,<a role="doc-backlink" href="#id6">3</a>,<a role="doc-backlink" href="#id7">4</a>,<a role="doc-backlink" href="#id9">5</a>,<a role="doc-backlink" href="#id15">6</a>,<a role="doc-backlink" href="#id16">7</a>,<a role="doc-backlink" href="#id17">8</a>,<a role="doc-backlink" href="#id31">9</a>,<a role="doc-backlink" href="#id32">10</a>)</span>
<p>J. Runge, P. Nowack, M. Kretschmer, S. Flaxman, D. Sejdinovic,
Detecting and quantifying causal associations in large nonlinear time
series datasets. Sci. Adv. 5, eaau4996 (2019)
<a class="reference external" href="https://advances.sciencemag.org/content/5/11/eaau4996">https://advances.sciencemag.org/content/5/11/eaau4996</a></p>
</aside>
<aside class="footnote brackets" id="id4" role="note">
<span class="label"><span class="fn-bracket">[</span>5<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id2">1</a>,<a role="doc-backlink" href="#id8">2</a>,<a role="doc-backlink" href="#id10">3</a>,<a role="doc-backlink" href="#id11">4</a>,<a role="doc-backlink" href="#id12">5</a>)</span>
<p>J. Runge,
Discovering contemporaneous and lagged causal relations in
autocorrelated nonlinear time series datasets
<a class="reference external" href="http://www.auai.org/~w-auai/uai2020/proceedings/579_main_paper.pdf">http://www.auai.org/~w-auai/uai2020/proceedings/579_main_paper.pdf</a></p>
</aside>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dataframe</strong> (<em>data object</em>) – This is the Tigramite dataframe object. Among others, it has the
attributes dataframe.values yielding a numpy array of shape (
observations T, variables N) and optionally a mask of the same shape.</p></li>
<li><p><strong>cond_ind_test</strong> (<em>conditional independence test object</em>) – This can be ParCorr or other classes from
<code class="docutils literal notranslate"><span class="pre">tigramite.independence_tests</span></code> or an external test passed as a
callable. This test can be based on the class
tigramite.independence_tests.CondIndTest.</p></li>
<li><p><strong>selected_variables</strong> (<em>list</em>) – Deprecated, just here to raise Error if not None.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Verbose levels 0, 1, …</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI.all_parents">
<span class="sig-name descname"><span class="pre">all_parents</span></span><a class="headerlink" href="#tigramite.pcmci.PCMCI.all_parents" title="Permalink to this definition">¶</a></dt>
<dd><p>Dictionary of form {0:[(0, -1), (3, -2), …], 1:[], …} containing
the conditioning-parents estimated with PC algorithm.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dictionary</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI.val_min">
<span class="sig-name descname"><span class="pre">val_min</span></span><a class="headerlink" href="#tigramite.pcmci.PCMCI.val_min" title="Permalink to this definition">¶</a></dt>
<dd><p>Dictionary of form val_min[j][(i, -tau)] = float
containing the minimum test statistic value for each link estimated in
the PC algorithm.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dictionary</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI.pval_max">
<span class="sig-name descname"><span class="pre">pval_max</span></span><a class="headerlink" href="#tigramite.pcmci.PCMCI.pval_max" title="Permalink to this definition">¶</a></dt>
<dd><p>Dictionary of form pval_max[j][(i, -tau)] = float containing the maximum
p-value for each link estimated in the PC algorithm.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dictionary</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI.iterations">
<span class="sig-name descname"><span class="pre">iterations</span></span><a class="headerlink" href="#tigramite.pcmci.PCMCI.iterations" title="Permalink to this definition">¶</a></dt>
<dd><p>Dictionary containing further information on algorithm steps.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dictionary</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI.N">
<span class="sig-name descname"><span class="pre">N</span></span><a class="headerlink" href="#tigramite.pcmci.PCMCI.N" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of variables.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI.T">
<span class="sig-name descname"><span class="pre">T</span></span><a class="headerlink" href="#tigramite.pcmci.PCMCI.T" title="Permalink to this definition">¶</a></dt>
<dd><p>Time series sample length of dataset(s).</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI.convert_to_string_graph">
<span class="sig-name descname"><span class="pre">convert_to_string_graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph_bool</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/pcmci.html#PCMCI.convert_to_string_graph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.pcmci.PCMCI.convert_to_string_graph" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts the 0,1-based graph returned by PCMCI to a string array
with links ‘–&gt;’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>graph_bool</strong> (<em>array</em>) – 0,1-based graph array output by PCMCI.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>graph</strong> – graph as string array with links ‘–&gt;’.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>array</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI.get_corrected_pvalues">
<span class="sig-name descname"><span class="pre">get_corrected_pvalues</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">p_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fdr_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fdr_bh'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_contemporaneous</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_min</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">selected_links</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/pcmci.html#PCMCI.get_corrected_pvalues"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.pcmci.PCMCI.get_corrected_pvalues" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns p-values corrected for multiple testing.</p>
<p>Currently implemented is Benjamini-Hochberg False Discovery Rate
method. Correction is performed either among all links if
exclude_contemporaneous==False, or only among lagged links.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>p_matrix</strong> (<em>array-like</em>) – Matrix of p-values. Must be of shape (N, N, tau_max + 1).</p></li>
<li><p><strong>tau_min</strong> (<em>int</em><em>, </em><em>default: 0</em>) – Minimum time lag. Only used as consistency check of selected_links.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em><em>, </em><em>default: 1</em>) – Maximum time lag. Must be larger or equal to tau_min. Only used as
consistency check of selected_links.</p></li>
<li><p><strong>selected_links</strong> (<em>dict</em><em> or </em><em>None</em>) – Dictionary of form {0: [(3, -2), …], 1:[], …}
specifying whether only selected links should be tested. If None is
passed, all links are tested.</p></li>
<li><p><strong>fdr_method</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'fdr_bh'</em><em>)</em>) – Correction method, currently implemented is Benjamini-Hochberg
False Discovery Rate method.</p></li>
<li><p><strong>exclude_contemporaneous</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: True</em><em>)</em>) – Whether to include contemporaneous links in correction.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>q_matrix</strong> – Matrix of shape (N, N, tau_max + 1) containing corrected p-values.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>array-like</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI.get_graph_from_pmatrix">
<span class="sig-name descname"><span class="pre">get_graph_from_pmatrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">p_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_level</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_min</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">selected_links</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/pcmci.html#PCMCI.get_graph_from_pmatrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.pcmci.PCMCI.get_graph_from_pmatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Construct graph from thresholding the p_matrix at an alpha-level.</p>
<p>Allows to take into account selected_links.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>p_matrix</strong> (<em>array of shape</em><em> [</em><em>N</em><em>, </em><em>N</em><em>, </em><em>tau_max+1</em><em>]</em>) – Estimated matrix of p-values, optionally adjusted if fdr_method is
not ‘none’.</p></li>
<li><p><strong>alpha_level</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 0.05</em><em>)</em>) – Significance level at which the p_matrix is thresholded to
get graph.</p></li>
<li><p><strong>tau_mix</strong> (<em>int</em>) – Minimum time delay to test.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em>) – Maximum time delay to test.</p></li>
<li><p><strong>selected_links</strong> (<em>dict</em><em> or </em><em>None</em>) – Dictionary of form {0: [(3, -2), …], 1:[], …}
specifying whether only selected links should be tested. If None is
passed, all links are tested.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>graph</strong> – Causal graph, see description above for interpretation.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>array of shape [N, N, tau_max+1]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI.get_lagged_dependencies">
<span class="sig-name descname"><span class="pre">get_lagged_dependencies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">selected_links</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_min</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fdr_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'none'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/pcmci.html#PCMCI.get_lagged_dependencies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.pcmci.PCMCI.get_lagged_dependencies" title="Permalink to this definition">¶</a></dt>
<dd><p>Unconditional lagged independence tests.</p>
<p>Implements the unconditional lagged independence test (see [ 1]_).</p>
<p>Returns the matrices of test statistic values, (optionally corrected)
p-values, and (optionally) confidence intervals. Also (new in 4.3)
returns graph based on alpha_level (and optional FDR-correction).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>selected_links</strong> (<em>dict</em><em> or </em><em>None</em>) – Dictionary of form {0: [(3, -2), …], 1:[], …}
specifying whether only selected links should be tested. If None is
passed, all links are tested.</p></li>
<li><p><strong>tau_min</strong> (<em>int</em><em>, </em><em>default: 0</em>) – Minimum time lag to test. Note that zero-lags are undirected.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em><em>, </em><em>default: 1</em>) – Maximum time lag. Must be larger or equal to tau_min.</p></li>
<li><p><strong>val_only</strong> (<em>bool</em><em>, </em><em>default: False</em>) – Option to only compute dependencies and not p-values.</p></li>
<li><p><strong>alpha_level</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 0.05</em><em>)</em>) – Significance level at which the p_matrix is thresholded to
get graph.</p></li>
<li><p><strong>fdr_method</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'none'</em><em>)</em>) – Correction method, currently implemented is Benjamini-Hochberg
False Discovery Rate method (‘fdr_bh’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>graph</strong> (<em>array of shape [N, N, tau_max+1]</em>) – Causal graph, see description above for interpretation.</p></li>
<li><p><strong>val_matrix</strong> (<em>array of shape [N, N, tau_max+1]</em>) – Estimated matrix of test statistic values.</p></li>
<li><p><strong>p_matrix</strong> (<em>array of shape [N, N, tau_max+1]</em>) – Estimated matrix of p-values, optionally adjusted if fdr_method is
not ‘none’.</p></li>
<li><p><strong>conf_matrix</strong> (<em>array of shape [N, N, tau_max+1,2]</em>) – Estimated matrix of confidence intervals of test statistic values.
Only computed if set in cond_ind_test, where also the percentiles
are set.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI.print_results">
<span class="sig-name descname"><span class="pre">print_results</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">return_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/pcmci.html#PCMCI.print_results"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.pcmci.PCMCI.print_results" title="Permalink to this definition">¶</a></dt>
<dd><p>Prints significant parents from output of MCI or PCMCI algorithms.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>return_dict</strong> (<em>dict</em>) – <dl class="simple">
<dt>Dictionary of return values, containing keys</dt><dd><ul>
<li><p>’p_matrix’</p></li>
<li><p>’val_matrix’</p></li>
<li><p>’conf_matrix’</p></li>
</ul>
</dd>
</dl>
</p></li>
<li><p><strong>alpha_level</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 0.05</em><em>)</em>) – Significance level.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI.print_significant_links">
<span class="sig-name descname"><span class="pre">print_significant_links</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">p_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conf_matrix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ambiguous_triples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/pcmci.html#PCMCI.print_significant_links"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.pcmci.PCMCI.print_significant_links" title="Permalink to this definition">¶</a></dt>
<dd><p>Prints significant links.</p>
<p>Used for output of PCMCI and PCMCIplus. For the latter also information
on ambiguous links and conflicts is returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alpha_level</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 0.05</em><em>)</em>) – Significance level.</p></li>
<li><p><strong>p_matrix</strong> (<em>array-like</em>) – Must be of shape (N, N, tau_max + 1).</p></li>
<li><p><strong>val_matrix</strong> (<em>array-like</em>) – Must be of shape (N, N, tau_max + 1).</p></li>
<li><p><strong>conf_matrix</strong> (<em>array-like</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Matrix of confidence intervals of shape (N, N, tau_max+1, 2).</p></li>
<li><p><strong>graph</strong> (<em>array-like</em>) – Must be of shape (N, N, tau_max + 1).</p></li>
<li><p><strong>ambiguous_triples</strong> (<em>list</em>) – List of ambiguous triples.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI.return_parents_dict">
<span class="sig-name descname"><span class="pre">return_parents_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_lagzero_parents</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/pcmci.html#PCMCI.return_parents_dict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.pcmci.PCMCI.return_parents_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns dictionary of parents sorted by val_matrix.</p>
<p>If parents are unclear (link with o or x), then no parent
is returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>graph</strong> (<em>array of shape</em><em> [</em><em>N</em><em>, </em><em>N</em><em>, </em><em>tau_max+1</em><em>]</em>) – Causal graph, see description above for interpretation.</p></li>
<li><p><strong>val_matrix</strong> (<em>array-like</em>) – Matrix of test statistic values. Must be of shape (N, N, tau_max +
1).</p></li>
<li><p><strong>include_lagzero_parents</strong> (<em>bool</em><em> (</em><em>default: False</em><em>)</em>) – Whether the dictionary should also return parents at lag
zero.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>parents_dict</strong> – Dictionary of form {0:[(0, -1), (3, -2), …], 1:[], …}
containing estimated parents.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI.return_significant_links">
<span class="sig-name descname"><span class="pre">return_significant_links</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pq_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_lagzero_links</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/pcmci.html#PCMCI.return_significant_links"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.pcmci.PCMCI.return_significant_links" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns list of significant links as well as a boolean matrix.</p>
<p>DEPRECATED. Will be removed in future.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI.run_bivci">
<span class="sig-name descname"><span class="pre">run_bivci</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">selected_links</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_min</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fdr_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'none'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/pcmci.html#PCMCI.run_bivci"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.pcmci.PCMCI.run_bivci" title="Permalink to this definition">¶</a></dt>
<dd><p>BivCI conditional independence tests.</p>
<p>Implements the BivCI test (see <a class="footnote-reference brackets" href="#id3" id="id5" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a>).</p>
<p>Returns the matrices of test statistic values, (optionally corrected)
p-values, and (optionally) confidence intervals. Also (new in 4.3)
returns graph based on alpha_level (and optional FDR-correction).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>selected_links</strong> (<em>dict</em><em> or </em><em>None</em>) – Dictionary of form {0: [(3, -2), …], 1:[], …}
specifying whether only selected links should be tested. If None is
passed, all links are tested.</p></li>
<li><p><strong>tau_min</strong> (<em>int</em><em>, </em><em>default: 0</em>) – Minimum time lag to test. Note that zero-lags are undirected.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em><em>, </em><em>default: 1</em>) – Maximum time lag. Must be larger or equal to tau_min.</p></li>
<li><p><strong>val_only</strong> (<em>bool</em><em>, </em><em>default: False</em>) – Option to only compute dependencies and not p-values.</p></li>
<li><p><strong>alpha_level</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 0.05</em><em>)</em>) – Significance level at which the p_matrix is thresholded to
get graph.</p></li>
<li><p><strong>fdr_method</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'fdr_bh'</em><em>)</em>) – Correction method, currently implemented is Benjamini-Hochberg
False Discovery Rate method.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>graph</strong> (<em>array of shape [N, N, tau_max+1]</em>) – Causal graph, see description above for interpretation.</p></li>
<li><p><strong>val_matrix</strong> (<em>array of shape [N, N, tau_max+1]</em>) – Estimated matrix of test statistic values.</p></li>
<li><p><strong>p_matrix</strong> (<em>array of shape [N, N, tau_max+1]</em>) – Estimated matrix of p-values, optionally adjusted if fdr_method is
not ‘none’.</p></li>
<li><p><strong>conf_matrix</strong> (<em>array of shape [N, N, tau_max+1,2]</em>) – Estimated matrix of confidence intervals of test statistic values.
Only computed if set in cond_ind_test, where also the percentiles
are set.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI.run_fullci">
<span class="sig-name descname"><span class="pre">run_fullci</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">selected_links</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_min</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fdr_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'none'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/pcmci.html#PCMCI.run_fullci"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.pcmci.PCMCI.run_fullci" title="Permalink to this definition">¶</a></dt>
<dd><p>FullCI conditional independence tests.</p>
<p>Implements the FullCI test (see <a class="footnote-reference brackets" href="#id3" id="id6" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a>).</p>
<p>Returns the matrices of test statistic values, (optionally corrected)
p-values, and (optionally) confidence intervals. Also (new in 4.3)
returns graph based on alpha_level (and optional FDR-correction).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>selected_links</strong> (<em>dict</em><em> or </em><em>None</em>) – Dictionary of form {0: [(3, -2), …], 1:[], …}
specifying whether only selected links should be tested. If None is
passed, all links are tested.</p></li>
<li><p><strong>tau_min</strong> (<em>int</em><em>, </em><em>default: 0</em>) – Minimum time lag to test. Note that zero-lags are undirected.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em><em>, </em><em>default: 1</em>) – Maximum time lag. Must be larger or equal to tau_min.</p></li>
<li><p><strong>val_only</strong> (<em>bool</em><em>, </em><em>default: False</em>) – Option to only compute dependencies and not p-values.</p></li>
<li><p><strong>alpha_level</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 0.05</em><em>)</em>) – Significance level at which the p_matrix is thresholded to
get graph.</p></li>
<li><p><strong>fdr_method</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'none'</em><em>)</em>) – Correction method, currently implemented is Benjamini-Hochberg
False Discovery Rate method (‘fdr_bh’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>graph</strong> (<em>array of shape [N, N, tau_max+1]</em>) – Causal graph, see description above for interpretation.</p></li>
<li><p><strong>val_matrix</strong> (<em>array of shape [N, N, tau_max+1]</em>) – Estimated matrix of test statistic values.</p></li>
<li><p><strong>p_matrix</strong> (<em>array of shape [N, N, tau_max+1]</em>) – Estimated matrix of p-values, optionally adjusted if fdr_method is
not ‘none’.</p></li>
<li><p><strong>conf_matrix</strong> (<em>array of shape [N, N, tau_max+1,2]</em>) – Estimated matrix of confidence intervals of test statistic values.
Only computed if set in cond_ind_test, where also the percentiles
are set.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI.run_mci">
<span class="sig-name descname"><span class="pre">run_mci</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">selected_links</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_min</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parents</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_conds_py</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_conds_px</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fdr_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'none'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/pcmci.html#PCMCI.run_mci"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.pcmci.PCMCI.run_mci" title="Permalink to this definition">¶</a></dt>
<dd><p>MCI conditional independence tests.</p>
<p>Implements the MCI test (Algorithm 2 in <a class="footnote-reference brackets" href="#id3" id="id7" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a>).</p>
<p>Returns the matrices of test statistic values, (optionally corrected)
p-values, and (optionally) confidence intervals. Also (new in 4.3)
returns graph based on alpha_level (and optional FDR-correction).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>selected_links</strong> (<em>dict</em><em> or </em><em>None</em>) – Dictionary of form {0: [(3, -2), …], 1:[], …}
specifying whether only selected links should be tested. If None is
passed, all links are tested.</p></li>
<li><p><strong>tau_min</strong> (<em>int</em><em>, </em><em>default: 0</em>) – Minimum time lag to test. Note that zero-lags are undirected.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em><em>, </em><em>default: 1</em>) – Maximum time lag. Must be larger or equal to tau_min.</p></li>
<li><p><strong>parents</strong> (<em>dict</em><em> or </em><em>None</em>) – Dictionary of form {0:[(0, -1), (3, -2), …], 1:[], …}
specifying the conditions for each variable. If None is
passed, no conditions are used.</p></li>
<li><p><strong>max_conds_py</strong> (<em>int</em><em> or </em><em>None</em>) – Maximum number of conditions of Y to use. If None is passed, this
number is unrestricted.</p></li>
<li><p><strong>max_conds_px</strong> (<em>int</em><em> or </em><em>None</em>) – Maximum number of conditions of Z to use. If None is passed, this
number is unrestricted.</p></li>
<li><p><strong>val_only</strong> (<em>bool</em><em>, </em><em>default: False</em>) – Option to only compute dependencies and not p-values.</p></li>
<li><p><strong>alpha_level</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 0.05</em><em>)</em>) – Significance level at which the p_matrix is thresholded to
get graph.</p></li>
<li><p><strong>fdr_method</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'none'</em><em>)</em>) – Correction method, currently implemented is Benjamini-Hochberg
False Discovery Rate method (‘fdr_bh’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>graph</strong> (<em>array of shape [N, N, tau_max+1]</em>) – Causal graph, see description above for interpretation.</p></li>
<li><p><strong>val_matrix</strong> (<em>array of shape [N, N, tau_max+1]</em>) – Estimated matrix of test statistic values.</p></li>
<li><p><strong>p_matrix</strong> (<em>array of shape [N, N, tau_max+1]</em>) – Estimated matrix of p-values, optionally adjusted if fdr_method is
not ‘none’.</p></li>
<li><p><strong>conf_matrix</strong> (<em>array of shape [N, N, tau_max+1,2]</em>) – Estimated matrix of confidence intervals of test statistic values.
Only computed if set in cond_ind_test, where also the percentiles
are set.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI.run_pc_stable">
<span class="sig-name descname"><span class="pre">run_pc_stable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">selected_links</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_min</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pc_alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_conds_dim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_combinations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/pcmci.html#PCMCI.run_pc_stable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.pcmci.PCMCI.run_pc_stable" title="Permalink to this definition">¶</a></dt>
<dd><p>Lagged PC algorithm for estimating lagged parents of all variables.</p>
<p>Parents are made available as self.all_parents</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>selected_links</strong> (<em>dict</em><em> or </em><em>None</em>) – Dictionary of form {0:[(0, -1), (3, -2), …], 1:[], …}
specifying whether only selected links should be tested. If None is
passed, all links are tested.</p></li>
<li><p><strong>tau_min</strong> (<em>int</em><em>, </em><em>default: 1</em>) – Minimum time lag to test. Useful for multi-step ahead predictions.
Must be greater zero.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em><em>, </em><em>default: 1</em>) – Maximum time lag. Must be larger or equal to tau_min.</p></li>
<li><p><strong>save_iterations</strong> (<em>bool</em><em>, </em><em>default: False</em>) – Whether to save iteration step results such as conditions used.</p></li>
<li><p><strong>pc_alpha</strong> (<em>float</em><em> or </em><em>list of floats</em><em>, </em><em>default:</em><em> [</em><em>0.05</em><em>, </em><em>0.1</em><em>, </em><em>0.2</em><em>, </em><em>...</em><em>, </em><em>0.5</em><em>]</em>) – Significance level in algorithm. If a list or None is passed, the
pc_alpha level is optimized for every variable across the given
pc_alpha values using the score computed in
cond_ind_test.get_model_selection_criterion().</p></li>
<li><p><strong>max_conds_dim</strong> (<em>int</em><em> or </em><em>None</em>) – Maximum number of conditions to test. If None is passed, this number
is unrestricted.</p></li>
<li><p><strong>max_combinations</strong> (<em>int</em><em>, </em><em>default: 1</em>) – Maximum number of combinations of conditions of current cardinality
to test. Defaults to 1 for PC_1 algorithm. For original PC algorithm
a larger number, such as 10, can be used.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>all_parents</strong> – Dictionary of form {0:[(0, -1), (3, -2), …], 1:[], …}
containing estimated parents.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI.run_pcalg">
<span class="sig-name descname"><span class="pre">run_pcalg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">selected_links</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pc_alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.01</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_min</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_conds_dim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_combinations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lagged_parents</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_conds_py</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_conds_px</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_conds_px_lagged</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'standard'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contemp_collider_rule</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'majority'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conflict_resolution</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/pcmci.html#PCMCI.run_pcalg"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.pcmci.PCMCI.run_pcalg" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs PC algorithm for time-lagged and contemporaneous causal
discovery for time series.</p>
<p>For <code class="docutils literal notranslate"><span class="pre">mode='contemp_conds'</span></code> this implements Steps 2-4 of the
PCMCIplus method described in <a class="footnote-reference brackets" href="#id4" id="id8" role="doc-noteref"><span class="fn-bracket">[</span>5<span class="fn-bracket">]</span></a>. For <code class="docutils literal notranslate"><span class="pre">mode='standard'</span></code> this
implements the standard PC algorithm adapted to time series.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>selected_links</strong> (<em>dict</em><em> or </em><em>None</em>) – Dictionary of form {0: [(3, 0), (0, -1), …], 1:[], …}
specifying whether only selected links should be tested. If None is
passed, all links are tested.</p></li>
<li><p><strong>lagged_parents</strong> (<em>dictionary</em>) – Dictionary of form {0:[(0, -1), (3, -2), …], 1:[], …} containing
additional conditions for each CI test. As part of PCMCIplus
these are the superset of lagged parents estimated with the PC1
algorithm.</p></li>
<li><p><strong>mode</strong> (<em>{'standard'</em><em>, </em><em>'contemp_conds'}</em>) – For <code class="docutils literal notranslate"><span class="pre">mode='contemp_conds'</span></code> this implements Steps 2-4 of the
PCMCIplus method. For <code class="docutils literal notranslate"><span class="pre">mode='standard'</span></code> this implements the
standard PC algorithm adapted to time series.</p></li>
<li><p><strong>tau_min</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Minimum time lag to test.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 1</em><em>)</em>) – Maximum time lag. Must be larger or equal to tau_min.</p></li>
<li><p><strong>pc_alpha</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 0.01</em><em>)</em>) – Significance level.</p></li>
<li><p><strong>contemp_collider_rule</strong> (<em>{'majority'</em><em>, </em><em>'conservative'</em><em>, </em><em>'none'}</em>) – Rule for collider phase to use. See the paper for details. Only
‘majority’ and ‘conservative’ lead to an order-independent
algorithm.</p></li>
<li><p><strong>conflict_resolution</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: True</em><em>)</em>) – Whether to mark conflicts in orientation rules. Only for True
this leads to an order-independent algorithm.</p></li>
<li><p><strong>max_conds_dim</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Maximum number of conditions to test. If None is passed, this number
is unrestricted.</p></li>
<li><p><strong>max_combinations</strong> (<em>int</em>) – Maximum number of combinations of conditions of current cardinality
to test.</p></li>
<li><p><strong>max_conds_py</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Maximum number of lagged conditions of Y to use in MCI tests. If
None is passed, this number is unrestricted.</p></li>
<li><p><strong>max_conds_px</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Maximum number of lagged conditions of X to use in MCI tests. If
None is passed, this number is unrestricted.</p></li>
<li><p><strong>max_conds_px_lagged</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Maximum number of lagged conditions of X when X is lagged in MCI
tests. If None is passed, this number is equal to max_conds_px.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>graph</strong> (<em>array of shape [N, N, tau_max+1]</em>) – Resulting causal graph, see description above for interpretation.</p></li>
<li><p><strong>val_matrix</strong> (<em>array of shape [N, N, tau_max+1]</em>) – Estimated matrix of test statistic values regarding adjacencies.</p></li>
<li><p><strong>p_matrix</strong> (<em>array of shape [N, N, tau_max+1]</em>) – Estimated matrix of p-values regarding adjacencies.</p></li>
<li><p><strong>sepset</strong> (<em>dictionary</em>) – Separating sets. See paper for details.</p></li>
<li><p><strong>ambiguous_triples</strong> (<em>list</em>) – List of ambiguous triples, only relevant for ‘majority’ and
‘conservative’ rules, see paper for details.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI.run_pcalg_non_timeseries_data">
<span class="sig-name descname"><span class="pre">run_pcalg_non_timeseries_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pc_alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.01</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_conds_dim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_combinations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contemp_collider_rule</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'majority'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conflict_resolution</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/pcmci.html#PCMCI.run_pcalg_non_timeseries_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.pcmci.PCMCI.run_pcalg_non_timeseries_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs PC algorithm for non-time series data.</p>
<p>Simply calls run_pcalg with tau_min = tau_max = 0.
Removes lags from ouput dictionaries.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pc_alpha</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 0.01</em><em>)</em>) – Significance level.</p></li>
<li><p><strong>contemp_collider_rule</strong> (<em>{'majority'</em><em>, </em><em>'conservative'</em><em>, </em><em>'none'}</em>) – Rule for collider phase to use. See the paper for details. Only
‘majority’ and ‘conservative’ lead to an order-independent
algorithm.</p></li>
<li><p><strong>conflict_resolution</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: True</em><em>)</em>) – Whether to mark conflicts in orientation rules. Only for True
this leads to an order-independent algorithm.</p></li>
<li><p><strong>max_conds_dim</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Maximum number of conditions to test. If None is passed, this number
is unrestricted.</p></li>
<li><p><strong>max_combinations</strong> (<em>int</em>) – Maximum number of combinations of conditions of current cardinality
to test.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>graph</strong> (<em>array of shape [N, N, 1]</em>) – Resulting causal graph, see description above for interpretation.</p></li>
<li><p><strong>val_matrix</strong> (<em>array of shape [N, N, 1]</em>) – Estimated matrix of test statistic values regarding adjacencies.</p></li>
<li><p><strong>p_matrix</strong> (<em>array of shape [N, N, 1]</em>) – Estimated matrix of p-values regarding adjacencies.</p></li>
<li><p><strong>sepset</strong> (<em>dictionary</em>) – Separating sets. See paper for details.</p></li>
<li><p><strong>ambiguous_triples</strong> (<em>list</em>) – List of ambiguous triples, only relevant for ‘majority’ and
‘conservative’ rules, see paper for details.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI.run_pcmci">
<span class="sig-name descname"><span class="pre">run_pcmci</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">selected_links</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_min</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pc_alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_conds_dim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_combinations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_conds_py</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_conds_px</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fdr_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'none'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/pcmci.html#PCMCI.run_pcmci"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.pcmci.PCMCI.run_pcmci" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs PCMCI time-lagged causal discovery for time series.</p>
<p>Wrapper around PC-algorithm function and MCI function.</p>
<p class="rubric">Notes</p>
<p>The PCMCI causal discovery method is comprehensively described in [
1]_, where also analytical and numerical results are presented. Here
we briefly summarize the method.</p>
<p>PCMCI estimates time-lagged causal links by a two-step procedure:</p>
<ol class="arabic simple">
<li><p>Condition-selection: For each variable <img class="math" src="_images/math/e3fc28292267f066fee7718c64f4bbfece521f24.png" alt="j"/>, estimate a
<em>superset</em> of parents <img class="math" src="_images/math/c4ed4f54ee00448249d2df22ad67f3e281df085b.png" alt="\tilde{\mathcal{P}}(X^j_t)"/> with the
iterative PC1 algorithm, implemented as <code class="docutils literal notranslate"><span class="pre">run_pc_stable</span></code>. The
condition-selection step reduces the dimensionality and avoids
conditioning on irrelevant variables.</p></li>
<li><p><em>Momentary conditional independence</em> (MCI)</p></li>
</ol>
<div class="math">
<p><img src="_images/math/e676f7877e03a3707387149ce203c5e79ed8f9a9.png" alt="X^i_{t-\tau} \perp X^j_{t} | \tilde{\mathcal{P}}(
X^j_t), \tilde{\mathcal{P}}(X^i_{t-\tau})"/></p>
</div><p>here implemented as <code class="docutils literal notranslate"><span class="pre">run_mci</span></code>. This step estimates the p-values and
test statistic values for all links accounting for common drivers,
indirect links, and autocorrelation.</p>
<p>NOTE: MCI test statistic values define a particular measure of causal
strength depending on the test statistic used. For example, ParCorr()
results in normalized values between -1 and 1. However, if you are
interested in quantifying causal effects, i.e., the effect of
hypothetical interventions, you may better look at the causal effect
estimation functionality of Tigramite.</p>
<p>PCMCI can be flexibly combined with any kind of conditional
independence test statistic adapted to the kind of data (continuous
or discrete) and its assumed dependency types. These are available in
<code class="docutils literal notranslate"><span class="pre">tigramite.independence_tests</span></code>.</p>
<p>The main free parameters of PCMCI (in addition to free parameters of
the conditional independence test statistic) are the maximum time
delay <img class="math" src="_images/math/2bf86ca220f43e569c6c7aefaf32742919222e6e.png" alt="\tau_{\max}"/> (<code class="docutils literal notranslate"><span class="pre">tau_max</span></code>) and the significance
threshold in the condition-selection step <img class="math" src="_images/math/2f5aa019312e1bbc969deab8dca8b00f76025404.png" alt="\alpha"/> (
<code class="docutils literal notranslate"><span class="pre">pc_alpha</span></code>). The maximum time delay depends on the application and
should be chosen according to the maximum causal time lag expected in
the complex system. We recommend a rather large choice that includes
peaks in the <code class="docutils literal notranslate"><span class="pre">get_lagged_dependencies</span></code> function. <img class="math" src="_images/math/2f5aa019312e1bbc969deab8dca8b00f76025404.png" alt="\alpha"/>
should not be seen as a significance test level in the
condition-selection step since the iterative hypothesis tests do not
allow for a precise assessment. <img class="math" src="_images/math/2f5aa019312e1bbc969deab8dca8b00f76025404.png" alt="\alpha"/> rather takes the role
of a regularization parameter in model-selection techniques. If a
list of values is given or <code class="docutils literal notranslate"><span class="pre">pc_alpha=None</span></code>, <img class="math" src="_images/math/2f5aa019312e1bbc969deab8dca8b00f76025404.png" alt="\alpha"/> is
optimized using model selection criteria implemented in the respective
<code class="docutils literal notranslate"><span class="pre">tigramite.independence_tests</span></code>.</p>
<p>Further optional parameters are discussed in <a class="footnote-reference brackets" href="#id3" id="id9" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a>.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">tigramite.pcmci</span> <span class="kn">import</span> <span class="n">PCMCI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">tigramite.independence_tests</span> <span class="kn">import</span> <span class="n">ParCorr</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">tigramite.data_processing</span> <span class="k">as</span> <span class="nn">pp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">tigramite.toymodels</span> <span class="kn">import</span> <span class="n">structural_causal_processes</span> <span class="k">as</span> <span class="n">toys</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">7</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Example process to play around with</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Each key refers to a variable and the incoming links are supplied</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># as a list of format [((driver, -lag), coeff), ...]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">links_coeffs</span> <span class="o">=</span> <span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="p">[((</span><span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">),</span> <span class="mf">0.8</span><span class="p">)],</span>
<span class="go">                    1: [((1, -1), 0.8), ((0, -1), 0.5)],</span>
<span class="go">                    2: [((2, -1), 0.8), ((1, -2), -0.6)]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">toys</span><span class="o">.</span><span class="n">var_process</span><span class="p">(</span><span class="n">links_coeffs</span><span class="p">,</span> <span class="n">T</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Data must be array of shape (time, variables)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="go">(1000, 3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dataframe</span> <span class="o">=</span> <span class="n">pp</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cond_ind_test</span> <span class="o">=</span> <span class="n">ParCorr</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcmci</span> <span class="o">=</span> <span class="n">PCMCI</span><span class="p">(</span><span class="n">dataframe</span><span class="o">=</span><span class="n">dataframe</span><span class="p">,</span> <span class="n">cond_ind_test</span><span class="o">=</span><span class="n">cond_ind_test</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">results</span> <span class="o">=</span> <span class="n">pcmci</span><span class="o">.</span><span class="n">run_pcmci</span><span class="p">(</span><span class="n">tau_max</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">pc_alpha</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcmci</span><span class="o">.</span><span class="n">print_significant_links</span><span class="p">(</span><span class="n">p_matrix</span><span class="o">=</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;p_matrix&#39;</span><span class="p">],</span>
<span class="go">                                 val_matrix=results[&#39;val_matrix&#39;],</span>
<span class="go">                                 alpha_level=0.05)</span>
<span class="go">## Significant parents at alpha = 0.05:</span>
</pre></div>
</div>
<blockquote>
<div><dl class="simple">
<dt>Variable 0 has 1 link(s):</dt><dd><p>(0 -1): pval = 0.00000 | val =  0.588</p>
</dd>
<dt>Variable 1 has 2 link(s):</dt><dd><p>(1 -1): pval = 0.00000 | val =  0.606
(0 -1): pval = 0.00000 | val =  0.447</p>
</dd>
<dt>Variable 2 has 2 link(s):</dt><dd><p>(2 -1): pval = 0.00000 | val =  0.618
(1 -2): pval = 0.00000 | val = -0.499</p>
</dd>
</dl>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>selected_links</strong> (<em>dict</em><em> or </em><em>None</em>) – Dictionary of form {0: [(3, -2), …], 1:[], …}
specifying whether only selected links should be tested. If None is
passed, all links are tested.</p></li>
<li><p><strong>tau_min</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Minimum time lag to test. Note that zero-lags are undirected.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 1</em><em>)</em>) – Maximum time lag. Must be larger or equal to tau_min.</p></li>
<li><p><strong>save_iterations</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: False</em><em>)</em>) – Whether to save iteration step results such as conditions used.</p></li>
<li><p><strong>pc_alpha</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 0.05</em><em>)</em>) – Significance level in algorithm.</p></li>
<li><p><strong>max_conds_dim</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Maximum number of conditions to test. If None is passed, this number
is unrestricted.</p></li>
<li><p><strong>max_combinations</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 1</em><em>)</em>) – Maximum number of combinations of conditions of current cardinality
to test. Defaults to 1 for PC_1 algorithm. For original PC algorithm
a larger number, such as 10, can be used.</p></li>
<li><p><strong>max_conds_py</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Maximum number of conditions of Y to use. If None is passed, this
number is unrestricted.</p></li>
<li><p><strong>max_conds_px</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Maximum number of conditions of Z to use. If None is passed, this
number is unrestricted.</p></li>
<li><p><strong>alpha_level</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 0.05</em><em>)</em>) – Significance level at which the p_matrix is thresholded to
get graph.</p></li>
<li><p><strong>fdr_method</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'fdr_bh'</em><em>)</em>) – Correction method, currently implemented is Benjamini-Hochberg
False Discovery Rate method.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>graph</strong> (<em>array of shape [N, N, tau_max+1]</em>) – Causal graph, see description above for interpretation.</p></li>
<li><p><strong>val_matrix</strong> (<em>array of shape [N, N, tau_max+1]</em>) – Estimated matrix of test statistic values.</p></li>
<li><p><strong>p_matrix</strong> (<em>array of shape [N, N, tau_max+1]</em>) – Estimated matrix of p-values, optionally adjusted if fdr_method is
not ‘none’.</p></li>
<li><p><strong>conf_matrix</strong> (<em>array of shape [N, N, tau_max+1,2]</em>) – Estimated matrix of confidence intervals of test statistic values.
Only computed if set in cond_ind_test, where also the percentiles
are set.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI.run_pcmciplus">
<span class="sig-name descname"><span class="pre">run_pcmciplus</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">selected_links</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_min</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pc_alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.01</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contemp_collider_rule</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'majority'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conflict_resolution</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reset_lagged_links</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_conds_dim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_conds_py</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_conds_px</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_conds_px_lagged</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fdr_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'none'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/pcmci.html#PCMCI.run_pcmciplus"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.pcmci.PCMCI.run_pcmciplus" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs PCMCIplus time-lagged and contemporaneous causal discovery for
time series.</p>
<p>Method described in <a class="footnote-reference brackets" href="#id4" id="id10" role="doc-noteref"><span class="fn-bracket">[</span>5<span class="fn-bracket">]</span></a>:
<a class="reference external" href="http://www.auai.org/~w-auai/uai2020/proceedings/579_main_paper.pdf">http://www.auai.org/~w-auai/uai2020/proceedings/579_main_paper.pdf</a></p>
<p class="rubric">Notes</p>
<p>The PCMCIplus causal discovery method is described in <a class="footnote-reference brackets" href="#id4" id="id11" role="doc-noteref"><span class="fn-bracket">[</span>5<span class="fn-bracket">]</span></a>, where
also analytical and numerical results are presented. In contrast to
PCMCI, PCMCIplus can identify the full, lagged and contemporaneous,
causal graph (up to the Markov equivalence class for contemporaneous
links) under the standard assumptions of Causal Sufficiency,
Faithfulness and the Markov condition.</p>
<p>PCMCIplus estimates time-lagged and contemporaneous causal links by a
four-step procedure:</p>
<p>1.  Condition-selection (same as for PCMCI): For each variable
<img class="math" src="_images/math/e3fc28292267f066fee7718c64f4bbfece521f24.png" alt="j"/>, estimate a <em>superset</em> of lagged parents <img class="math" src="_images/math/15af846e07c93e1c0d4cdbba8f4cefca67a811cb.png" alt="\widehat{
\mathcal{B}}_t^-( X^j_t)"/> with the iterative PC1 algorithm,
implemented as <code class="docutils literal notranslate"><span class="pre">run_pc_stable</span></code>. The condition-selection step
reduces the dimensionality and avoids conditioning on irrelevant
variables.</p>
<p>2.   PC skeleton phase with contemporaneous conditions and <em>Momentary
conditional independence</em> (MCI) tests: Iterate through subsets
<img class="math" src="_images/math/f852990c4a225a95d2c694a9f1351f54bac3ba86.png" alt="\mathcal{S}"/> of contemporaneous adjacencies and conduct MCI
conditional independence tests:</p>
<div class="math">
<p><img src="_images/math/d4ac9b60213cb2ffe02c6d5dacac8796221c73ac.png" alt="X^i_{t-\tau} ~\perp~ X^j_{t} ~|~ \mathcal{S},
\widehat{\mathcal{B}}_t^-(X^j_t),
\widehat{\mathcal{B}}_{t-\tau}^-(X^i_{t-{\tau}})"/></p>
</div><p>here implemented as <code class="docutils literal notranslate"><span class="pre">run_pcalg</span></code>. This step estimates the p-values and
test statistic values for all lagged and contemporaneous adjacencies
accounting for common drivers, indirect links, and autocorrelation.</p>
<p>3.   PC collider orientation phase: Orient contemporaneous collider
motifs based on unshielded triples. Optionally apply conservative or
majority rule (also based on MCI tests).</p>
<p>4.   PC rule orientation phase: Orient remaining contemporaneous
links based on PC rules.</p>
<p>In contrast to PCMCI, the relevant output of PCMCIplus is the
array <code class="docutils literal notranslate"><span class="pre">graph</span></code>. Its string entries are interpreted as follows:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">graph[i,j,tau]=--&gt;</span></code> for <img class="math" src="_images/math/2882aa5d9b56c4f42679de5f902a74d8ae88a624.png" alt="\tau&gt;0"/> denotes a directed, lagged
causal link from <img class="math" src="_images/math/5aa339d4daf45a810dda332e3c80a0698e526e04.png" alt="i"/> to <img class="math" src="_images/math/e3fc28292267f066fee7718c64f4bbfece521f24.png" alt="j"/> at lag <img class="math" src="_images/math/914b2d4b6659b86d3153d5510839dfb254dfc8a3.png" alt="\tau"/></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">graph[i,j,0]=--&gt;</span></code> (and <code class="docutils literal notranslate"><span class="pre">graph[j,i,0]=&lt;--</span></code>) denotes a directed,
contemporaneous causal link from <img class="math" src="_images/math/5aa339d4daf45a810dda332e3c80a0698e526e04.png" alt="i"/> to <img class="math" src="_images/math/e3fc28292267f066fee7718c64f4bbfece521f24.png" alt="j"/></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">graph[i,j,0]=o-o</span></code> (and <code class="docutils literal notranslate"><span class="pre">graph[j,i,0]=o-o</span></code>) denotes an unoriented,
contemporaneous adjacency between <img class="math" src="_images/math/5aa339d4daf45a810dda332e3c80a0698e526e04.png" alt="i"/> and <img class="math" src="_images/math/e3fc28292267f066fee7718c64f4bbfece521f24.png" alt="j"/> indicating
that the collider and orientation rules could not be applied (Markov
equivalence)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">graph[i,j,0]=x-x</span></code> and (<code class="docutils literal notranslate"><span class="pre">graph[j,i,0]=x-x</span></code>) denotes a conflicting,
contemporaneous adjacency between <img class="math" src="_images/math/5aa339d4daf45a810dda332e3c80a0698e526e04.png" alt="i"/> and <img class="math" src="_images/math/e3fc28292267f066fee7718c64f4bbfece521f24.png" alt="j"/> indicating
that the directionality is undecided due to conflicting orientation
rules</p></li>
</ul>
<p>Importantly, <code class="docutils literal notranslate"><span class="pre">p_matrix</span></code> and <code class="docutils literal notranslate"><span class="pre">val_matrix</span></code> for PCMCIplus quantify
the uncertainty and strength, respectively, only for the
adjacencies, but not for the directionality of contemporaneous links.
Note that lagged links are always oriented due to time order.</p>
<p>PCMCIplus can be flexibly combined with any kind of conditional
independence test statistic adapted to the kind of data (continuous
or discrete) and its assumed dependency types. These are available in
<code class="docutils literal notranslate"><span class="pre">tigramite.independence_tests</span></code>.</p>
<p>The main free parameters of PCMCIplus (in addition to free parameters of
the conditional independence tests) are the maximum time delay
<img class="math" src="_images/math/2bf86ca220f43e569c6c7aefaf32742919222e6e.png" alt="\tau_{\max}"/> (<code class="docutils literal notranslate"><span class="pre">tau_max</span></code>) and the significance threshold
<img class="math" src="_images/math/2f5aa019312e1bbc969deab8dca8b00f76025404.png" alt="\alpha"/> ( <code class="docutils literal notranslate"><span class="pre">pc_alpha</span></code>).</p>
<p>If a list or None is passed for <code class="docutils literal notranslate"><span class="pre">pc_alpha</span></code>, the significance level is
optimized for every graph across the given <code class="docutils literal notranslate"><span class="pre">pc_alpha</span></code> values using the
score computed in <code class="docutils literal notranslate"><span class="pre">cond_ind_test.get_model_selection_criterion()</span></code>.
Since PCMCIplus outputs not a DAG, but an equivalence class of DAGs,
first one member of this class is computed and then the score is
computed as the average over all models fits for each variable in <code class="docutils literal notranslate"><span class="pre">[0,</span>
<span class="pre">...,</span> <span class="pre">N]</span></code> for that member. The score is the same for all members of the
class.</p>
<p>The maximum time delay depends on the application and should be chosen
according to the maximum causal time lag expected in the complex system.
We recommend a rather large choice that includes peaks in the
<code class="docutils literal notranslate"><span class="pre">get_lagged_dependencies</span></code> function. Another important parameter is
<code class="docutils literal notranslate"><span class="pre">contemp_collider_rule</span></code>. Only if set to <code class="docutils literal notranslate"><span class="pre">majority</span></code> or
<code class="docutils literal notranslate"><span class="pre">conservative''</span> <span class="pre">and</span> <span class="pre">together</span> <span class="pre">with</span> <span class="pre">``conflict_resolution=True</span></code>,
PCMCIplus is fully <em>order independent</em> meaning that the order of the N
variables in the dataframe does not matter. Last, the default option
<code class="docutils literal notranslate"><span class="pre">reset_lagged_links=False</span></code> restricts the detection of lagged causal
links in Step 2 to the significant adjacencies found in Step 1, given by
<img class="math" src="_images/math/6ecffdf6c6b31ddba452a834568233441411465b.png" alt="\widehat{ \mathcal{B}}_t^-( X^j_t)"/>. For
<code class="docutils literal notranslate"><span class="pre">reset_lagged_links=True</span></code>, <em>all</em> lagged links are considered again,
which improves detection power for lagged links, but also leads to
larger runtimes.</p>
<p>Further optional parameters are discussed in <a class="footnote-reference brackets" href="#id4" id="id12" role="doc-noteref"><span class="fn-bracket">[</span>5<span class="fn-bracket">]</span></a>.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">tigramite.pcmci</span> <span class="kn">import</span> <span class="n">PCMCI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">tigramite.independence_tests</span> <span class="kn">import</span> <span class="n">ParCorr</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">tigramite.data_processing</span> <span class="k">as</span> <span class="nn">pp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">tigramite.toymodels</span> <span class="kn">import</span> <span class="n">structural_causal_processes</span> <span class="k">as</span> <span class="n">toys</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Example process to play around with</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Each key refers to a variable and the incoming links are supplied</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># as a list of format [((var, -lag), coeff, function), ...]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">def</span> <span class="nf">lin_f</span><span class="p">(</span><span class="n">x</span><span class="p">):</span> <span class="k">return</span> <span class="n">x</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">links</span> <span class="o">=</span> <span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="p">[((</span><span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">),</span> <span class="mf">0.9</span><span class="p">,</span> <span class="n">lin_f</span><span class="p">)],</span>
<span class="go">             1: [((1, -1), 0.8, lin_f), ((0, -1), 0.8, lin_f)],</span>
<span class="go">             2: [((2, -1), 0.7, lin_f), ((1, 0), 0.6, lin_f)],</span>
<span class="go">             3: [((3, -1), 0.7, lin_f), ((2, 0), -0.5, lin_f)],</span>
<span class="go">             }</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="p">,</span> <span class="n">nonstat</span> <span class="o">=</span> <span class="n">toys</span><span class="o">.</span><span class="n">structural_causal_process</span><span class="p">(</span><span class="n">links</span><span class="p">,</span>
<span class="go">                    T=1000, seed=7)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Data must be array of shape (time, variables)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="go">(1000, 4)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dataframe</span> <span class="o">=</span> <span class="n">pp</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cond_ind_test</span> <span class="o">=</span> <span class="n">ParCorr</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcmci</span> <span class="o">=</span> <span class="n">PCMCI</span><span class="p">(</span><span class="n">dataframe</span><span class="o">=</span><span class="n">dataframe</span><span class="p">,</span> <span class="n">cond_ind_test</span><span class="o">=</span><span class="n">cond_ind_test</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">results</span> <span class="o">=</span> <span class="n">pcmci</span><span class="o">.</span><span class="n">run_pcmciplus</span><span class="p">(</span><span class="n">tau_min</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">tau_max</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">pc_alpha</span><span class="o">=</span><span class="mf">0.01</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcmci</span><span class="o">.</span><span class="n">print_results</span><span class="p">(</span><span class="n">results</span><span class="p">,</span> <span class="n">alpha_level</span><span class="o">=</span><span class="mf">0.01</span><span class="p">)</span>
<span class="go">    ## Significant links at alpha = 0.01:</span>
</pre></div>
</div>
<blockquote>
<div><dl class="simple">
<dt>Variable 0 has 1 link(s):</dt><dd><p>(0 -1): pval = 0.00000 | val =  0.676</p>
</dd>
<dt>Variable 1 has 2 link(s):</dt><dd><p>(1 -1): pval = 0.00000 | val =  0.602
(0 -1): pval = 0.00000 | val =  0.599</p>
</dd>
<dt>Variable 2 has 2 link(s):</dt><dd><p>(1  0): pval = 0.00000 | val =  0.486
(2 -1): pval = 0.00000 | val =  0.466</p>
</dd>
<dt>Variable 3 has 2 link(s):</dt><dd><p>(3 -1): pval = 0.00000 | val =  0.524
(2  0): pval = 0.00000 | val = -0.449</p>
</dd>
</dl>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>selected_links</strong> (<em>dict</em><em> or </em><em>None</em>) – Dictionary of form {0: [(3, 0), (0, -1), …], 1:[], …}
specifying whether only selected links should be tested. If None is
passed, all links are tested.</p></li>
<li><p><strong>tau_min</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Minimum time lag to test.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 1</em><em>)</em>) – Maximum time lag. Must be larger or equal to tau_min.</p></li>
<li><p><strong>pc_alpha</strong> (<em>float</em><em> or </em><em>list of floats</em><em>, </em><em>default: 0.01</em>) – Significance level in algorithm. If a list or None is passed, the
pc_alpha level is optimized for every graph across the given
pc_alpha values ([0.001, 0.005, 0.01, 0.025, 0.05] for None) using
the score computed in cond_ind_test.get_model_selection_criterion().</p></li>
<li><p><strong>contemp_collider_rule</strong> (<em>{'majority'</em><em>, </em><em>'conservative'</em><em>, </em><em>'none'}</em>) – Rule for collider phase to use. See the paper for details. Only
‘majority’ and ‘conservative’ lead to an order-independent
algorithm.</p></li>
<li><p><strong>conflict_resolution</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: True</em><em>)</em>) – Whether to mark conflicts in orientation rules. Only for True
this leads to an order-independent algorithm.</p></li>
<li><p><strong>reset_lagged_links</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: False</em><em>)</em>) – Restricts the detection of lagged causal links in Step 2 to the
significant adjacencies found in the PC1 algorithm in Step 1. For
True, <em>all</em> lagged links are considered again, which improves
detection power for lagged links, but also leads to larger
runtimes.</p></li>
<li><p><strong>max_conds_dim</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Maximum number of conditions to test. If None is passed, this number
is unrestricted.</p></li>
<li><p><strong>max_conds_py</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Maximum number of lagged conditions of Y to use in MCI tests. If
None is passed, this number is unrestricted.</p></li>
<li><p><strong>max_conds_px</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Maximum number of lagged conditions of X to use in MCI tests. If
None is passed, this number is unrestricted.</p></li>
<li><p><strong>max_conds_px_lagged</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Maximum number of lagged conditions of X when X is lagged in MCI
tests. If None is passed, this number is equal to max_conds_px.</p></li>
<li><p><strong>fdr_method</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'none'</em><em>)</em>) – Correction method, default is Benjamini-Hochberg False Discovery
Rate method.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>graph</strong> (<em>array of shape [N, N, tau_max+1]</em>) – Resulting causal graph, see description above for interpretation.</p></li>
<li><p><strong>val_matrix</strong> (<em>array of shape [N, N, tau_max+1]</em>) – Estimated matrix of test statistic values regarding adjacencies.</p></li>
<li><p><strong>p_matrix</strong> (<em>array of shape [N, N, tau_max+1]</em>) – Estimated matrix of p-values regarding adjacencies.</p></li>
<li><p><strong>sepset</strong> (<em>dictionary</em>) – Separating sets. See paper for details.</p></li>
<li><p><strong>ambiguous_triples</strong> (<em>list</em>) – List of ambiguous triples, only relevant for ‘majority’ and
‘conservative’ rules, see paper for details.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI.run_sliding_window_of">
<span class="sig-name descname"><span class="pre">run_sliding_window_of</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method_args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window_step</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window_length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conf_lev</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.95</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/pcmci.html#PCMCI.run_sliding_window_of"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.pcmci.PCMCI.run_sliding_window_of" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs chosen method on sliding windows taken from DataFrame.</p>
<p>The function returns summary_results and all_results (containing the
individual window results). summary_results contains val_matrix_mean
and val_matrix_interval, the latter containing the confidence bounds for
conf_lev. If the method also returns a graph, then ‘most_frequent_links’
containing the most frequent link outcome (either 0 or 1 or a specific
link type) in each entry of graph, as well as ‘link_frequency’,
containing the occurence frequency of the most frequent link outcome,
are returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>method</strong> (<em>str</em>) – Chosen method among valid functions in PCMCI.</p></li>
<li><p><strong>method_args</strong> (<em>dict</em>) – Arguments passed to method.</p></li>
<li><p><strong>window_step</strong> (<em>int</em>) – Time step of windows.</p></li>
<li><p><strong>window_length</strong> (<em>int</em>) – Length of sliding window.</p></li>
<li><p><strong>conf_lev</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 0.9</em><em>)</em>) – Two-sided confidence interval for summary results.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary of results for every sliding window.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.pcmci.PCMCI.symmetrize_p_and_val_matrix">
<span class="sig-name descname"><span class="pre">symmetrize_p_and_val_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">p_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">selected_links</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conf_matrix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/pcmci.html#PCMCI.symmetrize_p_and_val_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.pcmci.PCMCI.symmetrize_p_and_val_matrix" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Symmetrizes the p_matrix, val_matrix, and conf_matrix based on selected_links</dt><dd><p>and the larger p-value.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>val_matrix</strong> (<em>array of shape</em><em> [</em><em>N</em><em>, </em><em>N</em><em>, </em><em>tau_max+1</em><em>]</em>) – Estimated matrix of test statistic values.</p></li>
<li><p><strong>p_matrix</strong> (<em>array of shape</em><em> [</em><em>N</em><em>, </em><em>N</em><em>, </em><em>tau_max+1</em><em>]</em>) – Estimated matrix of p-values. Set to 1 if val_only=True.</p></li>
<li><p><strong>conf_matrix</strong> (<em>array of shape</em><em> [</em><em>N</em><em>, </em><em>N</em><em>, </em><em>tau_max+1</em><em>,</em><em>2</em><em>]</em>) – Estimated matrix of confidence intervals of test statistic values.
Only computed if set in cond_ind_test, where also the percentiles
are set.</p></li>
<li><p><strong>selected_links</strong> (<em>dict</em><em> or </em><em>None</em>) – Dictionary of form {0: [(3, -2), …], 1:[], …}
specifying whether only selected links should be tested. If None is
passed, all links are tested.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>val_matrix</strong> (<em>array of shape [N, N, tau_max+1]</em>) – Estimated matrix of test statistic values.</p></li>
<li><p><strong>p_matrix</strong> (<em>array of shape [N, N, tau_max+1]</em>) – Estimated matrix of p-values. Set to 1 if val_only=True.</p></li>
<li><p><strong>conf_matrix</strong> (<em>array of shape [N, N, tau_max+1,2]</em>) – Estimated matrix of confidence intervals of test statistic values.
Only computed if set in cond_ind_test, where also the percentiles
are set.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="tigramite-lpcmci-lpcmci">
<h1><code class="xref py py-mod docutils literal notranslate"><span class="pre">tigramite.lpcmci</span></code>: LPCMCI<a class="headerlink" href="#tigramite-lpcmci-lpcmci" title="Permalink to this heading">¶</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="tigramite.lpcmci.LPCMCI">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tigramite.lpcmci.</span></span><span class="sig-name descname"><span class="pre">LPCMCI</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataframe</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cond_ind_test</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/lpcmci.html#LPCMCI"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.lpcmci.LPCMCI" title="Permalink to this definition">¶</a></dt>
<dd><p>LPCMCI is an algorithm for causal discovery in large-scale times series that allows for latent confounders and learns lag-specific
causal relationships.</p>
<p>The algorithm is introduced and explained in:
[1] Gerhardus, A. &amp; Runge, J. High-recall causal discovery for autocorrelated time series with latent confounders Advances in Neural
Information Processing Systems, 2020, 33. <a class="reference external" href="https://proceedings.neurips.cc/paper/2020/hash/94e70705efae423efda1088614128d0b-Abstract.html">https://proceedings.neurips.cc/paper/2020/hash/94e70705efae423efda1088614128d0b-Abstract.html</a></p>
<p>NOTE:
This method is still EXPERIMENTAL since the default settings of hyperparameters are still being fine-tuned. We actually invite
feedback on which work best in applications and numerical experiments.</p>
<p>The main function, which applies the algorithm, is ‘run_lpcmci’.</p>
<p>Parameters passed to the constructor:
- dataframe:</p>
<blockquote>
<div><p>Tigramite dataframe object that contains the the time series dataset bold{X}</p>
</div></blockquote>
<ul class="simple">
<li><dl class="simple">
<dt>cond_ind_test:</dt><dd><p>A conditional independence test object that specifies which conditional independence test CI is to be used</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>verbosity:</dt><dd><p>Controls the verbose output self.run_lpcmci() and the function it calls.</p>
</dd>
</dl>
</li>
</ul>
<p>Parameters passed to self.run_lpcmci():
Note: The default values are still being tuned and some parameters might be removed in the future.
- selected_links: dict or None</p>
<blockquote>
<div><p>Dictionary of the form {0: [(3, 0), (0, -1), …], 1:[], …} that specifys which links are potentially present. All other links
are assumed to be absent. If None is passed all links are potentially present.</p>
</div></blockquote>
<ul class="simple">
<li><dl class="simple">
<dt>tau_min:</dt><dd><p>The assumed minimum time lag, i.e., links with a lag smaller than tau_min are assumed to be absent.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>tau_max:</dt><dd><p>The maximum considered time lag, i.e., the algorithm learns a DPAG on a time window [t-taumax, t] with tau_max + 1 time steps.
It is <em>not</em> assumed that in the underlying time series DAG there are no links with a lag larger than tau_max.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>pc_alpha:</dt><dd><p>The significance level of conditional independence tests</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>n_preliminary_iterations:</dt><dd><p>Determines the number of iterations in the preliminary phase of LPCMCI, corresponding to the ‘k’ in LPCMCI(k) in [1].</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>max_cond_px:</dt><dd><p>Consider a pair of variables (X^i_{t-tau}, X^j_t) with tau &gt; 0. In Algorithm S2 in [1] (here this is
self._run_ancestral_removal_phase()), the algorithm does not test for conditional independence given subsets of
apds_t(X^i_{t-tau}, X^j_t, C(G)) of cardinality higher than max_cond_px. In Algorithm S3 in [1] (here this is
self._run_non_ancestral_removal_phase()), the algorithm does not test for conditional independence given subsets of
napds_t(X^i_{t-tau}, X^j_t, C(G)) of cardinality higher than max_cond_px.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>max_p_global:</dt><dd><p>Restricts all conditional independence tests to conditioning sets with cardinality smaller or equal to max_p_global</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>max_p_non_ancestral:</dt><dd><p>Restricts all conditional independence tests in the second removal phase (here this is self._run_dsep_removal_phase()) to
conditioning sets with cardinality smaller or equal to max_p_global</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>max_q_global:</dt><dd><p>For each ordered pair (X^i_{t-tau}, X^j_t) of adjacent variables and for each cardinality of the conditioning sets test at most
max_q_global many conditioning sets (when summing over all tested cardinalities more than max_q_global tests may be made)</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>max_pds_set:</dt><dd><p>In Algorithm S3 (here this is self._run_non_ancestral_removal_phase()), the algorithm tests for conditional independence given
subsets of the relevant napds_t sets. If for a given link the set napds_t(X^j_t, X^i_{t-tau}, C(G)) has more than max_pds_set many
elements (or, if the link is also tested in the opposite directed, if napds_t(X^i_{t-tau}, X^j_t, C(G)) has more than max_pds_set
elements), this link is not tested.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>prelim_with_collider_rules:</dt><dd><p>If True: As in pseudocode
If False: Line 22 of Algorithm S2 in [1] is replaced by line 18 of Algorithm S2 when Algorithm S2 is called from the preliminary
phase (not in the last application of Algorithm S2 directly before Algorithm S3 is applied)</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>parents_of_lagged:</dt><dd><p>If True: As in pseudocode
If False: The default conditioning set is pa(X^j_t, C(G)) rather than pa({X^j_t, X^i_{t-tau}, C(G)) for tau &gt; 0</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>prelim_only:</dt><dd><p>If True, stop after the preliminary phase. Can be used for detailed performance analysis</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>break_once_separated:</dt><dd><p>If True: As in pseudocode
If False: The break commands are removed from Algorithms S2 and S3 in in [1]</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>no_non_ancestral_phase:</dt><dd><p>If True, do not execute Algorithm S3. Can be used for detailed performance analysis</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>use_a_pds_t_for_majority:</dt><dd><p>If True: As in pseudocode
If False: The search for separating sets instructed by the majority rule is made given subsets adj(X^j_t, C(G)) rather than
subsets of apds_t(X^j_t, X^i_{t-tau}, C(G))</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>orient_contemp:</dt><dd><p>If orient_contemp == 1: As in pseudocode of Algorithm S2 in [1]
If orient_contemp == 2: Also orient contemporaneous links in line 18 of Algorithm S2
If orient_comtemp == 0: Also not orient contemporaneous links in line 22 of Algorithm S2</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>update_middle_marks:</dt><dd><p>If True: As in pseudoce of Algorithms S2 and S3 in [1]
If False: The MMR rule is not applied</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>prelim_rules:</dt><dd><p>If prelim_rules == 1: As in pseudocode of Algorithm S2 in [1]
If prelim_rules == 0: Exclude rules R9^prime and R10^prime from line 18 in Algorithm S2</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>fix_all_edges_before_final_orientation:</dt><dd><p>When one of max_p_global, max_p_non_ancestral, max_q_global or max_pds_set is not np.inf, the algorithm may terminate although not
all middle marks are empty. All orientation rules are nevertheless sound, since the rules always check for the appropriate middle
marks. If fix_all_edges_before_final_orientation is True, all middle marks are set to the empty middle mark by force, followed by
another application of the rules.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>auto_first:</dt><dd><p>If True: As in pseudcode of Algorithms S2 and S3 in [1]
If False: Autodependency links are not prioritized even before contemporaneous links</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>remember_only_parents:</dt><dd><p>If True: As in pseudocode of Algorithm 1
If False: If X^i_{t-tau} has been marked as ancestor of X^j_t at any point of a preliminary iteration but the link between
X^i_{t-tau} and X^j_t was removed later, the link is nevertheless initialized with a tail at X^i_{t-tau} in the re-initialization</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>no_apr:</dt><dd><p>If no_apr == 0: As in pseudcode of Algorithms S2 and S3 in [1]
If no_apr == 1: The APR is not applied by Algorithm S2, except in line 22 of its last call directly before the call of Algorithm S3
If no_apr == 2: The APR is never applied</p>
</dd>
</dl>
</li>
</ul>
<dl class="simple">
<dt>Return value of self.run_lpcmci():</dt><dd><dl class="simple">
<dt>graph<span class="classifier">array of shape (N, N, tau_max+1)</span></dt><dd><p>Resulting DPAG, representing the learned causal relationships.</p>
</dd>
<dt>val_matrix<span class="classifier">array of shape (N, N, tau_max+1)</span></dt><dd><p>Estimated matrix of test statistic values regarding adjacencies.</p>
</dd>
<dt>p_matrix<span class="classifier">array of shape [N, N, tau_max+1]</span></dt><dd><p>Estimated matrix of p-values regarding adjacencies.</p>
</dd>
</dl>
</dd>
<dt>A note on middle marks:</dt><dd><p>For convenience (to have strings of the same lengths) we here internally denote the empty middle mark by ‘-’.  For post-processing
purposes all middle marks are set to the empty middle mark (here ‘-‘).</p>
</dd>
<dt>A note on wildcards:</dt><dd><p>The middle mark wildcard ast and the edge mark wildcard are here represented as <a href="#id13"><span class="problematic" id="id14">*</span></a>, the edge mark wildcard star as +</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tigramite.lpcmci.LPCMCI.run_lpcmci">
<span class="sig-name descname"><span class="pre">run_lpcmci</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">selected_links</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_min</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pc_alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_preliminary_iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_cond_px</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_p_global</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">numpy.inf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_p_non_ancestral</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">numpy.inf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_q_global</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">numpy.inf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_pds_set</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">numpy.inf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prelim_with_collider_rules</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parents_of_lagged</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prelim_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">break_once_separated</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">no_non_ancestral_phase</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_a_pds_t_for_majority</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orient_contemp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">update_middle_marks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prelim_rules</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fix_all_edges_before_final_orientation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">auto_first</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remember_only_parents</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">no_apr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/lpcmci.html#LPCMCI.run_lpcmci"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.lpcmci.LPCMCI.run_lpcmci" title="Permalink to this definition">¶</a></dt>
<dd><p>Run LPCMCI on the dataset and with the conditional independence test passed to the class constructor and with the
options passed to this function.</p>
</dd></dl>

</dd></dl>

</section>
<section id="tigramite-independence-tests-conditional-independence-tests">
<h1><code class="xref py py-mod docutils literal notranslate"><span class="pre">tigramite.independence_tests</span></code>: Conditional independence tests<a class="headerlink" href="#tigramite-independence-tests-conditional-independence-tests" title="Permalink to this heading">¶</a></h1>
<p>Base class:</p>
<dl class="py class">
<dt class="sig sig-object py" id="tigramite.independence_tests.CondIndTest">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tigramite.independence_tests.</span></span><span class="sig-name descname"><span class="pre">CondIndTest</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">42</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">significance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'analytic'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fixed_thres</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_blocklength</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">confidence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conf_lev</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.9</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conf_samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conf_blocklength</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recycle_residuals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/independence_tests_base.html#CondIndTest"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.CondIndTest" title="Permalink to this definition">¶</a></dt>
<dd><p>Base class of conditional independence tests.</p>
<p>Provides useful general functions for different independence tests such as
shuffle significance testing and bootstrap confidence estimation. Also
handles masked samples. Other test classes can inherit from this class.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seed</strong> (<em>int</em><em>, </em><em>optional</em><em>(</em><em>default = 42</em><em>)</em>) – Seed for RandomState (default_rng)</p></li>
<li><p><strong>mask_type</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default = None</em><em>)</em>) – Must be in {None, ‘y’,’x’,’z’,’xy’,’xz’,’yz’,’xyz’}
Masking mode: Indicators for which variables in the dependence measure
I(X; Y | Z) the samples should be masked. If None, the mask is not used.
Explained in tutorial on masking and missing values.</p></li>
<li><p><strong>significance</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'analytic'</em><em>)</em>) – Type of significance test to use. In this package ‘analytic’,
‘fixed_thres’ and ‘shuffle_test’ are available.</p></li>
<li><p><strong>fixed_thres</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 0.1</em><em>)</em>) – If significance is ‘fixed_thres’, this specifies the threshold for the
absolute value of the dependence measure.</p></li>
<li><p><strong>sig_samples</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 1000</em><em>)</em>) – Number of samples for shuffle significance test.</p></li>
<li><p><strong>sig_blocklength</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Block length for block-shuffle significance test. If None, the
block length is determined from the decay of the autocovariance as
explained in <a class="footnote-reference brackets" href="#id3" id="id15" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a>.</p></li>
<li><p><strong>confidence</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Specify type of confidence estimation. If False, numpy.nan is returned.
‘bootstrap’ can be used with any test, for ParCorr also ‘analytic’ is
implemented.</p></li>
<li><p><strong>conf_lev</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 0.9</em><em>)</em>) – Two-sided confidence interval.</p></li>
<li><p><strong>conf_samples</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 100</em><em>)</em>) – Number of samples for bootstrap.</p></li>
<li><p><strong>conf_blocklength</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Block length for block-bootstrap. If None, the block length is
determined from the decay of the autocovariance as explained in <a class="footnote-reference brackets" href="#id3" id="id16" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a>.</p></li>
<li><p><strong>recycle_residuals</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: False</em><em>)</em>) – Specifies whether residuals should be stored. This may be faster, but
can cost considerable memory.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Level of verbosity.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.CondIndTest.get_analytic_confidence">
<span class="sig-name descname"><span class="pre">get_analytic_confidence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conf_lev</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/independence_tests_base.html#CondIndTest.get_analytic_confidence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.CondIndTest.get_analytic_confidence" title="Permalink to this definition">¶</a></dt>
<dd><p>Base class assumption that this is not implemented.  Concrete classes
should override when possible.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.CondIndTest.get_analytic_significance">
<span class="sig-name descname"><span class="pre">get_analytic_significance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">T</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dim</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/independence_tests_base.html#CondIndTest.get_analytic_significance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.CondIndTest.get_analytic_significance" title="Permalink to this definition">¶</a></dt>
<dd><p>Base class assumption that this is not implemented.  Concrete classes
should override when possible.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.CondIndTest.get_bootstrap_confidence">
<span class="sig-name descname"><span class="pre">get_bootstrap_confidence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xyz</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependence_measure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conf_samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conf_blocklength</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conf_lev</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.95</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/independence_tests_base.html#CondIndTest.get_bootstrap_confidence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.CondIndTest.get_bootstrap_confidence" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform bootstrap confidence interval estimation.</p>
<p>With conf_blocklength &gt; 1 or None a block-bootstrap is performed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>array</strong> (<em>array-like</em>) – data array with X, Y, Z in rows and observations in columns</p></li>
<li><p><strong>xyz</strong> (<em>array of ints</em>) – XYZ identifier array of shape (dim,).</p></li>
<li><p><strong>dependence_measure</strong> (<em>function</em><em> (</em><em>default = self.get_dependence_measure</em><em>)</em>) – Dependence measure function must be of form
dependence_measure(array, xyz) and return a numeric value</p></li>
<li><p><strong>conf_lev</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 0.9</em><em>)</em>) – Two-sided confidence interval.</p></li>
<li><p><strong>conf_samples</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 100</em><em>)</em>) – Number of samples for bootstrap.</p></li>
<li><p><strong>conf_blocklength</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Block length for block-bootstrap. If None, the block length is
determined from the decay of the autocovariance as explained in
<a class="footnote-reference brackets" href="#id3" id="id17" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a>.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Level of verbosity.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>(conf_lower, conf_upper)</strong> – Upper and lower confidence bound of confidence interval.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple of floats</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.CondIndTest.get_confidence">
<span class="sig-name descname"><span class="pre">get_confidence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Z</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/independence_tests_base.html#CondIndTest.get_confidence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.CondIndTest.get_confidence" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform confidence interval estimation.</p>
<p>Calls the dependence measure and confidence test functions. The child
classes can specify a function get_dependence_measure and
get_analytic_confidence or get_bootstrap_confidence. If confidence is
False, (numpy.nan, numpy.nan) is returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>list of tuples</em>) – X,Y,Z are of the form [(var, -tau)], where var specifies the
variable index and tau the time lag.</p></li>
<li><p><strong>Y</strong> (<em>list of tuples</em>) – X,Y,Z are of the form [(var, -tau)], where var specifies the
variable index and tau the time lag.</p></li>
<li><p><strong>Z</strong> (<em>list of tuples</em>) – X,Y,Z are of the form [(var, -tau)], where var specifies the
variable index and tau the time lag.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Maximum time lag. This may be used to make sure that estimates for
different lags in X, Z, all have the same sample size.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>(conf_lower, conf_upper)</strong> – Upper and lower confidence bound of confidence interval.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple of floats</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.CondIndTest.get_dependence_measure">
<em class="property"><span class="pre">abstract</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_dependence_measure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xyz</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/independence_tests_base.html#CondIndTest.get_dependence_measure"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.CondIndTest.get_dependence_measure" title="Permalink to this definition">¶</a></dt>
<dd><p>Abstract function that all concrete classes must instantiate.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.CondIndTest.get_fixed_thres_significance">
<span class="sig-name descname"><span class="pre">get_fixed_thres_significance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fixed_thres</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/independence_tests_base.html#CondIndTest.get_fixed_thres_significance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.CondIndTest.get_fixed_thres_significance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns signficance for thresholding test.</p>
<p>Returns 0 if numpy.abs(value) is smaller than fixed_thres and 1 else.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>value</strong> (<em>number</em>) – Value of test statistic for unshuffled estimate.</p></li>
<li><p><strong>fixed_thres</strong> (<em>number</em>) – Fixed threshold, is made positive.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>pval</strong> – Returns 0 if numpy.abs(value) is smaller than fixed_thres and 1
else.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.CondIndTest.get_measure">
<span class="sig-name descname"><span class="pre">get_measure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Z</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/independence_tests_base.html#CondIndTest.get_measure"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.CondIndTest.get_measure" title="Permalink to this definition">¶</a></dt>
<dd><p>Estimate dependence measure.</p>
<p>Calls the dependence measure function. The child classes must specify
a function get_dependence_measure.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>list of tuples</em>) – X,Y,Z are of the form [(var, -tau)], where var specifies the
variable index and tau the time lag.</p></li>
<li><p><strong>[</strong> (<em>Y</em>) – X,Y,Z are of the form [(var, -tau)], where var specifies the
variable index and tau the time lag.</p></li>
<li><p><strong>Z</strong><strong>]</strong> (<em>list of tuples</em>) – X,Y,Z are of the form [(var, -tau)], where var specifies the
variable index and tau the time lag.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Maximum time lag. This may be used to make sure that estimates for
different lags in X, Z, all have the same sample size.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>val</strong> – The test statistic value.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.CondIndTest.get_model_selection_criterion">
<span class="sig-name descname"><span class="pre">get_model_selection_criterion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">j</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parents</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/independence_tests_base.html#CondIndTest.get_model_selection_criterion"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.CondIndTest.get_model_selection_criterion" title="Permalink to this definition">¶</a></dt>
<dd><p>Base class assumption that this is not implemented.  Concrete classes
should override when possible.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.CondIndTest.get_shuffle_significance">
<span class="sig-name descname"><span class="pre">get_shuffle_significance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xyz</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_null_dist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/independence_tests_base.html#CondIndTest.get_shuffle_significance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.CondIndTest.get_shuffle_significance" title="Permalink to this definition">¶</a></dt>
<dd><p>Base class assumption that this is not implemented.  Concrete classes
should override when possible.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.CondIndTest.get_significance">
<span class="sig-name descname"><span class="pre">get_significance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">val</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xyz</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">T</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dim</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_override</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/independence_tests_base.html#CondIndTest.get_significance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.CondIndTest.get_significance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the p-value from whichever significance function is specified
for this test.  If an override is used, then it will call a different
function then specified by self.significance</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>val</strong> (<em>float</em>) – Test statistic value.</p></li>
<li><p><strong>array</strong> (<em>array-like</em>) – data array with X, Y, Z in rows and observations in columns</p></li>
<li><p><strong>xyz</strong> (<em>array of ints</em>) – XYZ identifier array of shape (dim,).</p></li>
<li><p><strong>T</strong> (<em>int</em>) – Sample length</p></li>
<li><p><strong>dim</strong> (<em>int</em>) – Dimensionality, ie, number of features.</p></li>
<li><p><strong>sig_override</strong> (<em>string</em>) – Must be in ‘analytic’, ‘shuffle_test’, ‘fixed_thres’</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>pval</strong> – P-value.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float or numpy.nan</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tigramite.independence_tests.CondIndTest.measure">
<em class="property"><span class="pre">abstract</span><span class="w"> </span><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">measure</span></span><a class="headerlink" href="#tigramite.independence_tests.CondIndTest.measure" title="Permalink to this definition">¶</a></dt>
<dd><p>Abstract property to store the type of independence test.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.CondIndTest.print_info">
<span class="sig-name descname"><span class="pre">print_info</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/independence_tests_base.html#CondIndTest.print_info"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.CondIndTest.print_info" title="Permalink to this definition">¶</a></dt>
<dd><p>Print information about the conditional independence test parameters</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.CondIndTest.run_test">
<span class="sig-name descname"><span class="pre">run_test</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Z</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut_off</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'2xtau_max'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/independence_tests_base.html#CondIndTest.run_test"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.CondIndTest.run_test" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform conditional independence test.</p>
<p>Calls the dependence measure and signficicance test functions. The child
classes must specify a function get_dependence_measure and either or
both functions get_analytic_significance and  get_shuffle_significance.
If recycle_residuals is True, also _get_single_residuals must be
available.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>list of tuples</em>) – X,Y,Z are of the form [(var, -tau)], where var specifies the
variable index and tau the time lag.</p></li>
<li><p><strong>Y</strong> (<em>list of tuples</em>) – X,Y,Z are of the form [(var, -tau)], where var specifies the
variable index and tau the time lag.</p></li>
<li><p><strong>Z</strong> (<em>list of tuples</em>) – X,Y,Z are of the form [(var, -tau)], where var specifies the
variable index and tau the time lag.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Maximum time lag. This may be used to make sure that estimates for
different lags in X, Z, all have the same sample size.</p></li>
<li><p><strong>cut_off</strong> (<em>{'2xtau_max'</em><em>, </em><em>'max_lag'</em><em>, </em><em>'max_lag_or_tau_max'}</em>) – How many samples to cutoff at the beginning. The default is
‘2xtau_max’, which guarantees that MCI tests are all conducted on
the same samples. For modeling, ‘max_lag_or_tau_max’ can be used,
which uses the maximum of tau_max and the conditions, which is
useful to compare multiple models on the same sample.  Last,
‘max_lag’ uses as much samples as possible.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>val, pval</strong> – The test statistic value and the p-value.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple of floats</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.CondIndTest.run_test_raw">
<span class="sig-name descname"><span class="pre">run_test_raw</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">z</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/independence_tests_base.html#CondIndTest.run_test_raw"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.CondIndTest.run_test_raw" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform conditional independence test directly on input arrays x, y, z.</p>
<p>Calls the dependence measure and signficicance test functions. The child
classes must specify a function get_dependence_measure and either or
both functions get_analytic_significance and  get_shuffle_significance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<em>arrays</em>) – x,y,z are of the form (samples, dimension).</p></li>
<li><p><strong>y</strong> (<em>arrays</em>) – x,y,z are of the form (samples, dimension).</p></li>
<li><p><strong>z</strong> (<em>arrays</em>) – x,y,z are of the form (samples, dimension).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>val, pval</strong> – The test statistic value and the p-value.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple of floats</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.CondIndTest.set_dataframe">
<span class="sig-name descname"><span class="pre">set_dataframe</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataframe</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/independence_tests_base.html#CondIndTest.set_dataframe"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.CondIndTest.set_dataframe" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize and check the dataframe.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>dataframe</strong> (<em>data object</em>) – Set tigramite dataframe object. It must have the attributes
dataframe.values yielding a numpy array of shape (observations T,
variables N) and optionally a mask of the same shape and a missing
values flag.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.CondIndTest.set_mask_type">
<span class="sig-name descname"><span class="pre">set_mask_type</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mask_type</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/independence_tests_base.html#CondIndTest.set_mask_type"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.CondIndTest.set_mask_type" title="Permalink to this definition">¶</a></dt>
<dd><p>Setter for mask type to ensure that this option does not clash with
recycle_residuals.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>mask_type</strong> (<em>str</em>) – Must be in {None, ‘y’,’x’,’z’,’xy’,’xz’,’yz’,’xyz’}
Masking mode: Indicators for which variables in the dependence measure
I(X; Y | Z) the samples should be masked. If None, the mask is not used.
Explained in tutorial on masking and missing values.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<p>Test statistics:</p>
<dl class="py class">
<dt class="sig sig-object py" id="tigramite.independence_tests.ParCorr">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tigramite.independence_tests.</span></span><span class="sig-name descname"><span class="pre">ParCorr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/parcorr.html#ParCorr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.ParCorr" title="Permalink to this definition">¶</a></dt>
<dd><p>Partial correlation test.</p>
<p>Partial correlation is estimated through linear ordinary least squares (OLS)
regression and a test for non-zero linear Pearson correlation on the
residuals.</p>
<p class="rubric">Notes</p>
<p>To test <img class="math" src="_images/math/667eda4bf3d5ce33b6cc785cadfef79bb95741ca.png" alt="X \perp Y | Z"/>, first <img class="math" src="_images/math/95f028ab2b20b895fa12d986e0d9f40f7b6e52d3.png" alt="Z"/> is regressed out from
<img class="math" src="_images/math/ed38fa24f1c94891bd312012aab3f6673be3eb83.png" alt="X"/> and <img class="math" src="_images/math/7daf0d4815e763eb90f0d5f1dc406f668c1e21db.png" alt="Y"/> assuming the  model</p>
<div class="math">
<p><img src="_images/math/78b6e77cc610bf56f5c64cb6dcc6d6ee49f886f9.png" alt="X &amp; =  Z \beta_X + \epsilon_{X} \\
Y &amp; =  Z \beta_Y + \epsilon_{Y}"/></p>
</div><p>using OLS regression. Then the dependency of the residuals is tested with
the Pearson correlation test.</p>
<div class="math">
<p><img src="_images/math/789735c1db036ea36cd0aa25a3af4b2528ed3abe.png" alt="\rho\left(r_X, r_Y\right)"/></p>
</div><p>For the <code class="docutils literal notranslate"><span class="pre">significance='analytic'</span></code> Student’s-<em>t</em> distribution with
<img class="math" src="_images/math/9c353382eebb42a8a9dec3a426d346d4842bd39d.png" alt="T-D_Z-2"/> degrees of freedom is implemented.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>**kwargs</strong> – Arguments passed on to Parent class CondIndTest.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.ParCorr.get_analytic_confidence">
<span class="sig-name descname"><span class="pre">get_analytic_confidence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conf_lev</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/parcorr.html#ParCorr.get_analytic_confidence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.ParCorr.get_analytic_confidence" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns analytic confidence interval for correlation coefficient.</p>
<p>Based on Student’s t-distribution.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>value</strong> (<em>float</em>) – Test statistic value.</p></li>
<li><p><strong>df</strong> (<em>int</em>) – degrees of freedom of the test</p></li>
<li><p><strong>conf_lev</strong> (<em>float</em>) – Confidence interval, eg, 0.9</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>(conf_lower, conf_upper)</strong> – Upper and lower confidence bound of confidence interval.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple of floats</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.ParCorr.get_analytic_significance">
<span class="sig-name descname"><span class="pre">get_analytic_significance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">T</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dim</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/parcorr.html#ParCorr.get_analytic_significance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.ParCorr.get_analytic_significance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns analytic p-value from Student’s t-test for the Pearson
correlation coefficient.</p>
<p>Assumes two-sided correlation. If the degrees of freedom are less than
1, numpy.nan is returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>value</strong> (<em>float</em>) – Test statistic value.</p></li>
<li><p><strong>T</strong> (<em>int</em>) – Sample length</p></li>
<li><p><strong>dim</strong> (<em>int</em>) – Dimensionality, ie, number of features.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>pval</strong> – P-value.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float or numpy.nan</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.ParCorr.get_dependence_measure">
<span class="sig-name descname"><span class="pre">get_dependence_measure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xyz</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/parcorr.html#ParCorr.get_dependence_measure"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.ParCorr.get_dependence_measure" title="Permalink to this definition">¶</a></dt>
<dd><p>Return partial correlation.</p>
<p>Estimated as the Pearson correlation of the residuals of a linear
OLS regression.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>array</strong> (<em>array-like</em>) – data array with X, Y, Z in rows and observations in columns</p></li>
<li><p><strong>xyz</strong> (<em>array of ints</em>) – XYZ identifier array of shape (dim,).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>val</strong> – Partial correlation coefficient.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.ParCorr.get_model_selection_criterion">
<span class="sig-name descname"><span class="pre">get_model_selection_criterion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">j</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parents</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">corrected_aic</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/parcorr.html#ParCorr.get_model_selection_criterion"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.ParCorr.get_model_selection_criterion" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns Akaike’s Information criterion modulo constants.</p>
<p>Fits a linear model of the parents to variable j and returns the
score. Leave-one-out cross-validation is asymptotically equivalent to
AIC for ordinary linear regression models. Here used to determine
optimal hyperparameters in PCMCI, in particular the pc_alpha value.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>j</strong> (<em>int</em>) – Index of target variable in data array.</p></li>
<li><p><strong>parents</strong> (<em>list</em>) – List of form [(0, -1), (3, -2), …] containing parents.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Maximum time lag. This may be used to make sure that estimates for
different lags in X, Z, all have the same sample size.</p></li>
<li><p><strong>Returns</strong> – </p></li>
<li><p><strong>score</strong> (<em>float</em>) – Model score.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.ParCorr.get_shuffle_significance">
<span class="sig-name descname"><span class="pre">get_shuffle_significance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xyz</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_null_dist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/parcorr.html#ParCorr.get_shuffle_significance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.ParCorr.get_shuffle_significance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns p-value for shuffle significance test.</p>
<p>For residual-based test statistics only the residuals are shuffled.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>array</strong> (<em>array-like</em>) – data array with X, Y, Z in rows and observations in columns</p></li>
<li><p><strong>xyz</strong> (<em>array of ints</em>) – XYZ identifier array of shape (dim,).</p></li>
<li><p><strong>value</strong> (<em>number</em>) – Value of test statistic for unshuffled estimate.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>pval</strong> – p-value</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tigramite.independence_tests.ParCorr.measure">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">measure</span></span><a class="headerlink" href="#tigramite.independence_tests.ParCorr.measure" title="Permalink to this definition">¶</a></dt>
<dd><p>Concrete property to return the measure of the independence test</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="tigramite.independence_tests.GPDC">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tigramite.independence_tests.</span></span><span class="sig-name descname"><span class="pre">GPDC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">null_dist_filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gp_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/gpdc.html#GPDC"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.GPDC" title="Permalink to this definition">¶</a></dt>
<dd><p>GPDC conditional independence test based on Gaussian processes and distance correlation.</p>
<p>GPDC is based on a Gaussian process (GP) regression and a distance
correlation test on the residuals <a class="footnote-reference brackets" href="#id20" id="id18" role="doc-noteref"><span class="fn-bracket">[</span>2<span class="fn-bracket">]</span></a>. GP is estimated with scikit-learn
and allows to flexibly specify kernels and hyperparameters or let them be
optimized automatically. The distance correlation test is implemented with
cython. Here the null distribution is not analytically available, but can be
precomputed with the function generate_and_save_nulldists(…) which saves a
*.npz file containing the null distribution for different sample sizes.
This file can then be supplied as null_dist_filename.</p>
<p class="rubric">Notes</p>
<p>GPDC is based on a Gaussian process (GP) regression and a distance
correlation test on the residuals. Distance correlation is described in
<a class="footnote-reference brackets" href="#id20" id="id19" role="doc-noteref"><span class="fn-bracket">[</span>2<span class="fn-bracket">]</span></a>. To test <img class="math" src="_images/math/667eda4bf3d5ce33b6cc785cadfef79bb95741ca.png" alt="X \perp Y | Z"/>, first <img class="math" src="_images/math/95f028ab2b20b895fa12d986e0d9f40f7b6e52d3.png" alt="Z"/> is regressed out from
<img class="math" src="_images/math/ed38fa24f1c94891bd312012aab3f6673be3eb83.png" alt="X"/> and <img class="math" src="_images/math/7daf0d4815e763eb90f0d5f1dc406f668c1e21db.png" alt="Y"/> assuming the  model</p>
<div class="math">
<p><img src="_images/math/db3d34854a6f48587cf5b9a41df90ad1c5e332d6.png" alt="X &amp; =  f_X(Z) + \epsilon_{X} \\
Y &amp; =  f_Y(Z) + \epsilon_{Y}  \\
\epsilon_{X,Y} &amp;\sim \mathcal{N}(0, \sigma^2)"/></p>
</div><p>using GP regression. Here <img class="math" src="_images/math/5406eadc281dbd20de843b0034c8497320dae5cb.png" alt="\sigma^2"/> and the kernel bandwidth are
optimzed using <code class="docutils literal notranslate"><span class="pre">sklearn</span></code>. Then the residuals  are transformed to uniform
marginals yielding <img class="math" src="_images/math/be07aa32325c7a6161c0cef04f9b702054873211.png" alt="r_X,r_Y"/> and their dependency is tested with</p>
<div class="math">
<p><img src="_images/math/5f8c562c89b6bf12d27dc6cdc9dc090f7bb78e9c.png" alt="\mathcal{R}\left(r_X, r_Y\right)"/></p>
</div><p>The null distribution of the distance correlation should be pre-computed.
Otherwise it is computed during runtime.</p>
<p class="rubric">References</p>
<aside class="footnote brackets" id="id20" role="note">
<span class="label"><span class="fn-bracket">[</span>2<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id18">1</a>,<a role="doc-backlink" href="#id19">2</a>,<a role="doc-backlink" href="#id21">3</a>,<a role="doc-backlink" href="#id22">4</a>)</span>
<p>Gabor J. Szekely, Maria L. Rizzo, and Nail K. Bakirov: Measuring and
testing dependence by correlation of distances,
<a class="reference external" href="https://arxiv.org/abs/0803.4101">https://arxiv.org/abs/0803.4101</a></p>
</aside>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>null_dist_filename</strong> (<em>str</em><em>, </em><em>otional</em><em> (</em><em>default: None</em><em>)</em>) – Path to file containing null distribution.</p></li>
<li><p><strong>gp_params</strong> (<em>dictionary</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Dictionary with parameters for <code class="docutils literal notranslate"><span class="pre">GaussianProcessRegressor</span></code>.</p></li>
<li><p><strong>**kwargs</strong> – Arguments passed on to parent class GaussProcReg.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.GPDC.generate_and_save_nulldists">
<span class="sig-name descname"><span class="pre">generate_and_save_nulldists</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_sizes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">null_dist_filename</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/gpdc.html#GPDC.generate_and_save_nulldists"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.GPDC.generate_and_save_nulldists" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates and saves null distribution for pairwise independence
tests.</p>
<p>Generates the null distribution for different sample sizes. Calls
generate_nulldist. Null dists are saved to disk as
self.null_dist_filename.npz. Also adds the null distributions to
self.gauss_pr.null_dists.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sizes</strong> (<em>list</em>) – List of sample sizes.</p></li>
<li><p><strong>null_dist_filename</strong> (<em>str</em>) – Name to save file containing null distributions.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.GPDC.generate_nulldist">
<span class="sig-name descname"><span class="pre">generate_nulldist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_to_null_dists</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/gpdc.html#GPDC.generate_nulldist"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.GPDC.generate_nulldist" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates null distribution for pairwise independence tests.</p>
<p>Generates the null distribution for sample size df. Assumes pairwise
samples transformed to uniform marginals. Uses get_dependence_measure
available in class and generates self.sig_samples random samples. Adds
the null distributions to self.gauss_pr.null_dists.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>int</em>) – Degrees of freedom / sample size to generate null distribution for.</p></li>
<li><p><strong>add_to_null_dists</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: True</em><em>)</em>) – Whether to add the null dist to the dictionary of null dists or
just return it.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>null_dist</strong> – Only returned,if add_to_null_dists is False.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>array of shape [df,]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.GPDC.get_analytic_significance">
<span class="sig-name descname"><span class="pre">get_analytic_significance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">T</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dim</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/gpdc.html#GPDC.get_analytic_significance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.GPDC.get_analytic_significance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns p-value for the distance correlation coefficient.</p>
<p>The null distribution for necessary degrees of freedom (df) is loaded.
If not available, the null distribution is generated with the function
generate_nulldist(). It is recommended to generate the nulldists for a
wide range of sample sizes beforehand with the function
generate_and_save_nulldists(…). The distance correlation coefficient
is one-sided. If the degrees of freedom are less than 1, numpy.nan is
returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>value</strong> (<em>float</em>) – Test statistic value.</p></li>
<li><p><strong>T</strong> (<em>int</em>) – Sample length</p></li>
<li><p><strong>dim</strong> (<em>int</em>) – Dimensionality, ie, number of features.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>pval</strong> – p-value.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float or numpy.nan</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.GPDC.get_dependence_measure">
<span class="sig-name descname"><span class="pre">get_dependence_measure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xyz</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/gpdc.html#GPDC.get_dependence_measure"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.GPDC.get_dependence_measure" title="Permalink to this definition">¶</a></dt>
<dd><p>Return GPDC measure.</p>
<p>Estimated as the distance correlation of the residuals of a GP
regression.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>array</strong> (<em>array-like</em>) – data array with X, Y, Z in rows and observations in columns</p></li>
<li><p><strong>xyz</strong> (<em>array of ints</em>) – XYZ identifier array of shape (dim,).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>val</strong> – GPDC test statistic.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.GPDC.get_model_selection_criterion">
<span class="sig-name descname"><span class="pre">get_model_selection_criterion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">j</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parents</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/gpdc.html#GPDC.get_model_selection_criterion"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.GPDC.get_model_selection_criterion" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns log marginal likelihood for GP regression.</p>
<p>Fits a GP model of the parents to variable j and returns the negative
log marginal likelihood as a model selection score. Is used to determine
optimal hyperparameters in PCMCI, in particular the pc_alpha value.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>j</strong> (<em>int</em>) – Index of target variable in data array.</p></li>
<li><p><strong>parents</strong> (<em>list</em>) – List of form [(0, -1), (3, -2), …] containing parents.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Maximum time lag. This may be used to make sure that estimates for
different lags in X, Z, all have the same sample size.</p></li>
<li><p><strong>Returns</strong> – </p></li>
<li><p><strong>score</strong> (<em>float</em>) – Model score.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.GPDC.get_shuffle_significance">
<span class="sig-name descname"><span class="pre">get_shuffle_significance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xyz</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_null_dist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/gpdc.html#GPDC.get_shuffle_significance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.GPDC.get_shuffle_significance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns p-value for shuffle significance test.</p>
<p>For residual-based test statistics only the residuals are shuffled.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>array</strong> (<em>array-like</em>) – data array with X, Y, Z in rows and observations in columns</p></li>
<li><p><strong>xyz</strong> (<em>array of ints</em>) – XYZ identifier array of shape (dim,).</p></li>
<li><p><strong>value</strong> (<em>number</em>) – Value of test statistic for unshuffled estimate.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>pval</strong> – p-value</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tigramite.independence_tests.GPDC.measure">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">measure</span></span><a class="headerlink" href="#tigramite.independence_tests.GPDC.measure" title="Permalink to this definition">¶</a></dt>
<dd><p>Concrete property to return the measure of the independence test</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="tigramite.independence_tests.GPDCtorch">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tigramite.independence_tests.</span></span><span class="sig-name descname"><span class="pre">GPDCtorch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">null_dist_filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/gpdc_torch.html#GPDCtorch"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.GPDCtorch" title="Permalink to this definition">¶</a></dt>
<dd><p>GPDC conditional independence test based on Gaussian processes and distance correlation. Here with gpytorch implementation.</p>
<p>GPDC is based on a Gaussian process (GP) regression and a distance
correlation test on the residuals <a class="footnote-reference brackets" href="#id20" id="id21" role="doc-noteref"><span class="fn-bracket">[</span>2<span class="fn-bracket">]</span></a>. GP is estimated with gpytorch.
The distance correlation test is implemented with the dcor package available
from pip. Here the null distribution is not analytically available, but can be
precomputed with the function generate_and_save_nulldists(…) which saves a
*.npz file containing the null distribution for different sample sizes.
This file can then be supplied as null_dist_filename.</p>
<p class="rubric">Notes</p>
<p>GPDC is based on a Gaussian process (GP) regression and a distance
correlation test on the residuals. Distance correlation is described in
<a class="footnote-reference brackets" href="#id20" id="id22" role="doc-noteref"><span class="fn-bracket">[</span>2<span class="fn-bracket">]</span></a>. To test <img class="math" src="_images/math/667eda4bf3d5ce33b6cc785cadfef79bb95741ca.png" alt="X \perp Y | Z"/>, first <img class="math" src="_images/math/95f028ab2b20b895fa12d986e0d9f40f7b6e52d3.png" alt="Z"/> is regressed out from
<img class="math" src="_images/math/ed38fa24f1c94891bd312012aab3f6673be3eb83.png" alt="X"/> and <img class="math" src="_images/math/7daf0d4815e763eb90f0d5f1dc406f668c1e21db.png" alt="Y"/> assuming the  model</p>
<div class="math">
<p><img src="_images/math/db3d34854a6f48587cf5b9a41df90ad1c5e332d6.png" alt="X &amp; =  f_X(Z) + \epsilon_{X} \\
Y &amp; =  f_Y(Z) + \epsilon_{Y}  \\
\epsilon_{X,Y} &amp;\sim \mathcal{N}(0, \sigma^2)"/></p>
</div><p>using GP regression. Here <img class="math" src="_images/math/5406eadc281dbd20de843b0034c8497320dae5cb.png" alt="\sigma^2"/> and the kernel bandwidth are
optimzed using <code class="docutils literal notranslate"><span class="pre">gpytorch</span></code>. Then the residuals  are transformed to uniform
marginals yielding <img class="math" src="_images/math/be07aa32325c7a6161c0cef04f9b702054873211.png" alt="r_X,r_Y"/> and their dependency is tested with</p>
<div class="math">
<p><img src="_images/math/5f8c562c89b6bf12d27dc6cdc9dc090f7bb78e9c.png" alt="\mathcal{R}\left(r_X, r_Y\right)"/></p>
</div><p>The null distribution of the distance correlation should be pre-computed.
Otherwise it is computed during runtime.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>null_dist_filename</strong> (<em>str</em><em>, </em><em>otional</em><em> (</em><em>default: None</em><em>)</em>) – Path to file containing null distribution.</p></li>
<li><p><strong>**kwargs</strong> – Arguments passed on to parent class GaussProcRegTorch.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.GPDCtorch.generate_and_save_nulldists">
<span class="sig-name descname"><span class="pre">generate_and_save_nulldists</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_sizes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">null_dist_filename</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/gpdc_torch.html#GPDCtorch.generate_and_save_nulldists"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.GPDCtorch.generate_and_save_nulldists" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates and saves null distribution for pairwise independence
tests.</p>
<p>Generates the null distribution for different sample sizes. Calls
generate_nulldist. Null dists are saved to disk as
self.null_dist_filename.npz. Also adds the null distributions to
self.gauss_pr.null_dists.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sizes</strong> (<em>list</em>) – List of sample sizes.</p></li>
<li><p><strong>null_dist_filename</strong> (<em>str</em>) – Name to save file containing null distributions.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.GPDCtorch.generate_nulldist">
<span class="sig-name descname"><span class="pre">generate_nulldist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_to_null_dists</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/gpdc_torch.html#GPDCtorch.generate_nulldist"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.GPDCtorch.generate_nulldist" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates null distribution for pairwise independence tests.</p>
<p>Generates the null distribution for sample size df. Assumes pairwise
samples transformed to uniform marginals. Uses get_dependence_measure
available in class and generates self.sig_samples random samples. Adds
the null distributions to self.gauss_pr.null_dists.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>int</em>) – Degrees of freedom / sample size to generate null distribution for.</p></li>
<li><p><strong>add_to_null_dists</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: True</em><em>)</em>) – Whether to add the null dist to the dictionary of null dists or
just return it.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>null_dist</strong> – Only returned,if add_to_null_dists is False.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>array of shape [df,]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.GPDCtorch.get_analytic_significance">
<span class="sig-name descname"><span class="pre">get_analytic_significance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">T</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dim</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/gpdc_torch.html#GPDCtorch.get_analytic_significance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.GPDCtorch.get_analytic_significance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns p-value for the distance correlation coefficient.</p>
<p>The null distribution for necessary degrees of freedom (df) is loaded.
If not available, the null distribution is generated with the function
generate_nulldist(). It is recommended to generate the nulldists for a
wide range of sample sizes beforehand with the function
generate_and_save_nulldists(…). The distance correlation coefficient
is one-sided. If the degrees of freedom are less than 1, numpy.nan is
returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>value</strong> (<em>float</em>) – Test statistic value.</p></li>
<li><p><strong>T</strong> (<em>int</em>) – Sample length</p></li>
<li><p><strong>dim</strong> (<em>int</em>) – Dimensionality, ie, number of features.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>pval</strong> – p-value.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float or numpy.nan</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.GPDCtorch.get_dependence_measure">
<span class="sig-name descname"><span class="pre">get_dependence_measure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xyz</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/gpdc_torch.html#GPDCtorch.get_dependence_measure"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.GPDCtorch.get_dependence_measure" title="Permalink to this definition">¶</a></dt>
<dd><p>Return GPDC measure.</p>
<p>Estimated as the distance correlation of the residuals of a GP
regression.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>array</strong> (<em>array-like</em>) – data array with X, Y, Z in rows and observations in columns</p></li>
<li><p><strong>xyz</strong> (<em>array of ints</em>) – XYZ identifier array of shape (dim,).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>val</strong> – GPDC test statistic.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.GPDCtorch.get_model_selection_criterion">
<span class="sig-name descname"><span class="pre">get_model_selection_criterion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">j</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parents</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/gpdc_torch.html#GPDCtorch.get_model_selection_criterion"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.GPDCtorch.get_model_selection_criterion" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns log marginal likelihood for GP regression.</p>
<p>Fits a GP model of the parents to variable j and returns the negative
log marginal likelihood as a model selection score. Is used to determine
optimal hyperparameters in PCMCI, in particular the pc_alpha value.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>j</strong> (<em>int</em>) – Index of target variable in data array.</p></li>
<li><p><strong>parents</strong> (<em>list</em>) – List of form [(0, -1), (3, -2), …] containing parents.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Maximum time lag. This may be used to make sure that estimates for
different lags in X, Z, all have the same sample size.</p></li>
<li><p><strong>Returns</strong> – </p></li>
<li><p><strong>score</strong> (<em>float</em>) – Model score.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.GPDCtorch.get_shuffle_significance">
<span class="sig-name descname"><span class="pre">get_shuffle_significance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xyz</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_null_dist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/gpdc_torch.html#GPDCtorch.get_shuffle_significance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.GPDCtorch.get_shuffle_significance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns p-value for shuffle significance test.</p>
<p>For residual-based test statistics only the residuals are shuffled.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>array</strong> (<em>array-like</em>) – data array with X, Y, Z in rows and observations in columns</p></li>
<li><p><strong>xyz</strong> (<em>array of ints</em>) – XYZ identifier array of shape (dim,).</p></li>
<li><p><strong>value</strong> (<em>number</em>) – Value of test statistic for unshuffled estimate.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>pval</strong> – p-value</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tigramite.independence_tests.GPDCtorch.measure">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">measure</span></span><a class="headerlink" href="#tigramite.independence_tests.GPDCtorch.measure" title="Permalink to this definition">¶</a></dt>
<dd><p>Concrete property to return the measure of the independence test</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="tigramite.independence_tests.CMIknn">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tigramite.independence_tests.</span></span><span class="sig-name descname"><span class="pre">CMIknn</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">knn</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shuffle_neighbors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">significance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'shuffle_test'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transform</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ranks'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">workers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-</span> <span class="pre">1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/cmiknn.html#CMIknn"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.CMIknn" title="Permalink to this definition">¶</a></dt>
<dd><p>Conditional mutual information test based on nearest-neighbor estimator.</p>
<p>Conditional mutual information is the most general dependency measure coming
from an information-theoretic framework. It makes no assumptions about the
parametric form of the dependencies by directly estimating the underlying
joint density. The test here is based on the estimator in  S. Frenzel and B.
Pompe, Phys. Rev. Lett. 99, 204101 (2007), combined with a shuffle test to
generate  the distribution under the null hypothesis of independence first
used in <a class="footnote-reference brackets" href="#id24" id="id23" role="doc-noteref"><span class="fn-bracket">[</span>3<span class="fn-bracket">]</span></a>. The knn-estimator is suitable only for variables taking a
continuous range of values. For discrete variables use the CMIsymb class.</p>
<p class="rubric">Notes</p>
<p>CMI is given by</p>
<div class="math">
<p><img src="_images/math/b62df33f46595ed60c51f255186ee346c1fcc0cb.png" alt="I(X;Y|Z) &amp;= \int p(z)  \iint  p(x,y|z) \log
\frac{ p(x,y |z)}{p(x|z)\cdot p(y |z)} \,dx dy dz"/></p>
</div><p>Its knn-estimator is given by</p>
<div class="math">
<p><img src="_images/math/42858b59a7270363c15ca14b0d5fc56d33af1f8a.png" alt="\widehat{I}(X;Y|Z)  &amp;=   \psi (k) + \frac{1}{T} \sum_{t=1}^T
\left[ \psi(k_{Z,t}) - \psi(k_{XZ,t}) - \psi(k_{YZ,t}) \right]"/></p>
</div><p>where <img class="math" src="_images/math/b6c245d487949782a89cab9ee83504a62fdc2337.png" alt="\psi"/> is the Digamma function.  This estimator has as a
parameter the number of nearest-neighbors <img class="math" src="_images/math/9630132210b904754c9ab272b61cb527d12263ca.png" alt="k"/> which determines the
size of hyper-cubes around each (high-dimensional) sample point. Then
<img class="math" src="_images/math/26b09a13f97c2e89eb7687980b95a54839775fc8.png" alt="k_{Z,},k_{XZ},k_{YZ}"/> are the numbers of neighbors in the respective
subspaces.</p>
<p><img class="math" src="_images/math/9630132210b904754c9ab272b61cb527d12263ca.png" alt="k"/> can be viewed as a density smoothing parameter (although it is
data-adaptive unlike fixed-bandwidth estimators). For large <img class="math" src="_images/math/9630132210b904754c9ab272b61cb527d12263ca.png" alt="k"/>, the
underlying dependencies are more smoothed and CMI has a larger bias,
but lower variance, which is more important for significance testing. Note
that the estimated CMI values can be slightly negative while CMI is a non-
negative quantity.</p>
<p>This method requires the scipy.spatial.cKDTree package.</p>
<p class="rubric">References</p>
<aside class="footnote brackets" id="id24" role="note">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id23">3</a><span class="fn-bracket">]</span></span>
<p>J. Runge (2018): Conditional Independence Testing Based on a
Nearest-Neighbor Estimator of Conditional Mutual Information.
In Proceedings of the 21st International Conference on Artificial
Intelligence and Statistics.
<a class="reference external" href="http://proceedings.mlr.press/v84/runge18a.html">http://proceedings.mlr.press/v84/runge18a.html</a></p>
</aside>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>knn</strong> (<em>int</em><em> or </em><em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 0.2</em><em>)</em>) – Number of nearest-neighbors which determines the size of hyper-cubes
around each (high-dimensional) sample point. If smaller than 1, this is
computed as a fraction of T, hence knn=knn*T. For knn larger or equal to
1, this is the absolute number.</p></li>
<li><p><strong>shuffle_neighbors</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 5</em><em>)</em>) – Number of nearest-neighbors within Z for the shuffle surrogates which
determines the size of hyper-cubes around each (high-dimensional) sample
point.</p></li>
<li><p><strong>transform</strong> (<em>{'ranks'</em><em>, </em><em>'standardize'</em><em>,  </em><em>'uniform'</em><em>, </em><em>False}</em><em>, </em><em>optional</em>) – (default: ‘ranks’)
Whether to transform the array beforehand by standardizing
or transforming to uniform marginals.</p></li>
<li><p><strong>workers</strong> (<em>int</em><em> (</em><em>optional</em><em>, </em><em>default = -1</em><em>)</em>) – Number of workers to use for parallel processing. If -1 is given
all processors are used. Default: -1.</p></li>
<li><p><strong>significance</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'shuffle_test'</em><em>)</em>) – Type of significance test to use. For CMIknn only ‘fixed_thres’ and
‘shuffle_test’ are available.</p></li>
<li><p><strong>**kwargs</strong> – Arguments passed on to parent class CondIndTest.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.CMIknn.get_conditional_entropy">
<span class="sig-name descname"><span class="pre">get_conditional_entropy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xyz</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/cmiknn.html#CMIknn.get_conditional_entropy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.CMIknn.get_conditional_entropy" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the nearest-neighbor conditional entropy estimate of H(X|Y).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>array</strong> (<em>array-like</em>) – data array with X, Y in rows and observations in columns</p></li>
<li><p><strong>xyz</strong> (<em>array of ints</em>) – XYZ identifier array of shape (dim,). Here only uses 0 for X and
1 for Y.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>val</strong> – Entropy estimate.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.CMIknn.get_dependence_measure">
<span class="sig-name descname"><span class="pre">get_dependence_measure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xyz</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/cmiknn.html#CMIknn.get_dependence_measure"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.CMIknn.get_dependence_measure" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns CMI estimate as described in Frenzel and Pompe PRL (2007).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>array</strong> (<em>array-like</em>) – data array with X, Y, Z in rows and observations in columns</p></li>
<li><p><strong>xyz</strong> (<em>array of ints</em>) – XYZ identifier array of shape (dim,).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>val</strong> – Conditional mutual information estimate.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.CMIknn.get_shuffle_significance">
<span class="sig-name descname"><span class="pre">get_shuffle_significance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xyz</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_null_dist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/cmiknn.html#CMIknn.get_shuffle_significance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.CMIknn.get_shuffle_significance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns p-value for nearest-neighbor shuffle significance test.</p>
<p>For non-empty Z, overwrites get_shuffle_significance from the parent
class  which is a block shuffle test, which does not preserve
dependencies of X and Y with Z. Here the parameter shuffle_neighbors is
used to permute only those values <img class="math" src="_images/math/c67734af70861b2bd4dedf5c41c9aad231466f84.png" alt="x_i"/> and <img class="math" src="_images/math/ab9afdaf786ce53318d75d81f050af8560822fcd.png" alt="x_j"/> for which
<img class="math" src="_images/math/1a2da53015c9d2f2c52257723f812bf512b6818e.png" alt="z_j"/> is among the nearest niehgbors of <img class="math" src="_images/math/c0d0f97cd9bb4e6571e2689163f9f2989b304f55.png" alt="z_i"/>. If Z is
empty, the block-shuffle test is used.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>array</strong> (<em>array-like</em>) – data array with X, Y, Z in rows and observations in columns</p></li>
<li><p><strong>xyz</strong> (<em>array of ints</em>) – XYZ identifier array of shape (dim,).</p></li>
<li><p><strong>value</strong> (<em>number</em>) – Value of test statistic for unshuffled estimate.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>pval</strong> – p-value</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tigramite.independence_tests.CMIknn.measure">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">measure</span></span><a class="headerlink" href="#tigramite.independence_tests.CMIknn.measure" title="Permalink to this definition">¶</a></dt>
<dd><p>Concrete property to return the measure of the independence test</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="tigramite.independence_tests.CMIsymb">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tigramite.independence_tests.</span></span><span class="sig-name descname"><span class="pre">CMIsymb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_symbs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">significance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'shuffle_test'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_blocklength</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conf_blocklength</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/cmisymb.html#CMIsymb"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.CMIsymb" title="Permalink to this definition">¶</a></dt>
<dd><p>Conditional mutual information test based on discrete estimator.</p>
<p>Conditional mutual information is the most general dependency measure
coming from an information-theoretic framework. It makes no assumptions
about the parametric form of the dependencies by directly estimating the
underlying joint density. The test here is based on directly estimating
the joint distribution assuming symbolic input, combined with a
shuffle test to generate  the distribution under the null hypothesis of
independence. The knn-estimator is suitable only for discrete variables.
For continuous variables, either pre-process the data using the functions
in data_processing or, better, use the CMIknn class.</p>
<p class="rubric">Notes</p>
<p>CMI and its estimator are given by</p>
<div class="math">
<p><img src="_images/math/d226d17c031c4b264fb8f59f953381951cc2e9b4.png" alt="I(X;Y|Z) &amp;= \sum p(z)  \sum \sum  p(x,y|z) \log
\frac{ p(x,y |z)}{p(x|z)\cdot p(y |z)} \,dx dy dz"/></p>
</div><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n_symbs</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Number of symbols in input data. Should be at least as large as the
maximum array entry + 1. If None, n_symbs is based on the
maximum value in the array (array.max() + 1).</p></li>
<li><p><strong>significance</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'shuffle_test'</em><em>)</em>) – Type of significance test to use. For CMIsymb only ‘fixed_thres’ and
‘shuffle_test’ are available.</p></li>
<li><p><strong>sig_blocklength</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 1</em><em>)</em>) – Block length for block-shuffle significance test.</p></li>
<li><p><strong>conf_blocklength</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 1</em><em>)</em>) – Block length for block-bootstrap.</p></li>
<li><p><strong>**kwargs</strong> – Arguments passed on to parent class CondIndTest.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.CMIsymb.get_dependence_measure">
<span class="sig-name descname"><span class="pre">get_dependence_measure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xyz</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/cmisymb.html#CMIsymb.get_dependence_measure"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.CMIsymb.get_dependence_measure" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns CMI estimate based on bincount histogram.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>array</strong> (<em>array-like</em>) – data array with X, Y, Z in rows and observations in columns</p></li>
<li><p><strong>xyz</strong> (<em>array of ints</em>) – XYZ identifier array of shape (dim,).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>val</strong> – Conditional mutual information estimate.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.CMIsymb.get_shuffle_significance">
<span class="sig-name descname"><span class="pre">get_shuffle_significance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xyz</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_null_dist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/cmisymb.html#CMIsymb.get_shuffle_significance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.CMIsymb.get_shuffle_significance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns p-value for shuffle significance test.</p>
<p>For residual-based test statistics only the residuals are shuffled.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>array</strong> (<em>array-like</em>) – data array with X, Y, Z in rows and observations in columns</p></li>
<li><p><strong>xyz</strong> (<em>array of ints</em>) – XYZ identifier array of shape (dim,).</p></li>
<li><p><strong>value</strong> (<em>number</em>) – Value of test statistic for unshuffled estimate.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>pval</strong> – p-value</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tigramite.independence_tests.CMIsymb.measure">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">measure</span></span><a class="headerlink" href="#tigramite.independence_tests.CMIsymb.measure" title="Permalink to this definition">¶</a></dt>
<dd><p>Concrete property to return the measure of the independence test</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="tigramite.independence_tests.OracleCI">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tigramite.independence_tests.</span></span><span class="sig-name descname"><span class="pre">OracleCI</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">links</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">observed_vars</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">selection_vars</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph_is_mag</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/oracle_conditional_independence.html#OracleCI"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.OracleCI" title="Permalink to this definition">¶</a></dt>
<dd><p>Oracle of conditional independence test X _|_ Y | Z given a graph.</p>
<p>Class around link_coeff causal ground truth. X _|_ Y | Z is based on
assessing whether X and Y are d-separated given Z in the graph.</p>
<p>Class can be used just like a Tigramite conditional independence class
(e.g., ParCorr). The main use is for unit testing of PCMCI methods.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>graph</strong> (<em>array of shape</em><em> [</em><em>N</em><em>, </em><em>N</em><em>, </em><em>tau_max+1</em><em>]</em>) – Causal graph.</p></li>
<li><p><strong>links</strong> (<em>dict</em>) – Dictionary of form {0:[(0, -1), …], 1:[…], …}.
Alternatively can also digest {0: [((0, -1), coeff, func)], …}.</p></li>
<li><p><strong>observed_vars</strong> (<em>None</em><em> or </em><em>list</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Subset of keys in links definining which variables are
observed. If None, then all variables are observed.</p></li>
<li><p><strong>selection_vars</strong> (<em>None</em><em> or </em><em>list</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Subset of keys in links definining which variables are
selected (= always conditioned on at every time lag).
If None, then no variables are selected.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Level of verbosity.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.OracleCI.check_shortest_path">
<span class="sig-name descname"><span class="pre">check_shortest_path</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Z</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_lag</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">starts_with</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ends_with</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">forbidden_nodes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">only_non_causal_paths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_optimality_cond</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">optimality_cond_des_YM</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">optimality_cond_Y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/oracle_conditional_independence.html#OracleCI.check_shortest_path"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.OracleCI.check_shortest_path" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns path between X and Y given Z in the graph.</p>
<p>X, Y, Z are of the form (var, lag) for lag &lt;= 0. D-separation is
based on:</p>
<p>1. Assessing maximum time lag max_lag of last ancestor of any X, Y, Z
with non-blocked (by Z), non-repeating directed path towards X, Y, Z
in the graph. ‘non_repeating’ means that an ancestor X^i_{ t-   au_i}
with link X^i_{t-       au_i} –&gt; X^j_{ t-      au_j} is only included if
X^i_{t’-        au_i} –&gt; X^j_{ t’-     au_j} for t’!=t is not already part of
the ancestors.</p>
<p>2. Using the time series graph truncated at max_lag we then test
d-separation between X and Y conditional on Z using breadth-first
search of non-blocked paths according to d-separation rules including
selection variables.</p>
<p>Optionally only considers paths starting/ending with specific marks)
and makes available the ancestors up to max_lag of X, Y, Z. This may take
a very long time, however.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>list of tuples</em>) – List of variables chosen for testing paths.</p></li>
<li><p><strong>Y</strong> (<em>list of tuples</em>) – List of variables chosen for testing paths.</p></li>
<li><p><strong>Z</strong> (<em>list of tuples</em>) – List of variables chosen for testing paths.</p></li>
<li><p><strong>max_lag</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Used here to constrain the has_path function to the graph
truncated at max_lag instead of identifying the max_lag from
ancestral search.</p></li>
<li><p><strong>compute_ancestors</strong> (<em>bool</em>) – Whether to also make available the ancestors for X, Y, Z as
self.anc_all_x, self.anc_all_y, and self.anc_all_z, respectively.</p></li>
<li><p><strong>starts_with</strong> (<em>{None</em><em>, </em><em>'tail'</em><em>, </em><em>'arrohead'}</em>) – Whether to only consider paths starting with particular mark at X.</p></li>
<li><p><strong>ends_with</strong> (<em>{None</em><em>, </em><em>'tail'</em><em>, </em><em>'arrohead'}</em>) – Whether to only consider paths ending with particular mark at Y.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>path</strong> – Returns path or False if no path exists.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list or False</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.OracleCI.get_confidence">
<span class="sig-name descname"><span class="pre">get_confidence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Z</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/oracle_conditional_independence.html#OracleCI.get_confidence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.OracleCI.get_confidence" title="Permalink to this definition">¶</a></dt>
<dd><p>For compatibility with PCMCI.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.OracleCI.get_graph_from_links">
<span class="sig-name descname"><span class="pre">get_graph_from_links</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/oracle_conditional_independence.html#OracleCI.get_graph_from_links"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.OracleCI.get_graph_from_links" title="Permalink to this definition">¶</a></dt>
<dd><p>Constructs graph (DAG or MAG or ADMG) from links, observed_vars,
and selection_vars.</p>
<p>For ADMGs uses the Latent projection operation (Pearl 2009).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.OracleCI.get_links_from_graph">
<span class="sig-name descname"><span class="pre">get_links_from_graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/oracle_conditional_independence.html#OracleCI.get_links_from_graph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.OracleCI.get_links_from_graph" title="Permalink to this definition">¶</a></dt>
<dd><p>Constructs links_coeffs dictionary, observed_vars,
and selection_vars from graph array (MAG or DAG).</p>
<p>In the case of MAGs, for every &lt;-&gt; or — link further
latent and selection variables, respectively, are added.
This corresponds to a canonical DAG (Richardson Spirtes 2002).</p>
<p>For ADMGs “—” are not supported, but also links of type “+-&gt;”
exist, which corresponds to having both “–&gt;” and “&lt;-&gt;”.</p>
<p>Can be used to evaluate d-separation in MAG/DAGs.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.OracleCI.get_measure">
<span class="sig-name descname"><span class="pre">get_measure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Z</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/oracle_conditional_independence.html#OracleCI.get_measure"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.OracleCI.get_measure" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns dependence measure.</p>
<p>Returns 0 if X and Y are d-separated given Z in the graph and 1 else.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>list of tuples</em>) – X,Y,Z are of the form [(var, -tau)], where var specifies the
variable index in the observed_vars and tau the time lag.</p></li>
<li><p><strong>[</strong> (<em>Y</em>) – X,Y,Z are of the form [(var, -tau)], where var specifies the
variable index in the observed_vars and tau the time lag.</p></li>
<li><p><strong>Z</strong><strong>]</strong> (<em>list of tuples</em>) – X,Y,Z are of the form [(var, -tau)], where var specifies the
variable index in the observed_vars and tau the time lag.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Maximum time lag. This may be used to make sure that estimates for
different lags in X, Z, all have the same sample size.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>val</strong> – The test statistic value.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.OracleCI.get_model_selection_criterion">
<span class="sig-name descname"><span class="pre">get_model_selection_criterion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">j</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parents</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/oracle_conditional_independence.html#OracleCI.get_model_selection_criterion"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.OracleCI.get_model_selection_criterion" title="Permalink to this definition">¶</a></dt>
<dd><p>Base class assumption that this is not implemented.  Concrete classes
should override when possible.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tigramite.independence_tests.OracleCI.measure">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">measure</span></span><a class="headerlink" href="#tigramite.independence_tests.OracleCI.measure" title="Permalink to this definition">¶</a></dt>
<dd><p>Concrete property to return the measure of the independence test</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.OracleCI.run_test">
<span class="sig-name descname"><span class="pre">run_test</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Z</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut_off</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'2xtau_max'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/oracle_conditional_independence.html#OracleCI.run_test"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.OracleCI.run_test" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform oracle conditional independence test.</p>
<p>Calls the d-separation function.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>list of tuples</em>) – X,Y,Z are of the form [(var, -tau)], where var specifies the
variable index in the observed_vars and tau the time lag.</p></li>
<li><p><strong>Y</strong> (<em>list of tuples</em>) – X,Y,Z are of the form [(var, -tau)], where var specifies the
variable index in the observed_vars and tau the time lag.</p></li>
<li><p><strong>Z</strong> (<em>list of tuples</em>) – X,Y,Z are of the form [(var, -tau)], where var specifies the
variable index in the observed_vars and tau the time lag.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Not used here.</p></li>
<li><p><strong>cut_off</strong> (<em>{'2xtau_max'</em><em>, </em><em>'max_lag'</em><em>, </em><em>'max_lag_or_tau_max'}</em>) – Not used here.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>val, pval</strong> – The test statistic value and the p-value.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple of floats</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.OracleCI.set_dataframe">
<span class="sig-name descname"><span class="pre">set_dataframe</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataframe</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/oracle_conditional_independence.html#OracleCI.set_dataframe"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.OracleCI.set_dataframe" title="Permalink to this definition">¶</a></dt>
<dd><p>Dummy function.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="tigramite.independence_tests.ParCorrMult">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tigramite.independence_tests.</span></span><span class="sig-name descname"><span class="pre">ParCorrMult</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">correlation_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'max_corr'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/parcorr_mult.html#ParCorrMult"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.ParCorrMult" title="Permalink to this definition">¶</a></dt>
<dd><p>Partial correlation test for multivariate X and Y.</p>
<p>Multivariate partial correlation is estimated through ordinary least squares (OLS)
regression and some test for multivariate dependency among the residuals.</p>
<p class="rubric">Notes</p>
<p>To test <img class="math" src="_images/math/667eda4bf3d5ce33b6cc785cadfef79bb95741ca.png" alt="X \perp Y | Z"/>, first <img class="math" src="_images/math/95f028ab2b20b895fa12d986e0d9f40f7b6e52d3.png" alt="Z"/> is regressed out from
<img class="math" src="_images/math/ed38fa24f1c94891bd312012aab3f6673be3eb83.png" alt="X"/> and <img class="math" src="_images/math/7daf0d4815e763eb90f0d5f1dc406f668c1e21db.png" alt="Y"/> assuming the  model</p>
<div class="math">
<p><img src="_images/math/78b6e77cc610bf56f5c64cb6dcc6d6ee49f886f9.png" alt="X &amp; =  Z \beta_X + \epsilon_{X} \\
Y &amp; =  Z \beta_Y + \epsilon_{Y}"/></p>
</div><p>using OLS regression. Then different measures for the dependency among the residuals
can be used. Currently only a test for zero correlation on the maximum of the residuals’
correlation is performed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>correlation_type</strong> (<em>{'max_corr'}</em>) – Which dependency measure to use on residuals.</p></li>
<li><p><strong>**kwargs</strong> – Arguments passed on to Parent class CondIndTest.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.ParCorrMult.get_analytic_confidence">
<span class="sig-name descname"><span class="pre">get_analytic_confidence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conf_lev</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/parcorr_mult.html#ParCorrMult.get_analytic_confidence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.ParCorrMult.get_analytic_confidence" title="Permalink to this definition">¶</a></dt>
<dd><p>Base class assumption that this is not implemented.  Concrete classes
should override when possible.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.ParCorrMult.get_analytic_significance">
<span class="sig-name descname"><span class="pre">get_analytic_significance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">T</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dim</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/parcorr_mult.html#ParCorrMult.get_analytic_significance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.ParCorrMult.get_analytic_significance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns analytic p-value depending on correlation_type.</p>
<p>If the degrees of freedom are less than
1, numpy.nan is returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>value</strong> (<em>float</em>) – Test statistic value.</p></li>
<li><p><strong>T</strong> (<em>int</em>) – Sample length</p></li>
<li><p><strong>dim</strong> (<em>int</em>) – Dimensionality, ie, number of features.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>pval</strong> – P-value.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float or numpy.nan</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.ParCorrMult.get_dependence_measure">
<span class="sig-name descname"><span class="pre">get_dependence_measure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xyz</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/parcorr_mult.html#ParCorrMult.get_dependence_measure"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.ParCorrMult.get_dependence_measure" title="Permalink to this definition">¶</a></dt>
<dd><p>Return multivariate kernel correlation coefficient.</p>
<p>Estimated as some dependency measure on the
residuals of a linear OLS regression.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>array</strong> (<em>array-like</em>) – data array with X, Y, Z in rows and observations in columns</p></li>
<li><p><strong>xyz</strong> (<em>array of ints</em>) – XYZ identifier array of shape (dim,).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>val</strong> – Partial correlation coefficient.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.ParCorrMult.get_model_selection_criterion">
<span class="sig-name descname"><span class="pre">get_model_selection_criterion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">j</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parents</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/parcorr_mult.html#ParCorrMult.get_model_selection_criterion"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.ParCorrMult.get_model_selection_criterion" title="Permalink to this definition">¶</a></dt>
<dd><p>Base class assumption that this is not implemented.  Concrete classes
should override when possible.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.ParCorrMult.get_shuffle_significance">
<span class="sig-name descname"><span class="pre">get_shuffle_significance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xyz</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_null_dist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/parcorr_mult.html#ParCorrMult.get_shuffle_significance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.ParCorrMult.get_shuffle_significance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns p-value for shuffle significance test.</p>
<p>For residual-based test statistics only the residuals are shuffled.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>array</strong> (<em>array-like</em>) – data array with X, Y, Z in rows and observations in columns</p></li>
<li><p><strong>xyz</strong> (<em>array of ints</em>) – XYZ identifier array of shape (dim,).</p></li>
<li><p><strong>value</strong> (<em>number</em>) – Value of test statistic for unshuffled estimate.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>pval</strong> – p-value</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tigramite.independence_tests.ParCorrMult.measure">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">measure</span></span><a class="headerlink" href="#tigramite.independence_tests.ParCorrMult.measure" title="Permalink to this definition">¶</a></dt>
<dd><p>Concrete property to return the measure of the independence test</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.independence_tests.ParCorrMult.mult_corr">
<span class="sig-name descname"><span class="pre">mult_corr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xyz</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">standardize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/independence_tests/parcorr_mult.html#ParCorrMult.mult_corr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.independence_tests.ParCorrMult.mult_corr" title="Permalink to this definition">¶</a></dt>
<dd><p>Return multivariate dependency measure.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>array</strong> (<em>array-like</em>) – data array with X, Y in rows and observations in columns</p></li>
<li><p><strong>xyz</strong> (<em>array of ints</em>) – XYZ identifier array of shape (dim,).</p></li>
<li><p><strong>standardize</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: True</em><em>)</em>) – Whether to standardize the array beforehand. Must be used for
partial correlation.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>val</strong> – Multivariate dependency measure.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="tigramite-causal-effects-causal-effect-analysis">
<h1><code class="xref py py-mod docutils literal notranslate"><span class="pre">tigramite.causal_effects</span></code>: Causal Effect analysis<a class="headerlink" href="#tigramite-causal-effects-causal-effect-analysis" title="Permalink to this heading">¶</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="tigramite.causal_effects.CausalEffects">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tigramite.causal_effects.</span></span><span class="sig-name descname"><span class="pre">CausalEffects</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">S</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hidden_variables</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_SM_overlap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/causal_effects.html#CausalEffects"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.causal_effects.CausalEffects" title="Permalink to this definition">¶</a></dt>
<dd><p>Causal effect estimation.</p>
<p>Methods for the estimation of linear or non-parametric causal effects
between (potentially multivariate) X and Y (potentially conditional
on S) by (generalized) backdoor adjustment. Various graph types are
supported, also including hidden variables.</p>
<p>Linear and non-parametric estimators are based on sklearn. For the
linear case without hidden variables also an efficient estimation
based on Wright’s path coefficients is available. This estimator
also allows to estimate mediation effects.</p>
<p>See the corresponding paper <a class="footnote-reference brackets" href="#id26" id="id25" role="doc-noteref"><span class="fn-bracket">[</span>6<span class="fn-bracket">]</span></a> and tigramite tutorial for an
in-depth introduction.</p>
<p class="rubric">References</p>
<aside class="footnote brackets" id="id26" role="note">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id25">6</a><span class="fn-bracket">]</span></span>
<p>J. Runge, Necessary and sufficient graphical conditions for
optimal adjustment sets in causal graphical models with
hidden variables, Advances in Neural Information Processing
Systems, 2021, 34
<a class="reference external" href="https://proceedings.neurips.cc/paper/2021/hash/8485ae387a981d783f8764e508151cd9-Abstract.html">https://proceedings.neurips.cc/paper/2021/hash/8485ae387a981d783f8764e508151cd9-Abstract.html</a></p>
</aside>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>graph</strong> (<em>array of either shape</em><em> [</em><em>N</em><em>, </em><em>N</em><em>]</em><em>, </em><em>[</em><em>N</em><em>, </em><em>N</em><em>, </em><em>tau_max+1</em><em>]</em><em>, or </em><em>[</em><em>N</em><em>, </em><em>N</em><em>, </em><em>tau_max+1</em><em>, </em><em>tau_max+1</em><em>]</em>) – Different graph types are supported, see tutorial.</p></li>
<li><p><strong>X</strong> (<em>list of tuples</em>) – List of tuples [(i, -tau), …] containing cause variables.</p></li>
<li><p><strong>Y</strong> (<em>list of tuples</em>) – List of tuples [(j, 0), …] containing effect variables.</p></li>
<li><p><strong>S</strong> (<em>list of tuples</em>) – List of tuples [(i, -tau), …] containing conditioned variables.</p></li>
<li><p><strong>graph_type</strong> (<em>str</em>) – Type of graph.</p></li>
<li><p><strong>hidden_variables</strong> (<em>list of tuples</em>) – Hidden variables in format [(i, -tau), …]. The internal graph is
constructed by a latent projection.</p></li>
<li><p><strong>check_SM_overlap</strong> (<em>bool</em>) – Whether to check whether S overlaps with M.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Level of verbosity.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tigramite.causal_effects.CausalEffects.check_XYS_paths">
<span class="sig-name descname"><span class="pre">check_XYS_paths</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/causal_effects.html#CausalEffects.check_XYS_paths"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.causal_effects.CausalEffects.check_XYS_paths" title="Permalink to this definition">¶</a></dt>
<dd><p>Check whether one can remove nodes from X and Y with no proper causal paths.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>X, Y</strong></p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>cleaned lists of X and Y with irrelevant nodes removed.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.causal_effects.CausalEffects.check_optimality">
<span class="sig-name descname"><span class="pre">check_optimality</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/causal_effects.html#CausalEffects.check_optimality"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.causal_effects.CausalEffects.check_optimality" title="Permalink to this definition">¶</a></dt>
<dd><p>Check whether optimal adjustment set exists according to Thm. 3 in Runge NeurIPS 2021.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>optimality</strong> – Returns True if an optimal adjustment set exists, otherwise False.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.causal_effects.CausalEffects.fit_bootstrap_of">
<span class="sig-name descname"><span class="pre">fit_bootstrap_of</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method_args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">boot_samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">boot_blocklength</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/causal_effects.html#CausalEffects.fit_bootstrap_of"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.causal_effects.CausalEffects.fit_bootstrap_of" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs chosen method on bootstrap samples drawn from DataFrame.</p>
<p>Bootstraps for tau=0 are drawn from [max_lag, …, T] and all lagged
variables constructed in DataFrame.construct_array are consistently
shifted with respect to this bootsrap sample to ensure that lagged
relations in the bootstrap sample are preserved.</p>
<p>This function fits the models, predict_bootstrap_of can then be used
to get confidence intervals for the effect of interventions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>method</strong> (<em>str</em>) – Chosen method among valid functions in this class.</p></li>
<li><p><strong>method_args</strong> (<em>dict</em>) – Arguments passed to method.</p></li>
<li><p><strong>boot_samples</strong> (<em>int</em>) – Number of bootstrap samples to draw.</p></li>
<li><p><strong>boot_blocklength</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 1</em><em>)</em>) – Block length for block-bootstrap.</p></li>
<li><p><strong>seed</strong> (<em>int</em><em>, </em><em>optional</em><em>(</em><em>default = None</em><em>)</em>) – Seed for RandomState (default_rng)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.causal_effects.CausalEffects.fit_total_effect">
<span class="sig-name descname"><span class="pre">fit_total_effect</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataframe</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">estimator</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adjustment_set</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'optimal'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conditional_estimator</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_transform</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/causal_effects.html#CausalEffects.fit_total_effect"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.causal_effects.CausalEffects.fit_total_effect" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Returns a fitted model for the total causal effect of X on Y</dt><dd><p>conditional on S.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dataframe</strong> (<em>data object</em>) – Tigramite dataframe object. It must have the attributes dataframe.values
yielding a numpy array of shape (observations T, variables N) and
optionally a mask of the same shape and a missing values flag.</p></li>
<li><p><strong>estimator</strong> (<em>sklearn model object</em>) – For example, sklearn.linear_model.LinearRegression() for a linear
regression model.</p></li>
<li><p><strong>adjustment_set</strong> (<em>str</em><em> or </em><em>list of tuples</em>) – If ‘optimal’ the Oset is used, if ‘minimized_optimal’ the minimized Oset,
and if ‘colliders_minimized_optimal’, the colliders-minimized Oset.
If a list of tuples is passed, this set is used.</p></li>
<li><p><strong>conditional_estimator</strong> (<em>sklearn model object</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Used to fit conditional causal effects in nested regression.
If None, the same model as for estimator is used.</p></li>
<li><p><strong>data_transform</strong> (<em>sklearn preprocessing object</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Used to transform data prior to fitting. For example,
sklearn.preprocessing.StandardScaler for simple standardization. The
fitted parameters are stored.</p></li>
<li><p><strong>mask_type</strong> (<em>{None</em><em>, </em><em>'y'</em><em>,</em><em>'x'</em><em>,</em><em>'z'</em><em>,</em><em>'xy'</em><em>,</em><em>'xz'</em><em>,</em><em>'yz'</em><em>,</em><em>'xyz'}</em>) – Masking mode: Indicators for which variables in the dependence
measure I(X; Y | Z) the samples should be masked. If None, the mask
is not used. Explained in tutorial on masking and missing values.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.causal_effects.CausalEffects.fit_wright_effect">
<span class="sig-name descname"><span class="pre">fit_wright_effect</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataframe</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mediation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'parents'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">links_coeffs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_transform</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/causal_effects.html#CausalEffects.fit_wright_effect"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.causal_effects.CausalEffects.fit_wright_effect" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Returns a fitted model for the total or mediated causal effect of X on Y</dt><dd><p>potentially through mediator variables.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dataframe</strong> (<em>data object</em>) – Tigramite dataframe object. It must have the attributes dataframe.values
yielding a numpy array of shape (observations T, variables N) and
optionally a mask of the same shape and a missing values flag.</p></li>
<li><p><strong>mediation</strong> (<em>None</em><em>, </em><em>'direct'</em><em>, or </em><em>list of tuples</em>) – If None, total effect is estimated, if ‘direct’ then only the direct effect is estimated,
else only those causal paths are considerd that pass at least through one of these mediator nodes.</p></li>
<li><p><strong>method</strong> (<em>{'parents'</em><em>, </em><em>'links_coeffs'</em><em>, </em><em>'optimal'}</em>) – Method to use for estimating Wright’s path coefficients. If ‘optimal’,
the Oset is used, if ‘links_coeffs’, the coefficients in links_coeffs are used,
if ‘parents’, the parents are used (only valid for DAGs).</p></li>
<li><p><strong>links_coeffs</strong> (<em>dict</em>) – Only used if method = ‘links_coeffs’.
Dictionary of format: {0:[((i, -tau), coeff),…], 1:[…],
…} for all variables where i must be in [0..N-1] and tau &gt;= 0 with
number of variables N. coeff must be a float.</p></li>
<li><p><strong>data_transform</strong> (<em>None</em>) – Not implemented for Wright estimator. Complicated for missing samples.</p></li>
<li><p><strong>mask_type</strong> (<em>{None</em><em>, </em><em>'y'</em><em>,</em><em>'x'</em><em>,</em><em>'z'</em><em>,</em><em>'xy'</em><em>,</em><em>'xz'</em><em>,</em><em>'yz'</em><em>,</em><em>'xyz'}</em>) – Masking mode: Indicators for which variables in the dependence
measure I(X; Y | Z) the samples should be masked. If None, the mask
is not used. Explained in tutorial on masking and missing values.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.causal_effects.CausalEffects.get_graph_from_dict">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_graph_from_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">links</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/causal_effects.html#CausalEffects.get_graph_from_dict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.causal_effects.CausalEffects.get_graph_from_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Helper function to convert dictionary of links to graph array format.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>links</strong> (<em>dict</em>) – Dictionary of form {0:[((0, -1), coeff, func), …], 1:[…], …}.
Also format {0:[(0, -1), …], 1:[…], …} is allowed.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em><em> or </em><em>None</em>) – Maximum lag. If None, the maximum lag in links is used.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>graph</strong> – Matrix format of graph with 1 for true links and 0 else.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>array of shape (N, N, tau_max+1)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.causal_effects.CausalEffects.get_mediators">
<span class="sig-name descname"><span class="pre">get_mediators</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/causal_effects.html#CausalEffects.get_mediators"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.causal_effects.CausalEffects.get_mediators" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns mediator variables on proper causal paths.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> (<em>set</em>) – Set of start nodes.</p></li>
<li><p><strong>end</strong> (<em>set</em>) – Set of end nodes.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>mediators</strong> – Mediators on causal paths from start to end.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>set</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.causal_effects.CausalEffects.get_optimal_set">
<span class="sig-name descname"><span class="pre">get_optimal_set</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alternative_conditions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">minimize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_separate_sets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/causal_effects.html#CausalEffects.get_optimal_set"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.causal_effects.CausalEffects.get_optimal_set" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns optimal adjustment set.</p>
<p>See Runge NeurIPS 2021.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alternative_conditions</strong> (<em>set of tuples</em>) – Used only internally in optimality theorem. If None, self.S is used.</p></li>
<li><p><strong>minimize</strong> (<em>{False</em><em>, </em><em>True</em><em>, </em><em>'colliders_only'}</em>) – Minimize optimal set. If True, minimize such that no subset
can be removed without making it invalid. If ‘colliders_only’,
only colliders are minimized.</p></li>
<li><p><strong>return_separate_sets</strong> (<em>bool</em>) – Whether to return tuple of parents, colliders, collider_parents, and S.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>Oset_S</strong> – Returns optimal adjustment set if a valid set exists, otherwise False.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>False or list or tuple of lists</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.causal_effects.CausalEffects.predict_bootstrap_of">
<span class="sig-name descname"><span class="pre">predict_bootstrap_of</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method_args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conf_lev</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.9</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_individual_bootstrap_results</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/causal_effects.html#CausalEffects.predict_bootstrap_of"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.causal_effects.CausalEffects.predict_bootstrap_of" title="Permalink to this definition">¶</a></dt>
<dd><p>Predicts with fitted bootstraps.</p>
<p>To be used after fitting with fit_bootstrap_of. Only uses the
expected values of the predict function, not potential other output.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>method</strong> (<em>str</em>) – Chosen method among valid functions in this class.</p></li>
<li><p><strong>method_args</strong> (<em>dict</em>) – Arguments passed to method.</p></li>
<li><p><strong>conf_lev</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 0.9</em><em>)</em>) – Two-sided confidence interval.</p></li>
<li><p><strong>return_individual_bootstrap_results</strong> (<em>bool</em>) – Returns the individual bootstrap predictions.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>confidence_intervals</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy array of</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.causal_effects.CausalEffects.predict_total_effect">
<span class="sig-name descname"><span class="pre">predict_total_effect</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervention_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conditions_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pred_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_further_pred_results</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/causal_effects.html#CausalEffects.predict_total_effect"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.causal_effects.CausalEffects.predict_total_effect" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict effect of intervention with fitted model.</p>
<p>Uses the model.predict() function of the sklearn model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervention_data</strong> (<em>numpy array</em>) – Numpy array of shape (time, len(X)) that contains the do(X) values.</p></li>
<li><p><strong>conditions_data</strong> (<em>data object</em><em>, </em><em>optional</em>) – Numpy array of shape (time, len(S)) that contains the S=s values.</p></li>
<li><p><strong>pred_params</strong> (<em>dict</em><em>, </em><em>optional</em>) – Optional parameters passed on to sklearn prediction function.</p></li>
<li><p><strong>return_further_pred_results</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: False</em><em>)</em>) – In case the predictor class returns more than just the expected value,
the entire results can be returned.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>Results from prediction</strong> (<em>an array of shape  (time, len(Y)).</em>)</p></li>
<li><p><em>If estimate_confidence = True, then a tuple is returned.</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.causal_effects.CausalEffects.predict_wright_effect">
<span class="sig-name descname"><span class="pre">predict_wright_effect</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervention_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pred_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/causal_effects.html#CausalEffects.predict_wright_effect"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.causal_effects.CausalEffects.predict_wright_effect" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict linear effect of intervention with fitted Wright-model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervention_data</strong> (<em>numpy array</em>) – Numpy array of shape (time, len(X)) that contains the do(X) values.</p></li>
<li><p><strong>pred_params</strong> (<em>dict</em><em>, </em><em>optional</em>) – Optional parameters passed on to sklearn prediction function.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>Results from prediction</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>an array of shape  (time, len(Y)).</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="tigramite-models-time-series-modeling-mediation-and-prediction">
<h1><code class="xref py py-mod docutils literal notranslate"><span class="pre">tigramite.models</span></code>: Time series modeling, mediation, and prediction<a class="headerlink" href="#tigramite-models-time-series-modeling-mediation-and-prediction" title="Permalink to this heading">¶</a></h1>
<p>Base class:</p>
<dl class="py class">
<dt class="sig sig-object py" id="tigramite.models.Models">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tigramite.models.</span></span><span class="sig-name descname"><span class="pre">Models</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataframe</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conditional_model</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_transform</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">sklearn.preprocessing.StandardScaler</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#Models"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.Models" title="Permalink to this definition">¶</a></dt>
<dd><p>Base class for time series models.</p>
<p>Allows to fit any model from sklearn to the parents of a target variable.
Also takes care of missing values, masking and preprocessing.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dataframe</strong> (<em>data object</em>) – Tigramite dataframe object. It must have the attributes dataframe.values
yielding a numpy array of shape (observations T, variables N) and
optionally a mask of the same shape and a missing values flag.</p></li>
<li><p><strong>model</strong> (<em>sklearn model object</em>) – For example, sklearn.linear_model.LinearRegression() for a linear
regression model.</p></li>
<li><p><strong>conditional_model</strong> (<em>sklearn model object</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Used to fit conditional causal effects in nested regression.
If None, model is used.</p></li>
<li><p><strong>data_transform</strong> (<em>sklearn preprocessing object</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Used to transform data prior to fitting. For example,
sklearn.preprocessing.StandardScaler for simple standardization. The
fitted parameters are stored. Note that the inverse_transform is then
applied to the predicted data.</p></li>
<li><p><strong>mask_type</strong> (<em>{None</em><em>, </em><em>'y'</em><em>,</em><em>'x'</em><em>,</em><em>'z'</em><em>,</em><em>'xy'</em><em>,</em><em>'xz'</em><em>,</em><em>'yz'</em><em>,</em><em>'xyz'}</em>) – Masking mode: Indicators for which variables in the dependence
measure I(X; Y | Z) the samples should be masked. If None, the mask
is not used. Explained in tutorial on masking and missing values.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Level of verbosity.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.Models.get_coefs">
<span class="sig-name descname"><span class="pre">get_coefs</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#Models.get_coefs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.Models.get_coefs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns dictionary of coefficients for linear models.</p>
<p>Only for models from sklearn.linear_model</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>coeffs</strong> – Dictionary of dictionaries for each variable with keys given by the
parents and the regression coefficients as values.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dictionary</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.Models.get_fit">
<span class="sig-name descname"><span class="pre">get_fit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">all_parents</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">selected_variables</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut_off</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'max_lag_or_tau_max'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#Models.get_fit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.Models.get_fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit time series model.</p>
<p>For each variable in selected_variables, the sklearn model is fitted
with <img class="math" src="_images/math/1b5e577d6216dca3af7d87aa122a0b9b360d6cb3.png" alt="y"/> given by the target variable, and <img class="math" src="_images/math/ed38fa24f1c94891bd312012aab3f6673be3eb83.png" alt="X"/> given by its
parents. The fitted model class is returned for later use.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>all_parents</strong> (<em>dictionary</em>) – Dictionary of form {0:[(0, -1), (3, 0), …], 1:[], …} containing
the parents estimated with PCMCI.</p></li>
<li><p><strong>selected_variables</strong> (<em>list of integers</em><em>, </em><em>optional</em><em> (</em><em>default: range</em><em>(</em><em>N</em><em>)</em><em>)</em>) – Specify to estimate parents only for selected variables. If None is
passed, parents are estimated for all variables.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Maximum time lag. If None, the maximum lag in all_parents is used.</p></li>
<li><p><strong>cut_off</strong> (<em>{'max_lag_or_tau_max'</em><em>, </em><em>'2xtau_max'</em><em>, </em><em>'max_lag'}</em>) – How many samples to cutoff at the beginning. The default is
‘max_lag_or_tau_max’, which uses the maximum of tau_max and the
conditions. This is useful to compare multiple models on the same
sample. Other options are ‘2xtau_max’, which guarantees that MCI
tests are all conducted on the same samples. Last, ‘max_lag’ uses
as much samples as possible.</p></li>
<li><p><strong>return_data</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: False</em><em>)</em>) – Whether to save the data array.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>fit_results</strong> – Returns the sklearn model after fitting. Also returns the data
transformation parameters.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dictionary of sklearn model objects for each variable</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.Models.get_general_fitted_model">
<span class="sig-name descname"><span class="pre">get_general_fitted_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">Y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Z</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conditions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut_off</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'max_lag_or_tau_max'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#Models.get_general_fitted_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.Models.get_general_fitted_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit time series model.</p>
<p>For each variable in selected_variables, the sklearn model is fitted
with <img class="math" src="_images/math/1b5e577d6216dca3af7d87aa122a0b9b360d6cb3.png" alt="y"/> given by the target variable, and <img class="math" src="_images/math/ed38fa24f1c94891bd312012aab3f6673be3eb83.png" alt="X"/> given by its
parents. The fitted model class is returned for later use.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>lists of tuples</em>) – List of variables for estimating model Y = f(X,Z)</p></li>
<li><p><strong>Y</strong> (<em>lists of tuples</em>) – List of variables for estimating model Y = f(X,Z)</p></li>
<li><p><strong>Z</strong> (<em>lists of tuples</em>) – List of variables for estimating model Y = f(X,Z)</p></li>
<li><p><strong>conditions</strong> (<em>list of tuples.</em>) – Conditions for estimating conditional causal effects.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Maximum time lag. If None, the maximum lag in all_parents is used.</p></li>
<li><p><strong>cut_off</strong> (<em>{'max_lag_or_tau_max'</em><em>, </em><em>'2xtau_max'</em><em>, </em><em>'max_lag'}</em>) – How many samples to cutoff at the beginning. The default is
‘max_lag_or_tau_max’, which uses the maximum of tau_max and the
conditions. This is useful to compare multiple models on the same
sample. Other options are ‘2xtau_max’, which guarantees that MCI
tests are all conducted on the same samples. Last, ‘max_lag’ uses
as much samples as possible.</p></li>
<li><p><strong>return_data</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: False</em><em>)</em>) – Whether to save the data array.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>fit_results</strong> – Returns the sklearn model after fitting. Also returns the data
transformation parameters.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dictionary of sklearn model objects for each variable</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.Models.get_general_prediction">
<span class="sig-name descname"><span class="pre">get_general_prediction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervention_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conditions_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pred_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_further_pred_results</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#Models.get_general_prediction"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.Models.get_general_prediction" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict effect of intervention with fitted model.</p>
<p>Uses the model.predict() function of the sklearn model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervention_data</strong> (<em>numpy array</em>) – Numpy array of shape (time, len(X)) that contains the do(X) values.</p></li>
<li><p><strong>conditions_data</strong> (<em>data object</em><em>, </em><em>optional</em>) – Numpy array of shape (time, len(S)) that contains the S=s values.</p></li>
<li><p><strong>pred_params</strong> (<em>dict</em><em>, </em><em>optional</em>) – Optional parameters passed on to sklearn prediction function.</p></li>
<li><p><strong>return_further_pred_results</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: False</em><em>)</em>) – In case the predictor class returns more than just the expected value,
the entire results can be returned.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>Results from prediction.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.Models.get_val_matrix">
<span class="sig-name descname"><span class="pre">get_val_matrix</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#Models.get_val_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.Models.get_val_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the coefficient array for different lags for linear model.</p>
<p>Requires fit_model() before. An entry val_matrix[i,j,tau] gives the
coefficient of the link from i to j at lag tau, including tau=0.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>val_matrix</strong> – Array of coefficients for each time lag, including lag-zero.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>array-like, shape (N, N, tau_max + 1)</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<p>Derived classes:</p>
<dl class="py class">
<dt class="sig sig-object py" id="tigramite.models.LinearMediation">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tigramite.models.</span></span><span class="sig-name descname"><span class="pre">LinearMediation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataframe</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_transform</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">sklearn.preprocessing.StandardScaler</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#LinearMediation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.LinearMediation" title="Permalink to this definition">¶</a></dt>
<dd><p>Linear mediation analysis for time series models.</p>
<p>Fits linear model to parents and provides functions to return measures such
as causal effect, mediated causal effect, average causal effect, etc. as
described in <a class="footnote-reference brackets" href="#id29" id="id27" role="doc-noteref"><span class="fn-bracket">[</span>4<span class="fn-bracket">]</span></a>.</p>
<p>For general nonlinear, lagged and contemporaneous causal effect analysis,
use the CausalEffects class.</p>
<p class="rubric">Notes</p>
<p>This class implements the following causal mediation measures introduced in
<a class="footnote-reference brackets" href="#id29" id="id28" role="doc-noteref"><span class="fn-bracket">[</span>4<span class="fn-bracket">]</span></a>:</p>
<blockquote>
<div><ul class="simple">
<li><p>causal effect (CE)</p></li>
<li><p>mediated causal effect (MCE)</p></li>
<li><p>average causal effect (ACE)</p></li>
<li><p>average causal susceptibility (ACS)</p></li>
<li><p>average mediated causal effect (AMCE)</p></li>
</ul>
</div></blockquote>
<p>Consider a simple model of a causal chain as given in the Example with</p>
<div class="math">
<p><img src="_images/math/7b5b384bfd47bd6e8d707b3189aaab3a46c5ed04.png" alt="X_t &amp;= \eta^X_t \\
Y_t &amp;= 0.5 X_{t-1} +  \eta^Y_t \\
Z_t &amp;= 0.5 Y_{t-1} +  \eta^Z_t"/></p>
</div><p>Here the link coefficient of <img class="math" src="_images/math/64932505aadaa1eac6316ff09c2c3b101b068168.png" alt="X_{t-2} \to Z_t"/> is zero while the
causal effect is 0.25. MCE through <img class="math" src="_images/math/7daf0d4815e763eb90f0d5f1dc406f668c1e21db.png" alt="Y"/> is 0.25 implying that <em>all</em>
of the the CE is explained by <img class="math" src="_images/math/7daf0d4815e763eb90f0d5f1dc406f668c1e21db.png" alt="Y"/>. ACE from <img class="math" src="_images/math/ed38fa24f1c94891bd312012aab3f6673be3eb83.png" alt="X"/> is 0.37 since it
has CE 0.5 on <img class="math" src="_images/math/7daf0d4815e763eb90f0d5f1dc406f668c1e21db.png" alt="Y"/> and 0.25 on <img class="math" src="_images/math/95f028ab2b20b895fa12d986e0d9f40f7b6e52d3.png" alt="Z"/>.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">42</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">links_coeffs</span> <span class="o">=</span> <span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="p">[],</span> <span class="mi">1</span><span class="p">:</span> <span class="p">[((</span><span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">),</span> <span class="mf">0.5</span><span class="p">)],</span> <span class="mi">2</span><span class="p">:</span> <span class="p">[((</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">),</span> <span class="mf">0.5</span><span class="p">)]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="p">,</span> <span class="n">true_parents</span> <span class="o">=</span> <span class="n">toys</span><span class="o">.</span><span class="n">var_process</span><span class="p">(</span><span class="n">links_coeffs</span><span class="p">,</span> <span class="n">T</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dataframe</span> <span class="o">=</span> <span class="n">pp</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">med</span> <span class="o">=</span> <span class="n">LinearMediation</span><span class="p">(</span><span class="n">dataframe</span><span class="o">=</span><span class="n">dataframe</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">med</span><span class="o">.</span><span class="n">fit_model</span><span class="p">(</span><span class="n">all_parents</span><span class="o">=</span><span class="n">true_parents</span><span class="p">,</span> <span class="n">tau_max</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="s2">&quot;Link coefficient (0, -2) --&gt; 2: &quot;</span><span class="p">,</span> <span class="n">med</span><span class="o">.</span><span class="n">get_coeff</span><span class="p">(</span>
<span class="go">i=0, tau=-2, j=2)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="s2">&quot;Causal effect (0, -2) --&gt; 2: &quot;</span><span class="p">,</span> <span class="n">med</span><span class="o">.</span><span class="n">get_ce</span><span class="p">(</span><span class="n">i</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">tau</span><span class="o">=-</span><span class="mi">2</span><span class="p">,</span> <span class="n">j</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="s2">&quot;Mediated Causal effect (0, -2) --&gt; 2 through 1: &quot;</span><span class="p">,</span> <span class="n">med</span><span class="o">.</span><span class="n">get_mce</span><span class="p">(</span>
<span class="go">i=0, tau=-2, j=2, k=1)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="s2">&quot;Average Causal Effect: &quot;</span><span class="p">,</span> <span class="n">med</span><span class="o">.</span><span class="n">get_all_ace</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="s2">&quot;Average Causal Susceptibility: &quot;</span><span class="p">,</span> <span class="n">med</span><span class="o">.</span><span class="n">get_all_acs</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="s2">&quot;Average Mediated Causal Effect: &quot;</span><span class="p">,</span> <span class="n">med</span><span class="o">.</span><span class="n">get_all_amce</span><span class="p">()</span>
<span class="go">Link coefficient (0, -2) --&gt; 2:  0.0</span>
<span class="go">Causal effect (0, -2) --&gt; 2:  0.250648072987</span>
<span class="go">Mediated Causal effect (0, -2) --&gt; 2 through 1:  0.250648072987</span>
<span class="go">Average Causal Effect:  [ 0.36897445  0.25718002  0.        ]</span>
<span class="go">Average Causal Susceptibility:  [ 0.          0.24365041  0.38250406]</span>
<span class="go">Average Mediated Causal Effect:  [ 0.          0.12532404  0.        ]</span>
</pre></div>
</div>
<p class="rubric">References</p>
<aside class="footnote brackets" id="id29" role="note">
<span class="label"><span class="fn-bracket">[</span>4<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id27">1</a>,<a role="doc-backlink" href="#id28">2</a>,<a role="doc-backlink" href="#id30">3</a>)</span>
<p>J. Runge et al. (2015): Identifying causal gateways and mediators in
complex spatio-temporal systems.
Nature Communications, 6, 8502. <a class="reference external" href="http://doi.org/10.1038/ncomms9502">http://doi.org/10.1038/ncomms9502</a></p>
</aside>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dataframe</strong> (<em>data object</em>) – Tigramite dataframe object. It must have the attributes dataframe.values
yielding a numpy array of shape (observations T, variables N) and
optionally a mask of the same shape and a missing values flag.</p></li>
<li><p><strong>model_params</strong> (<em>dictionary</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Optional parameters passed on to sklearn model</p></li>
<li><p><strong>data_transform</strong> (<em>sklearn preprocessing object</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Used to transform data prior to fitting. For example,
sklearn.preprocessing.StandardScaler for simple standardization. The
fitted parameters are stored.</p></li>
<li><p><strong>mask_type</strong> (<em>{None</em><em>, </em><em>'y'</em><em>,</em><em>'x'</em><em>,</em><em>'z'</em><em>,</em><em>'xy'</em><em>,</em><em>'xz'</em><em>,</em><em>'yz'</em><em>,</em><em>'xyz'}</em>) – Masking mode: Indicators for which variables in the dependence
measure I(X; Y | Z) the samples should be masked. If None, the mask
is not used. Explained in tutorial on masking and missing values.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Level of verbosity.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.LinearMediation.fit_model">
<span class="sig-name descname"><span class="pre">fit_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">all_parents</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#LinearMediation.fit_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.LinearMediation.fit_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit linear time series model.</p>
<p>Fits a sklearn.linear_model.LinearRegression model to the parents of
each variable and computes the coefficient matrices <img class="math" src="_images/math/d8ee85ac4d75924cdf4b18f5fb3b46550932fc26.png" alt="\Phi"/> and
<img class="math" src="_images/math/20582dab63cb7f6604f5bf70224030ad3411ae16.png" alt="\Psi"/> as described in <a class="footnote-reference brackets" href="#id29" id="id30" role="doc-noteref"><span class="fn-bracket">[</span>4<span class="fn-bracket">]</span></a>. Does not accepted
contemporaneous links.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>all_parents</strong> (<em>dictionary</em>) – Dictionary of form {0:[(0, -1), (3, -2), …], 1:[], …} containing
the parents estimated with PCMCI.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Maximum time lag. If None, the maximum lag in all_parents is used.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.LinearMediation.get_ace">
<span class="sig-name descname"><span class="pre">get_ace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lag_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'absmax'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_i</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#LinearMediation.get_ace"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.LinearMediation.get_ace" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the average causal effect.</p>
<p>This is the average causal effect (ACE) emanating from variable i to any
other variable. With lag_mode=’absmax’ this is based on the lag of
maximum CE for each pair.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>i</strong> (<em>int</em>) – Index of cause variable.</p></li>
<li><p><strong>lag_mode</strong> (<em>{'absmax'</em><em>, </em><em>'all_lags'}</em>) – Lag mode. Either average across all lags between each pair or only
at the lag of maximum absolute causal effect.</p></li>
<li><p><strong>exclude_i</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: True</em><em>)</em>) – Whether to exclude causal effects on the variable itself at later
lags.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ace</strong> – Average Causal Effect.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.LinearMediation.get_acs">
<span class="sig-name descname"><span class="pre">get_acs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">j</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lag_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'absmax'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_j</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#LinearMediation.get_acs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.LinearMediation.get_acs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the average causal susceptibility.</p>
<p>This is the Average Causal Susceptibility (ACS) affecting a variable j
from any other variable. With lag_mode=’absmax’ this is based on the lag
of maximum CE for each pair.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>j</strong> (<em>int</em>) – Index of variable.</p></li>
<li><p><strong>lag_mode</strong> (<em>{'absmax'</em><em>, </em><em>'all_lags'}</em>) – Lag mode. Either average across all lags between each pair or only
at the lag of maximum absolute causal effect.</p></li>
<li><p><strong>exclude_j</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: True</em><em>)</em>) – Whether to exclude causal effects on the variable itself at previous
lags.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>acs</strong> – Average Causal Susceptibility.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.LinearMediation.get_all_ace">
<span class="sig-name descname"><span class="pre">get_all_ace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lag_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'absmax'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_i</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#LinearMediation.get_all_ace"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.LinearMediation.get_all_ace" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the average causal effect for all variables.</p>
<p>This is the average causal effect (ACE) emanating from variable i to any
other variable. With lag_mode=’absmax’ this is based on the lag of
maximum CE for each pair.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>lag_mode</strong> (<em>{'absmax'</em><em>, </em><em>'all_lags'}</em>) – Lag mode. Either average across all lags between each pair or only
at the lag of maximum absolute causal effect.</p></li>
<li><p><strong>exclude_i</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: True</em><em>)</em>) – Whether to exclude causal effects on the variable itself at later
lags.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ace</strong> – Average Causal Effect for each variable.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>array of shape (N,)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.LinearMediation.get_all_acs">
<span class="sig-name descname"><span class="pre">get_all_acs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lag_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'absmax'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_j</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#LinearMediation.get_all_acs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.LinearMediation.get_all_acs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the average causal susceptibility.</p>
<p>This is the Average Causal Susceptibility (ACS) for each variable from
any other variable. With lag_mode=’absmax’ this is based on the lag of
maximum CE for each pair.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>lag_mode</strong> (<em>{'absmax'</em><em>, </em><em>'all_lags'}</em>) – Lag mode. Either average across all lags between each pair or only
at the lag of maximum absolute causal effect.</p></li>
<li><p><strong>exclude_j</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: True</em><em>)</em>) – Whether to exclude causal effects on the variable itself at previous
lags.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>acs</strong> – Average Causal Susceptibility.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>array of shape (N,)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.LinearMediation.get_all_amce">
<span class="sig-name descname"><span class="pre">get_all_amce</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lag_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'absmax'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_self_effects</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#LinearMediation.get_all_amce"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.LinearMediation.get_all_amce" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the average mediated causal effect.</p>
<p>This is the Average Mediated Causal Effect (AMCE) through all variables
With lag_mode=’absmax’ this is based on the lag of maximum CE for each
pair.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>lag_mode</strong> (<em>{'absmax'</em><em>, </em><em>'all_lags'}</em>) – Lag mode. Either average across all lags between each pair or only
at the lag of maximum absolute causal effect.</p></li>
<li><p><strong>exclude_k</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: True</em><em>)</em>) – Whether to exclude causal effects through the variable itself at
previous lags.</p></li>
<li><p><strong>exclude_self_effects</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: True</em><em>)</em>) – Whether to exclude causal self effects of variables on themselves.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>amce</strong> – Average Mediated Causal Effect.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>array of shape (N,)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.LinearMediation.get_amce">
<span class="sig-name descname"><span class="pre">get_amce</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">k</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lag_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'absmax'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_self_effects</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#LinearMediation.get_amce"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.LinearMediation.get_amce" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the average mediated causal effect.</p>
<p>This is the Average Mediated Causal Effect (AMCE) through a variable k
With lag_mode=’absmax’ this is based on the lag of maximum CE for each
pair.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>k</strong> (<em>int</em>) – Index of variable.</p></li>
<li><p><strong>lag_mode</strong> (<em>{'absmax'</em><em>, </em><em>'all_lags'}</em>) – Lag mode. Either average across all lags between each pair or only
at the lag of maximum absolute causal effect.</p></li>
<li><p><strong>exclude_k</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: True</em><em>)</em>) – Whether to exclude causal effects through the variable itself at
previous lags.</p></li>
<li><p><strong>exclude_self_effects</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: True</em><em>)</em>) – Whether to exclude causal self effects of variables on themselves.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>amce</strong> – Average Mediated Causal Effect.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.LinearMediation.get_ce">
<span class="sig-name descname"><span class="pre">get_ce</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#LinearMediation.get_ce"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.LinearMediation.get_ce" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the causal effect.</p>
<p>This is the causal effect for  (i, tau) – –&gt; j.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>i</strong> (<em>int</em>) – Index of cause variable.</p></li>
<li><p><strong>tau</strong> (<em>int</em>) – Lag of cause variable.</p></li>
<li><p><strong>j</strong> (<em>int</em>) – Index of effect variable.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ce</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.LinearMediation.get_ce_max">
<span class="sig-name descname"><span class="pre">get_ce_max</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#LinearMediation.get_ce_max"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.LinearMediation.get_ce_max" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the causal effect.</p>
<p>This is the maximum absolute causal effect for  i –&gt; j across all lags.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>i</strong> (<em>int</em>) – Index of cause variable.</p></li>
<li><p><strong>j</strong> (<em>int</em>) – Index of effect variable.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ce</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.LinearMediation.get_coeff">
<span class="sig-name descname"><span class="pre">get_coeff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#LinearMediation.get_coeff"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.LinearMediation.get_coeff" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns link coefficient.</p>
<p>This is the direct causal effect for a particular link (i, tau) –&gt; j.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>i</strong> (<em>int</em>) – Index of cause variable.</p></li>
<li><p><strong>tau</strong> (<em>int</em>) – Lag of cause variable.</p></li>
<li><p><strong>j</strong> (<em>int</em>) – Index of effect variable.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>coeff</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.LinearMediation.get_mce">
<span class="sig-name descname"><span class="pre">get_mce</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#LinearMediation.get_mce"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.LinearMediation.get_mce" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the mediated causal effect.</p>
<p>This is the causal effect for  i –&gt; j minus the causal effect not going
through k.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>i</strong> (<em>int</em>) – Index of cause variable.</p></li>
<li><p><strong>tau</strong> (<em>int</em>) – Lag of cause variable.</p></li>
<li><p><strong>j</strong> (<em>int</em>) – Index of effect variable.</p></li>
<li><p><strong>k</strong> (<em>int</em>) – Index of mediator variable.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>mce</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.LinearMediation.get_mediation_graph_data">
<span class="sig-name descname"><span class="pre">get_mediation_graph_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_neighbors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#LinearMediation.get_mediation_graph_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.LinearMediation.get_mediation_graph_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns link and node weights for mediation analysis.</p>
<p>Returns array with non-zero entries for links that are on causal
paths between <img class="math" src="_images/math/5aa339d4daf45a810dda332e3c80a0698e526e04.png" alt="i"/> and <img class="math" src="_images/math/e3fc28292267f066fee7718c64f4bbfece521f24.png" alt="j"/> at lag <img class="math" src="_images/math/914b2d4b6659b86d3153d5510839dfb254dfc8a3.png" alt="\tau"/>.
<code class="docutils literal notranslate"><span class="pre">path_val_matrix</span></code> contains the corresponding path coefficients and
<code class="docutils literal notranslate"><span class="pre">path_node_array</span></code> the MCE values. <code class="docutils literal notranslate"><span class="pre">tsg_path_val_matrix</span></code> contains the
corresponding values in the time series graph format.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>i</strong> (<em>int</em>) – Index of cause variable.</p></li>
<li><p><strong>tau</strong> (<em>int</em>) – Lag of cause variable.</p></li>
<li><p><strong>j</strong> (<em>int</em>) – Index of effect variable.</p></li>
<li><p><strong>include_neighbors</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: False</em><em>)</em>) – Whether to include causal paths emanating from neighbors of i</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>graph_data</strong> – Dictionary of matrices for coloring mediation graph plots.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dictionary</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.LinearMediation.get_tsg">
<span class="sig-name descname"><span class="pre">get_tsg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">link_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val_matrix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_neighbors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#LinearMediation.get_tsg"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.LinearMediation.get_tsg" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns time series graph matrix.</p>
<p>Constructs a matrix of shape (N*tau_max, N*tau_max) from link_matrix.
This matrix can be used for plotting the time series graph and analyzing
causal pathways.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>link_matrix</strong> (<em>bool array-like</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Matrix of significant links. Must be of same shape as val_matrix.
Either sig_thres or link_matrix has to be provided.</p></li>
<li><p><strong>val_matrix</strong> (<em>array_like</em>) – Matrix of shape (N, N, tau_max+1) containing test statistic values.</p></li>
<li><p><strong>include_neighbors</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: False</em><em>)</em>) – Whether to include causal paths emanating from neighbors of i</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>tsg</strong> – Time series graph matrix.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>array of shape (N*tau_max, N*tau_max)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.LinearMediation.get_val_matrix">
<span class="sig-name descname"><span class="pre">get_val_matrix</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#LinearMediation.get_val_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.LinearMediation.get_val_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the matrix of linear coefficients.</p>
<p>Requires fit_model() before. An entry val_matrix[i,j,tau] gives the
coefficient of the link from i to j at lag tau. Lag=0 is always set
to zero for LinearMediation, use Models class for contemporaneous
models.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>val_matrix</strong> – Matrix of linear coefficients, shape (N, N, tau_max + 1).</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>array</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.LinearMediation.net_to_tsg">
<span class="sig-name descname"><span class="pre">net_to_tsg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lag</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_lag</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#LinearMediation.net_to_tsg"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.LinearMediation.net_to_tsg" title="Permalink to this definition">¶</a></dt>
<dd><p>Helper function to translate from network to time series graph.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.LinearMediation.tsg_to_net">
<span class="sig-name descname"><span class="pre">tsg_to_net</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">node</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_lag</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#LinearMediation.tsg_to_net"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.LinearMediation.tsg_to_net" title="Permalink to this definition">¶</a></dt>
<dd><p>Helper function to translate from time series graph to network.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="tigramite.models.Prediction">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tigramite.models.</span></span><span class="sig-name descname"><span class="pre">Prediction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataframe</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">train_indices</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_indices</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prediction_model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cond_ind_test</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_transform</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#Prediction"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.Prediction" title="Permalink to this definition">¶</a></dt>
<dd><p>Prediction class for time series models.</p>
<p>Allows to fit and predict from any sklearn model. The optimal predictors can
be estimated using PCMCI. Also takes care of missing values, masking and
preprocessing.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dataframe</strong> (<em>data object</em>) – Tigramite dataframe object. It must have the attributes dataframe.values
yielding a numpy array of shape (observations T, variables N) and
optionally a mask of the same shape and a missing values flag.</p></li>
<li><p><strong>train_indices</strong> (<em>array-like</em>) – Either boolean array or time indices marking the training data.</p></li>
<li><p><strong>test_indices</strong> (<em>array-like</em>) – Either boolean array or time indices marking the test data.</p></li>
<li><p><strong>prediction_model</strong> (<em>sklearn model object</em>) – For example, sklearn.linear_model.LinearRegression() for a linear
regression model.</p></li>
<li><p><strong>cond_ind_test</strong> (<em>Conditional independence test object</em><em>, </em><em>optional</em>) – Only needed if predictors are estimated with causal algorithm.
The class will be initialized with masking set to the training data.</p></li>
<li><p><strong>data_transform</strong> (<em>sklearn preprocessing object</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Used to transform data prior to fitting. For example,
sklearn.preprocessing.StandardScaler for simple standardization. The
fitted parameters are stored.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Level of verbosity.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.Prediction.fit">
<span class="sig-name descname"><span class="pre">fit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target_predictors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">selected_targets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#Prediction.fit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.Prediction.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit time series model.</p>
<p>Wrapper around <code class="docutils literal notranslate"><span class="pre">Models.get_fit()</span></code>. To each variable in
<code class="docutils literal notranslate"><span class="pre">selected_targets</span></code>, the sklearn model is fitted with <img class="math" src="_images/math/1b5e577d6216dca3af7d87aa122a0b9b360d6cb3.png" alt="y"/> given
by the target variable, and <img class="math" src="_images/math/ed38fa24f1c94891bd312012aab3f6673be3eb83.png" alt="X"/> given by its predictors. The
fitted model class is returned for later use.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>target_predictors</strong> (<em>dictionary</em>) – Dictionary of form {0:[(0, -1), (3, -2), …], 1:[], …} containing
the predictors estimated with PCMCI.</p></li>
<li><p><strong>selected_targets</strong> (<em>list of integers</em><em>, </em><em>optional</em><em> (</em><em>default: range</em><em>(</em><em>N</em><em>)</em><em>)</em>) – Specify to fit model only for selected targets. If None is
passed, models are estimated for all variables.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Maximum time lag. If None, the maximum lag in target_predictors is
used.</p></li>
<li><p><strong>return_data</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: False</em><em>)</em>) – Whether to save the data array.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>self</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>instance of self</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.Prediction.get_predictors">
<span class="sig-name descname"><span class="pre">get_predictors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">selected_targets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">selected_links</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">steps_ahead</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pc_alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_conds_dim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_combinations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#Prediction.get_predictors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.Prediction.get_predictors" title="Permalink to this definition">¶</a></dt>
<dd><p>Estimate predictors using PC1 algorithm.</p>
<p>Wrapper around PCMCI.run_pc_stable that estimates causal predictors.
The lead time can be specified by <code class="docutils literal notranslate"><span class="pre">steps_ahead</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>selected_targets</strong> (<em>list of ints</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – List of variables to estimate predictors of. If None, predictors of
all variables are estimated.</p></li>
<li><p><strong>selected_links</strong> (<em>dict</em><em> or </em><em>None</em>) – Dictionary of form {0:[(0, -1), (3, -2), …], 1:[], …}
specifying whether only selected links should be tested. If None is
passed, all links are tested</p></li>
<li><p><strong>steps_ahead</strong> (<em>int</em><em>, </em><em>default: 1</em>) – Minimum time lag to test. Useful for multi-step ahead predictions.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em><em>, </em><em>default: 1</em>) – Maximum time lag. Must be larger or equal to tau_min.</p></li>
<li><p><strong>pc_alpha</strong> (<em>float</em><em> or </em><em>list of floats</em><em>, </em><em>default: 0.2</em>) – Significance level in algorithm. If a list or None is passed, the
pc_alpha level is optimized for every variable across the given
pc_alpha values using the score computed in
cond_ind_test.get_model_selection_criterion()</p></li>
<li><p><strong>max_conds_dim</strong> (<em>int</em><em> or </em><em>None</em>) – Maximum number of conditions to test. If None is passed, this number
is unrestricted.</p></li>
<li><p><strong>max_combinations</strong> (<em>int</em><em>, </em><em>default: 1</em>) – Maximum number of combinations of conditions of current cardinality
to test. Defaults to 1 for PC_1 algorithm. For original PC algorithm
a larger number, such as 10, can be used.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>predictors</strong> – Dictionary of form {0:[(0, -1), (3, -2), …], 1:[], …}
containing estimated predictors.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.Prediction.get_test_array">
<span class="sig-name descname"><span class="pre">get_test_array</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#Prediction.get_test_array"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.Prediction.get_test_array" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns test array.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.Prediction.get_train_array">
<span class="sig-name descname"><span class="pre">get_train_array</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">j</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#Prediction.get_train_array"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.Prediction.get_train_array" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns training array.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.models.Prediction.predict">
<span class="sig-name descname"><span class="pre">predict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pred_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut_off</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'max_lag_or_tau_max'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/models.html#Prediction.predict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.models.Prediction.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict target variable with fitted model.</p>
<p>Uses the model.predict() function of the sklearn model.</p>
<p>If target is an int, the predicted time series is returned. If target
is a list of integers, then a list of predicted time series is returned.
If the list of integers equals range(N), then an array of shape (T, N)
of the predicted series is returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>target</strong> (<em>int</em><em> or </em><em>list of integers</em>) – Index or indices of target variable(s).</p></li>
<li><p><strong>new_data</strong> (<em>data object</em><em>, </em><em>optional</em>) – New Tigramite dataframe object with optional new mask.</p></li>
<li><p><strong>pred_params</strong> (<em>dict</em><em>, </em><em>optional</em>) – Optional parameters passed on to sklearn prediction function.</p></li>
<li><p><strong>cut_off</strong> (<em>{'2xtau_max'</em><em>, </em><em>'max_lag'</em><em>, </em><em>'max_lag_or_tau_max'}</em>) – How many samples to cutoff at the beginning. The default is
‘2xtau_max’, which guarantees that MCI tests are all conducted on
the same samples.  For modeling, ‘max_lag_or_tau_max’ can be used,
which uses the maximum of tau_max and the conditions, which is
useful to compare multiple models on the same sample. Last,
‘max_lag’ uses as much samples as possible.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>Results from prediction.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-tigramite.data_processing">
<span id="tigramite-data-processing-data-processing-functions"></span><h1><a class="reference internal" href="#module-tigramite.data_processing" title="tigramite.data_processing"><code class="xref py py-mod docutils literal notranslate"><span class="pre">tigramite.data_processing</span></code></a>: Data processing functions<a class="headerlink" href="#module-tigramite.data_processing" title="Permalink to this heading">¶</a></h1>
<p>Tigramite data processing functions.</p>
<dl class="py class">
<dt class="sig sig-object py" id="tigramite.data_processing.DataFrame">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tigramite.data_processing.</span></span><span class="sig-name descname"><span class="pre">DataFrame</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">missing_flag</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vector_vars</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datatime</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">analysis_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'single'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_points</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_offsets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_missing_upto_maxlag</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/data_processing.html#DataFrame"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.data_processing.DataFrame" title="Permalink to this definition">¶</a></dt>
<dd><p>Data object containing single or multiple time series arrays and optional
mask.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>array-like</em>) – <dl class="simple">
<dt>if analysis_mode == ‘single’:</dt><dd><p>1) Numpy array of shape (observations T, variables N)
OR
2) Dictionary with a single entry whose value is a numpy array of
shape (observations T, variables N)</p>
</dd>
<dt>if analysis_mode == ‘multiple’:</dt><dd><p>1) Numpy array of shape (multiple datasets M, observations T,
variables N)
OR
2) Dictionary whose values are numpy arrays of shape
(observations T_i, variables N), where the number of observations
T_i may vary across the multiple datasets but the number of variables
N is fixed.</p>
</dd>
</dl>
</p></li>
<li><p><strong>mask</strong> (<em>array-like</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Optional mask array, must be of same format and shape as data.</p></li>
<li><p><strong>missing_flag</strong> (<em>number</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Flag for missing values in dataframe. Dismisses all time slices of
samples where missing values occur in any variable. For
remove_missing_upto_maxlag=True also flags samples for all lags up
to 2*tau_max (more precisely, this depends on
the cut_off argument in self.construct_array(), see further below).
This avoids biases, see section on masking in Supplement of <a class="footnote-reference brackets" href="#id3" id="id31" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a>.</p></li>
<li><p><strong>var_names</strong> (<em>list of strings</em><em>, </em><em>optional</em><em> (</em><em>default: range</em><em>(</em><em>N</em><em>)</em><em>)</em>) – Names of variables, must match the number of variables. If None is
passed, variables are enumerated as [0, 1, …]</p></li>
<li><p><strong>datatime</strong> (<em>array-like</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Timelabel array. If None, range(T) is used.</p></li>
<li><p><strong>remove_missing_upto_maxlag</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: False</em><em>)</em>) – Whether to remove not only missing samples, but also all neighboring
samples up to max_lag (as given by cut_off in construct_array).</p></li>
<li><p><strong>analysis_mode</strong> (<em>string</em><em>, </em><em>optional</em><em> (</em><em>default: 'single'</em><em>)</em>) – Must be ‘single’ or ‘multiple’.
Determines whether data contains a single (potentially multivariate)
time series (–&gt; ‘single’) or multiple time series (–&gt; ‘multiple’).</p></li>
<li><p><strong>reference_points</strong> (<em>None</em><em>, </em><em>int</em><em>, or </em><em>list</em><em> (or </em><em>1D array</em><em>) </em><em>of integers</em><em>,</em>) – optional (default:None)
Determines the time steps — relative to the shared time axis as
defined by the optional time_offset argument (see below) — that are
used to create samples for conditional independence testing.
Set to [0, 1, …, T_max-1] if None is passed, where T_max is
self.largest_time_step, see below.
All values smaller than 0 and bigger than T_max-1 will be ignored.
At least one value must be in [0, 1, …, T_max-1].</p></li>
<li><p><strong>time_offsets</strong> (<em>None</em><em> or </em><em>dict</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – <dl class="simple">
<dt>if analysis_mode == ‘single’:</dt><dd><p>Must be None.
Shared time axis defined by the time indices of the single time series</p>
</dd>
<dt>if analysis_mode == ‘multiple’ and data is numpy array:</dt><dd><p>Must be None.
All datasets are assumed to be already aligned in time with
respect to a shared time axis, which is the time axis of data</p>
</dd>
<dt>if analysis_mode == ‘multiple’ and data is dictionary:</dt><dd><p>Must be dictionary of the form {key(m): time_offset(m), …} whose
set of keys agrees with the set of keys of data and whose values are
non-negative integers, at least of one which is 0. The value
time_offset(m) defines the time offset of dataset m with
respect to a shared time axis.</p>
</dd>
</dl>
</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="tigramite.data_processing.DataFrame.self._initialized_from">
<span class="sig-prename descclassname"><span class="pre">self.</span></span><span class="sig-name descname"><span class="pre">_initialized_from</span></span><a class="headerlink" href="#tigramite.data_processing.DataFrame.self._initialized_from" title="Permalink to this definition">¶</a></dt>
<dd><p>Specifies the data format in which data was given at instantiation.
Possible values: ‘2d numpy array’, ‘3d numpy array’, ‘dict’.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>string</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tigramite.data_processing.DataFrame.self.values">
<span class="sig-prename descclassname"><span class="pre">self.</span></span><span class="sig-name descname"><span class="pre">values</span></span><a class="headerlink" href="#tigramite.data_processing.DataFrame.self.values" title="Permalink to this definition">¶</a></dt>
<dd><p>Dictionary holding the observations given by data internally mapped to a
dictionary representation as follows:
If analysis_mode == ‘single’:</p>
<blockquote>
<div><dl class="simple">
<dt>If self._initialized_from == ‘2d numpy array’:</dt><dd><p>Is {0: data}</p>
</dd>
<dt>If self._initialized_from == ‘dict’:</dt><dd><p>Is data</p>
</dd>
</dl>
</div></blockquote>
<dl class="simple">
<dt>If analysis_mode == ‘multiple’:</dt><dd><dl class="simple">
<dt>If self._initialized_from == ‘3d numpy array’:</dt><dd><p>Is {m: data[m, :, :] for m in range(data.shape[0])}</p>
</dd>
<dt>If self._initialized_from == ‘dict’:</dt><dd><p>Is data</p>
</dd>
</dl>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dictionary</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tigramite.data_processing.DataFrame.self.datasets">
<span class="sig-prename descclassname"><span class="pre">self.</span></span><span class="sig-name descname"><span class="pre">datasets</span></span><a class="headerlink" href="#tigramite.data_processing.DataFrame.self.datasets" title="Permalink to this definition">¶</a></dt>
<dd><p>List of the keys identifiying the multiple datasets, i.e.,
list(self.values.keys())</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tigramite.data_processing.DataFrame.self.mask">
<span class="sig-prename descclassname"><span class="pre">self.</span></span><span class="sig-name descname"><span class="pre">mask</span></span><a class="headerlink" href="#tigramite.data_processing.DataFrame.self.mask" title="Permalink to this definition">¶</a></dt>
<dd><p>Mask internally mapped to a dictionary representation in the same way as
data is mapped to self.values</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dictionary</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tigramite.data_processing.DataFrame.self.missing_flag">
<span class="sig-prename descclassname"><span class="pre">self.</span></span><span class="sig-name descname"><span class="pre">missing_flag</span></span><a class="headerlink" href="#tigramite.data_processing.DataFrame.self.missing_flag" title="Permalink to this definition">¶</a></dt>
<dd><p>Is missing_flag</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tigramite.data_processing.DataFrame.self.var_names">
<span class="sig-prename descclassname"><span class="pre">self.</span></span><span class="sig-name descname"><span class="pre">var_names</span></span><a class="headerlink" href="#tigramite.data_processing.DataFrame.self.var_names" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>If var_names is not None:</dt><dd><p>Is var_names</p>
</dd>
<dt>If var_names is None:</dt><dd><p>Is {i: i for i in range(self.N)}</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tigramite.data_processing.DataFrame.self.datatime">
<span class="sig-prename descclassname"><span class="pre">self.</span></span><span class="sig-name descname"><span class="pre">datatime</span></span><a class="headerlink" href="#tigramite.data_processing.DataFrame.self.datatime" title="Permalink to this definition">¶</a></dt>
<dd><p>Time axis for each of the multiple datasets.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dictionary</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tigramite.data_processing.DataFrame.self.analysis_mode">
<span class="sig-prename descclassname"><span class="pre">self.</span></span><span class="sig-name descname"><span class="pre">analysis_mode</span></span><a class="headerlink" href="#tigramite.data_processing.DataFrame.self.analysis_mode" title="Permalink to this definition">¶</a></dt>
<dd><p>Is analysis_mode</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>string</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tigramite.data_processing.DataFrame.self.reference_points">
<span class="sig-prename descclassname"><span class="pre">self.</span></span><span class="sig-name descname"><span class="pre">reference_points</span></span><a class="headerlink" href="#tigramite.data_processing.DataFrame.self.reference_points" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>If reference_points is not None:</dt><dd><p>1D numpy array holding all specified reference_points, less those
smaller than 0 and larger than self.largest_time_step-1</p>
</dd>
<dt>If reference_points is None:</dt><dd><p>Is np.array(range(self.largest_time_step))</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>array-like</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tigramite.data_processing.DataFrame.self.time_offsets">
<span class="sig-prename descclassname"><span class="pre">self.</span></span><span class="sig-name descname"><span class="pre">time_offsets</span></span><a class="headerlink" href="#tigramite.data_processing.DataFrame.self.time_offsets" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>If time_offsets is not None:</dt><dd><p>Is time_offsets</p>
</dd>
<dt>If time_offsets is None:</dt><dd><p>Is {key: 0 for key in self.values.keys()}</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dictionary</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tigramite.data_processing.DataFrame.self.M">
<span class="sig-prename descclassname"><span class="pre">self.</span></span><span class="sig-name descname"><span class="pre">M</span></span><a class="headerlink" href="#tigramite.data_processing.DataFrame.self.M" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of datasets</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tigramite.data_processing.DataFrame.self.N">
<span class="sig-prename descclassname"><span class="pre">self.</span></span><span class="sig-name descname"><span class="pre">N</span></span><a class="headerlink" href="#tigramite.data_processing.DataFrame.self.N" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of variables (constant across datasets)</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tigramite.data_processing.DataFrame.self.T">
<span class="sig-prename descclassname"><span class="pre">self.</span></span><span class="sig-name descname"><span class="pre">T</span></span><a class="headerlink" href="#tigramite.data_processing.DataFrame.self.T" title="Permalink to this definition">¶</a></dt>
<dd><p>Dictionary {key(m): T(m), …}, where T(m) is the time length of
datasets m and key(m) its identifier as in self.values</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dictionary</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tigramite.data_processing.DataFrame.self.largest_time_step">
<span class="sig-prename descclassname"><span class="pre">self.</span></span><span class="sig-name descname"><span class="pre">largest_time_step</span></span><a class="headerlink" href="#tigramite.data_processing.DataFrame.self.largest_time_step" title="Permalink to this definition">¶</a></dt>
<dd><p>max_{0 &lt;= m &lt;= M} [ T(m) + time_offset(m)], i.e., the largest (latest)
time step relative to the shared time axis for which at least one
observation exists in the dataset.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tigramite.data_processing.DataFrame.self.bootstrap">
<span class="sig-prename descclassname"><span class="pre">self.</span></span><span class="sig-name descname"><span class="pre">bootstrap</span></span><a class="headerlink" href="#tigramite.data_processing.DataFrame.self.bootstrap" title="Permalink to this definition">¶</a></dt>
<dd><p>Whether to use bootstrap. Must be a dictionary with keys random_state,
boot_samples, and boot_blocklength.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dictionary</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.data_processing.DataFrame.construct_array">
<span class="sig-name descname"><span class="pre">construct_array</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Z</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extraZ</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_cleaned_xyz</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">do_checks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_overlaps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut_off</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'2xtau_max'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/data_processing.html#DataFrame.construct_array"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.data_processing.DataFrame.construct_array" title="Permalink to this definition">¶</a></dt>
<dd><p>Constructs array from variables X, Y, Z from data.
Data is of shape (T, N) if analysis_mode == ‘single’, where T is the
time series length and N the number of variables, and of (n_ens, T, N)
if analysis_mode == ‘multiple’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>list of tuples</em>) – For a dependence measure I(X;Y|Z), X, Y, Z can be multivariate of
the form [(var1, -lag), (var2, -lag), …]. At least one varlag in Y
has to be at lag zero. extraZ is only used in CausalEffects class.</p></li>
<li><p><strong>Y</strong> (<em>list of tuples</em>) – For a dependence measure I(X;Y|Z), X, Y, Z can be multivariate of
the form [(var1, -lag), (var2, -lag), …]. At least one varlag in Y
has to be at lag zero. extraZ is only used in CausalEffects class.</p></li>
<li><p><strong>Z</strong> (<em>list of tuples</em>) – For a dependence measure I(X;Y|Z), X, Y, Z can be multivariate of
the form [(var1, -lag), (var2, -lag), …]. At least one varlag in Y
has to be at lag zero. extraZ is only used in CausalEffects class.</p></li>
<li><p><strong>extraZ</strong> (<em>list of tuples</em>) – For a dependence measure I(X;Y|Z), X, Y, Z can be multivariate of
the form [(var1, -lag), (var2, -lag), …]. At least one varlag in Y
has to be at lag zero. extraZ is only used in CausalEffects class.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em>) – Maximum time lag. This may be used to make sure that estimates for
different lags in X and Z all have the same sample size.</p></li>
<li><p><strong>mask</strong> (<em>array-like</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Optional mask array, must be of same shape as data.  If it is set,
then it overrides the self.mask assigned to the dataframe. If it is
None, then the self.mask is used, if it exists.</p></li>
<li><p><strong>mask_type</strong> (<em>{None</em><em>, </em><em>'y'</em><em>,</em><em>'x'</em><em>,</em><em>'z'</em><em>,</em><em>'xy'</em><em>,</em><em>'xz'</em><em>,</em><em>'yz'</em><em>,</em><em>'xyz'}</em>) – Masking mode: Indicators for which variables in the dependence
measure I(X; Y | Z) the samples should be masked. If None, the mask
is not used. Explained in tutorial on masking and missing values.</p></li>
<li><p><strong>return_cleaned_xyz</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: False</em><em>)</em>) – Whether to return cleaned X,Y,Z, where possible duplicates are
removed.</p></li>
<li><p><strong>do_checks</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: True</em><em>)</em>) – Whether to perform sanity checks on input X,Y,Z</p></li>
<li><p><strong>remove_overlaps</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: True</em><em>)</em>) – Whether to remove variables from Z/extraZ if they overlap with X or Y.</p></li>
<li><p><strong>cut_off</strong> (<em>{'2xtau_max'</em><em>, </em><em>'tau_max'</em><em>, </em><em>'max_lag'</em><em>, </em><em>'max_lag_or_tau_max'</em><em>,</em>) – <p>2xtau_max_future}
If cut_off == ‘2xtau_max’:</p>
<blockquote>
<div><ul>
<li><p>2*tau_max samples are cut off at the beginning of the time</p></li>
</ul>
<p>series (‘beginning’ here refers to the temporally first time
steps). This guarantees that (as long as no mask is used) all
MCI tests are conducted on the same samples, independent of X,
Y, and Z.
- If at time step t_missing a data value is missing, then the
time steps t_missing, …, t_missing + 2*tau_max are cut out.
The latter part only holds if remove_missing_upto_maxlag=True.</p>
</div></blockquote>
<dl>
<dt>If cut_off ==  ‘max_lag’:</dt><dd><ul>
<li><p>max_lag(X, Y, Z) samples are cut off at the beginning of the</p></li>
</ul>
<p>time series, where max_lag(X, Y, Z) is the maximum lag of all
nodes in X, Y, and Z. These are all samples that can in
principle be used.
- If at time step t_missing a data value is missing, then the
time steps t_missing, …, t_missing + max_lag(X, Y, Z) are cut
out.
The latter part only holds if remove_missing_upto_maxlag=True.</p>
</dd>
<dt>If cut_off == ‘max_lag_or_tau_max’:</dt><dd><ul>
<li><p>max(max_lag(X, Y, Z), tau_max) are cut off at the beginning.</p></li>
</ul>
<p>This may be useful for modeling by comparing multiple models on
the same samples.
- If at time step t_missing a data value is missing, then the
time steps
t_missing, …, t_missing + max(max_lag(X, Y, Z), tau_max)
are cut out.
The latter part only holds if remove_missing_upto_maxlag=True.</p>
</dd>
<dt>If cut_off == ‘tau_max’:</dt><dd><ul>
<li><p>tau_max samples are cut off at the beginning.</p></li>
</ul>
<p>This may be useful for modeling by comparing multiple models on
the same samples.
- If at time step t_missing a data value is missing, then the
time steps
t_missing, …, t_missing + max(max_lag(X, Y, Z), tau_max)
are cut out.
The latter part only holds if remove_missing_upto_maxlag=True.</p>
</dd>
<dt>If cut_off == ‘2xtau_max_future’:</dt><dd><p>First, the relevant time steps are determined as for cut_off ==
‘max_lag’. Then, the temporally latest time steps are removed
such that the same number of time steps remains as there would
be for cut_off == ‘2xtau_max’. This may be useful when one is
mostly interested in the temporally first time steps and would
like all MCI tests to be performed on the same <em>number</em> of
samples. Note, however, that while the <em>number</em> of samples is
the same for all MCI tests, the samples themselves may be
different.</p>
</dd>
</dl>
</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Level of verbosity.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>array, xyz [,XYZ]</strong> – xyz identifier array of shape (dim,) identifying which row in array
corresponds to X, Y, and Z. For example:: X = [(0, -1)],
Y = [(1, 0)], Z = [(1, -1), (0, -2)] yields an array of shape
(4, n_samples) and xyz is xyz = numpy.array([0,1,2,2]). If
return_cleaned_xyz is True, also outputs the cleaned XYZ lists.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple of data array of shape (dim, n_samples) and</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tigramite.data_processing.DataFrame.print_array_info">
<span class="sig-name descname"><span class="pre">print_array_info</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Z</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">missing_flag</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extraZ</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/data_processing.html#DataFrame.print_array_info"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.data_processing.DataFrame.print_array_info" title="Permalink to this definition">¶</a></dt>
<dd><p>Print info about the constructed array</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>array</strong> (<em>Data array of shape</em><em> (</em><em>dim</em><em>, </em><em>T</em><em>)</em>) – Data array.</p></li>
<li><p><strong>X</strong> (<em>list of tuples</em>) – For a dependence measure I(X;Y|Z), Y is of the form [(varY, 0)],
where var specifies the variable index. X typically is of the form
[(varX, -tau)] with tau denoting the time lag and Z can be
multivariate [(var1, -lag), (var2, -lag), …] .</p></li>
<li><p><strong>Y</strong> (<em>list of tuples</em>) – For a dependence measure I(X;Y|Z), Y is of the form [(varY, 0)],
where var specifies the variable index. X typically is of the form
[(varX, -tau)] with tau denoting the time lag and Z can be
multivariate [(var1, -lag), (var2, -lag), …] .</p></li>
<li><p><strong>Z</strong> (<em>list of tuples</em>) – For a dependence measure I(X;Y|Z), Y is of the form [(varY, 0)],
where var specifies the variable index. X typically is of the form
[(varX, -tau)] with tau denoting the time lag and Z can be
multivariate [(var1, -lag), (var2, -lag), …] .</p></li>
<li><p><strong>extraZ</strong> (<em>list of tuples</em>) – For a dependence measure I(X;Y|Z), Y is of the form [(varY, 0)],
where var specifies the variable index. X typically is of the form
[(varX, -tau)] with tau denoting the time lag and Z can be
multivariate [(var1, -lag), (var2, -lag), …] .</p></li>
<li><p><strong>missing_flag</strong> (<em>number</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Flag for missing values. Dismisses all time slices of samples where
missing values occur in any variable and also flags samples for all
lags up to 2*tau_max. This avoids biases, see section on masking in
Supplement of <a class="footnote-reference brackets" href="#id3" id="id32" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a>.</p></li>
<li><p><strong>mask_type</strong> (<em>{'y'</em><em>,</em><em>'x'</em><em>,</em><em>'z'</em><em>,</em><em>'xy'</em><em>,</em><em>'xz'</em><em>,</em><em>'yz'</em><em>,</em><em>'xyz'}</em>) – Masking mode: Indicators for which variables in the dependence
measure I(X; Y | Z) the samples should be masked. If None, the mask
is not used. Explained in tutorial on masking and missing values.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tigramite.data_processing.lowhighpass_filter">
<span class="sig-prename descclassname"><span class="pre">tigramite.data_processing.</span></span><span class="sig-name descname"><span class="pre">lowhighpass_filter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutperiod</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pass_periods</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'low'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/data_processing.html#lowhighpass_filter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.data_processing.lowhighpass_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Butterworth low- or high pass filter.</p>
<p>This function applies a linear filter twice, once forward and once
backwards. The combined filter has linear phase.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>array</em>) – Data array of shape (time, variables).</p></li>
<li><p><strong>cutperiod</strong> (<em>int</em>) – Period of cutoff.</p></li>
<li><p><strong>pass_periods</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'low'</em><em>)</em>) – Either ‘low’ or ‘high’ to act as a low- or high-pass filter</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>data</strong> – Filtered data array.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>array</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tigramite.data_processing.ordinal_patt_array">
<span class="sig-prename descclassname"><span class="pre">tigramite.data_processing.</span></span><span class="sig-name descname"><span class="pre">ordinal_patt_array</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">array_mask</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">step</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/data_processing.html#ordinal_patt_array"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.data_processing.ordinal_patt_array" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns symbolified array of ordinal patterns.</p>
<p>Each data vector (X_t, …, X_t+(dim-1)*step) is converted to its rank
vector. E.g., (0.2, -.6, 1.2) –&gt; (1,0,2) which is then assigned to a
unique integer (see Article). There are faculty(dim) possible rank vectors.</p>
<p>Note that the symb_array is step*(dim-1) shorter than the original array!</p>
<p>Reference: B. Pompe and J. Runge (2011). Momentary information transfer as
a coupling measure of time series. Phys. Rev. E, 83(5), 1-12.
doi:10.1103/PhysRevE.83.051122</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>array</strong> (<em>array-like</em>) – Data array of shape (time, variables).</p></li>
<li><p><strong>array_mask</strong> (<em>bool array</em>) – Data mask where True labels masked samples.</p></li>
<li><p><strong>dim</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 2</em><em>)</em>) – Pattern dimension</p></li>
<li><p><strong>step</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 1</em><em>)</em>) – Delay of pattern embedding vector.</p></li>
<li><p><strong>weights</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: False</em><em>)</em>) – Whether to return array of variances of embedding vectors as weights.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Level of verbosity.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>patt, patt_mask [, patt_time]</strong> – Tuple of converted pattern array and new length</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple of arrays</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tigramite.data_processing.quantile_bin_array">
<span class="sig-prename descclassname"><span class="pre">tigramite.data_processing.</span></span><span class="sig-name descname"><span class="pre">quantile_bin_array</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">6</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/data_processing.html#quantile_bin_array"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.data_processing.quantile_bin_array" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns symbolified array with equal-quantile binning.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>array</em>) – Data array of shape (time, variables).</p></li>
<li><p><strong>bins</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 6</em><em>)</em>) – Number of bins.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>symb_array</strong> – Converted data of integer type.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>array</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tigramite.data_processing.smooth">
<span class="sig-prename descclassname"><span class="pre">tigramite.data_processing.</span></span><span class="sig-name descname"><span class="pre">smooth</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smooth_width</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kernel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gaussian'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">residuals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/data_processing.html#smooth"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.data_processing.smooth" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns either smoothed time series or its residuals.</p>
<p>the difference between the original and the smoothed time series
(=residuals) of a kernel smoothing with gaussian (smoothing kernel width =
twice the sigma!) or heaviside window, equivalent to a running mean.</p>
<p>Assumes data of shape (T, N) or (T,)
:rtype: array
:returns: smoothed/residual data</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>array</em>) – Data array of shape (time, variables).</p></li>
<li><p><strong>smooth_width</strong> (<em>float</em>) – Window width of smoothing, 2*sigma for a gaussian.</p></li>
<li><p><strong>kernel</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'gaussian'</em><em>)</em>) – Smoothing kernel, ‘gaussian’ or ‘heaviside’ for a running mean.</p></li>
<li><p><strong>mask</strong> (<em>bool array</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Data mask where True labels masked samples.</p></li>
<li><p><strong>residuals</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: False</em><em>)</em>) – True if residuals should be returned instead of smoothed data.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>data</strong> – Smoothed/residual data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>array-like</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tigramite.data_processing.structural_causal_process">
<span class="sig-prename descclassname"><span class="pre">tigramite.data_processing.</span></span><span class="sig-name descname"><span class="pre">structural_causal_process</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">links</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">T</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">noises</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">intervention</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">intervention_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hard'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/data_processing.html#structural_causal_process"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.data_processing.structural_causal_process" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a structural causal process with contemporaneous and lagged
dependencies.</p>
<p>DEPRECATED. Will be removed in future.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tigramite.data_processing.time_bin_with_mask">
<span class="sig-prename descclassname"><span class="pre">tigramite.data_processing.</span></span><span class="sig-name descname"><span class="pre">time_bin_with_mask</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_bin_length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/data_processing.html#time_bin_with_mask"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.data_processing.time_bin_with_mask" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns time binned data where only about non-masked values is averaged.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>array</em>) – Data array of shape (time, variables).</p></li>
<li><p><strong>time_bin_length</strong> (<em>int</em>) – Length of time bin.</p></li>
<li><p><strong>mask</strong> (<em>bool array</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Data mask where True labels masked samples.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>(bindata, T)</strong> – Tuple of time-binned data array and new length of array.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple of array and int</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tigramite.data_processing.var_process">
<span class="sig-prename descclassname"><span class="pre">tigramite.data_processing.</span></span><span class="sig-name descname"><span class="pre">var_process</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parents_neighbors_coeffs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">T</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'inv_inno_cov'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">initial_values</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/data_processing.html#var_process"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.data_processing.var_process" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a vector-autoregressive process with correlated innovations.</p>
<p>Wrapper around var_network with possibly more user-friendly input options.</p>
<p>DEPRECATED. Will be removed in future.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tigramite.data_processing.weighted_avg_and_std">
<span class="sig-prename descclassname"><span class="pre">tigramite.data_processing.</span></span><span class="sig-name descname"><span class="pre">weighted_avg_and_std</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">axis</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weights</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/data_processing.html#weighted_avg_and_std"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.data_processing.weighted_avg_and_std" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the weighted average and standard deviation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>values</strong> (<em>array</em>) – Data array of shape (time, variables).</p></li>
<li><p><strong>axis</strong> (<em>int</em>) – Axis to average/std about</p></li>
<li><p><strong>weights</strong> (<em>array</em>) – Weight array of shape (time, variables).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>(average, std)</strong> – Tuple of weighted average and standard deviation along axis.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple of arrays</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-tigramite.toymodels.structural_causal_processes">
<span id="tigramite-toymodels-toy-model-generators"></span><h1><code class="xref py py-mod docutils literal notranslate"><span class="pre">tigramite.toymodels</span></code>: Toy model generators<a class="headerlink" href="#module-tigramite.toymodels.structural_causal_processes" title="Permalink to this heading">¶</a></h1>
<p>Tigramite toymodels.</p>
<dl class="py function">
<dt class="sig sig-object py" id="tigramite.toymodels.structural_causal_processes.dag_to_links">
<span class="sig-prename descclassname"><span class="pre">tigramite.toymodels.structural_causal_processes.</span></span><span class="sig-name descname"><span class="pre">dag_to_links</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dag</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/toymodels/structural_causal_processes.html#dag_to_links"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.toymodels.structural_causal_processes.dag_to_links" title="Permalink to this definition">¶</a></dt>
<dd><p>Helper function to convert DAG graph to dictionary of parents.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>dag</strong> (<em>array of shape</em><em> (</em><em>N</em><em>, </em><em>N</em><em>, </em><em>tau_max+1</em><em>)</em>) – Matrix format of graph in string format. Must be DAG.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>parents</strong> – Dictionary of form {0:[(0, -1), …], 1:[…], …}.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tigramite.toymodels.structural_causal_processes.links_to_graph">
<span class="sig-prename descclassname"><span class="pre">tigramite.toymodels.structural_causal_processes.</span></span><span class="sig-name descname"><span class="pre">links_to_graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">links</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/toymodels/structural_causal_processes.html#links_to_graph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.toymodels.structural_causal_processes.links_to_graph" title="Permalink to this definition">¶</a></dt>
<dd><p>Helper function to convert dictionary of links to graph array format.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>links</strong> (<em>dict</em>) – Dictionary of form {0:[((0, -1), coeff, func), …], 1:[…], …}.
Also format {0:[(0, -1), …], 1:[…], …} is allowed.</p></li>
<li><p><strong>tau_max</strong> (<em>int</em><em> or </em><em>None</em>) – Maximum lag. If None, the maximum lag in links is used.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>graph</strong> – Matrix format of graph with 1 for true links and 0 else.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>array of shape (N, N, tau_max+1)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tigramite.toymodels.structural_causal_processes.structural_causal_process">
<span class="sig-prename descclassname"><span class="pre">tigramite.toymodels.structural_causal_processes.</span></span><span class="sig-name descname"><span class="pre">structural_causal_process</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">links</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">T</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">noises</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">intervention</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">intervention_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hard'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transient_fraction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/toymodels/structural_causal_processes.html#structural_causal_process"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.toymodels.structural_causal_processes.structural_causal_process" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a time series generated from a structural causal process.</p>
<p>Allows lagged and contemporaneous dependencies and includes the option
to have intervened variables or particular samples.</p>
<p>The interventional data is in particular useful for generating ground
truth for the CausalEffects class.</p>
<p>In more detail, the method implements a generalized additive noise model process of the form</p>
<div class="math">
<p><img src="_images/math/04effbe50a9c1421627153afe77c1c013193dd90.png" alt="X^j_t = \eta^j_t + \sum_{X^i_{t-\tau}\in \mathcal{P}(X^j_t)}
c^i_{\tau} f^i_{\tau}(X^i_{t-\tau})"/></p>
</div><p>Links have the format <code class="docutils literal notranslate"><span class="pre">{0:[((i,</span> <span class="pre">-tau),</span> <span class="pre">coeff,</span> <span class="pre">func),...],</span> <span class="pre">1:[...],</span>
<span class="pre">...}</span></code> where <code class="docutils literal notranslate"><span class="pre">func</span></code> can be an arbitrary (nonlinear) function provided
as a python callable with one argument and coeff is the multiplication
factor. The noise distributions of <img class="math" src="_images/math/be38eb4004ec8179f569f36692f9132289acf8b1.png" alt="\eta^j"/> can be specified in
<code class="docutils literal notranslate"><span class="pre">noises</span></code>.</p>
<p>Through the parameters <code class="docutils literal notranslate"><span class="pre">intervention</span></code> and <code class="docutils literal notranslate"><span class="pre">intervention_type</span></code> the model
can also be generated with intervened variables.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>links</strong> (<em>dict</em>) – Dictionary of format: {0:[((i, -tau), coeff, func),…], 1:[…],
…} for all variables where i must be in [0..N-1] and tau &gt;= 0 with
number of variables N. coeff must be a float and func a python
callable of one argument.</p></li>
<li><p><strong>T</strong> (<em>int</em>) – Sample size.</p></li>
<li><p><strong>noises</strong> (<em>list of callables</em><em> or </em><em>array</em><em>, </em><em>optional</em><em> (</em><em>default: 'np.random.randn'</em><em>)</em>) – Random distribution function that is called with noises[j](T). If an array,
it must be of shape ((transient_fraction + 1)*T, N).</p></li>
<li><p><strong>intervention</strong> (<em>dict</em>) – Dictionary of format: {1:np.array, …} containing only keys of intervened
variables with the value being the array of length T with interventional values.
Set values to np.nan to leave specific time points of a variable un-intervened.</p></li>
<li><p><strong>intervention_type</strong> (<em>str</em><em> or </em><em>dict</em>) – Dictionary of format: {1:’hard’,  3:’soft’, …} to specify whether intervention is
hard (set value) or soft (add value) for variable j. If str, all interventions have
the same type.</p></li>
<li><p><strong>transient_fraction</strong> (<em>float</em>) – Added percentage of T used as a transient. In total a realization of length
(transient_fraction + 1)*T will be generated, but then transient_fraction*T will be
cut off.</p></li>
<li><p><strong>seed</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Random seed.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>data</strong> (<em>array-like</em>) – Data generated from this process, shape (T, N).</p></li>
<li><p><strong>nonstationary</strong> (<em>bool</em>) – Indicates whether data has NaNs or infinities.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tigramite.toymodels.structural_causal_processes.var_process">
<span class="sig-prename descclassname"><span class="pre">tigramite.toymodels.structural_causal_processes.</span></span><span class="sig-name descname"><span class="pre">var_process</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parents_neighbors_coeffs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">T</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'inv_inno_cov'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">initial_values</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tigramite/toymodels/structural_causal_processes.html#var_process"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tigramite.toymodels.structural_causal_processes.var_process" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a vector-autoregressive process with correlated innovations.</p>
<p>Wrapper around var_network with possibly more user-friendly input options.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>parents_neighbors_coeffs</strong> (<em>dict</em>) – Dictionary of format: {…, j:[((var1, lag1), coef1), ((var2, lag2),
coef2), …], …} for all variables where vars must be in [0..N-1]
and lags &lt;= 0 with number of variables N. If lag=0, a nonzero value
in the covariance matrix (or its inverse) is implied. These should be
the same for (i, j) and (j, i).</p></li>
<li><p><strong>use</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'inv_inno_cov'</em><em>)</em>) – Specifier, either ‘inno_cov’ or ‘inv_inno_cov’.
Any other specifier will result in non-correlated noise.
For debugging, ‘no_noise’ can also be specified, in which case random
noise will be disabled.</p></li>
<li><p><strong>T</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 1000</em><em>)</em>) – Sample size.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Level of verbosity.</p></li>
<li><p><strong>initial_values</strong> (<em>array</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Initial values for each node. Shape must be (N, max_delay+1)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>data</strong> (<em>array-like</em>) – Data generated from this process</p></li>
<li><p><strong>true_parent_neighbor</strong> (<em>dict</em>) – Dictionary of lists of tuples.  The dictionary is keyed by node ID, the
list stores the tuple values (parent_node_id, time_lag)</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="tigramite-plotting-plotting-functions">
<h1><code class="xref py py-mod docutils literal notranslate"><span class="pre">tigramite.plotting</span></code>: Plotting functions<a class="headerlink" href="#tigramite-plotting-plotting-functions" title="Permalink to this heading">¶</a></h1>
</section>
<section id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this heading">¶</a></h1>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <div>
    <h3><a href="#">Table of Contents</a></h3>
    <ul>
<li><a class="reference internal" href="#">TIGRAMITE</a></li>
<li><a class="reference internal" href="#tigramite-pcmci-pcmci"><code class="xref py py-mod docutils literal notranslate"><span class="pre">tigramite.pcmci</span></code>: PCMCI</a></li>
<li><a class="reference internal" href="#tigramite-lpcmci-lpcmci"><code class="xref py py-mod docutils literal notranslate"><span class="pre">tigramite.lpcmci</span></code>: LPCMCI</a></li>
<li><a class="reference internal" href="#tigramite-independence-tests-conditional-independence-tests"><code class="xref py py-mod docutils literal notranslate"><span class="pre">tigramite.independence_tests</span></code>: Conditional independence tests</a></li>
<li><a class="reference internal" href="#tigramite-causal-effects-causal-effect-analysis"><code class="xref py py-mod docutils literal notranslate"><span class="pre">tigramite.causal_effects</span></code>: Causal Effect analysis</a></li>
<li><a class="reference internal" href="#tigramite-models-time-series-modeling-mediation-and-prediction"><code class="xref py py-mod docutils literal notranslate"><span class="pre">tigramite.models</span></code>: Time series modeling, mediation, and prediction</a></li>
<li><a class="reference internal" href="#module-tigramite.data_processing"><code class="xref py py-mod docutils literal notranslate"><span class="pre">tigramite.data_processing</span></code>: Data processing functions</a></li>
<li><a class="reference internal" href="#module-tigramite.toymodels.structural_causal_processes"><code class="xref py py-mod docutils literal notranslate"><span class="pre">tigramite.toymodels</span></code>: Toy model generators</a></li>
<li><a class="reference internal" href="#tigramite-plotting-plotting-functions"><code class="xref py py-mod docutils literal notranslate"><span class="pre">tigramite.plotting</span></code>: Plotting functions</a></li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>

  </div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/index.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">Tigramite 5.1 documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">TIGRAMITE</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2021, Jakob Runge.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 5.0.2.
    </div>
  </body>
</html>